{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "abfa86c4-b036-463d-aa86-e78b1d3bf1d6",
   "metadata": {},
   "source": [
    "# Simulation Data Gene Program Inference with BANKSY"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5972a05a-6ee4-401f-8f3b-f5948017b73b",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (sebastian.birk@helmholtz-munich.de)\n",
    "- **Date of Creation:** 12.07.2024\n",
    "- **Date of Last Modification:** 13.08.2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c124fc61-369f-46ed-ba56-c697d7153fe8",
   "metadata": {},
   "source": [
    "- Perform gene set enrichment analysis with Decoupler as described at https://www.sc-best-practices.org/conditions/gsea_pathway.html#cluster-level-gene-set-enrichment-analysis-with-decoupler"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caaacf02-aee2-4eef-a713-afe6a57d975a",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0e77fbd-b371-4ed7-8a7c-35a96013c4c5",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1ece400f-cdad-4dfe-af37-9e61e31b4854",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "36629fc4-4928-4d41-9f88-126e5f8b4002",
   "metadata": {},
   "outputs": [],
   "source": [
    "import decoupler\n",
    "import liana as li\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from sklearn.metrics import f1_score\n",
    "from nichecompass.models import NicheCompass\n",
    "from nichecompass.utils import (extract_gp_dict_from_mebocost_ms_interactions,\n",
    "                                extract_gp_dict_from_nichenet_lrt_interactions,\n",
    "                                extract_gp_dict_from_omnipath_lr_interactions,\n",
    "                                filter_and_combine_gp_dict_gps_v2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f14b6a9-91af-45aa-95ba-e7a8ce541b20",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b4c5db9f-370f-4508-b676-c4ee1e4ea874",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "species = \"mouse\"\n",
    "sample_key = \"niche_type\"\n",
    "groupby = \"cell_type\"\n",
    "differential_gp_test_results_key = \"nichecompass_differential_gp_test_results\"\n",
    "gp_names_key = \"nichecompass_gp_names\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c944b9d-96e0-4e08-b42f-75c4aff95e38",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "274fa353-9717-47de-8535-ffb0e14138c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(6, 6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2c0e0fc-5e05-49ba-89e7-a929adc1398f",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths and Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "df8c140d-d5aa-4885-bd6b-925637886c6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ga_data_folder_path = \"../../../datasets/ga_data\"\n",
    "gp_data_folder_path = \"../../../datasets/gp_data\"\n",
    "st_data_folder_path = \"../../../datasets/st_data\"\n",
    "st_data_gold_folder_path = st_data_folder_path + \"/gold\"\n",
    "omnipath_lr_network_file_path = f\"{gp_data_folder_path}/omnipath_lr_network.csv\"\n",
    "nichenet_lr_network_file_path = f\"{gp_data_folder_path}/nichenet_lr_network_v2_{species}.csv\"\n",
    "nichenet_ligand_target_matrix_file_path = f\"{gp_data_folder_path}/nichenet_ligand_target_matrix_v2_{species}.csv\"\n",
    "mebocost_enzyme_sensor_interactions_folder_path = f\"{gp_data_folder_path}/metabolite_enzyme_sensor_gps\"\n",
    "gene_orthologs_mapping_file_path = f\"{ga_data_folder_path}/human_mouse_gene_orthologs.csv\"\n",
    "artifacts_folder_path = f\"../../../artifacts\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35e28759-bc54-4ffc-8b55-89063434b34e",
   "metadata": {},
   "source": [
    "### 1.5 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "98266299-c6da-42a6-8c03-db265166e146",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_f1(set1, set2):\n",
    "    # Convert sets to binary vectors\n",
    "    all_elements = list(set1.union(set2))\n",
    "    vec1 = [1 if elem in set1 else 0 for elem in all_elements]\n",
    "    vec2 = [1 if elem in set2 else 0 for elem in all_elements]\n",
    "    \n",
    "    # Compute F1 score\n",
    "    return f1_score(vec1, vec2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "851bbb7a-7c25-4004-a1b4-cae69fc23719",
   "metadata": {},
   "source": [
    "## 2. GP Inference\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "205de08d-7aec-480f-84bb-5eacb9f95503",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 2.1 Load Cell Embeddings from BANKSY\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "027a1cb8-980b-4eb6-958a-61e88045fd39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get embeddings from BANKSY\n",
    "adata_annotated_banksy = sc.read_h5ad(f\"{artifacts_folder_path}/single_sample_method_benchmarking/sim1_1105genes_10000locs_strongincrements_banksy.h5ad\")\n",
    "\n",
    "# Get raw gene expression counts\n",
    "adata = sc.read_h5ad(f\"{st_data_gold_folder_path}/sim1_1105genes_10000locs_strongincrements.h5ad\")\n",
    "\n",
    "for run_number in range(1,9):\n",
    "    adata.obsm[f\"banksy_latent_run{run_number}\"] = adata_annotated_banksy.obsm[f\"banksy_latent_run{run_number}\"]\n",
    "del(adata_annotated_banksy)\n",
    "\n",
    "# Preprocess counts\n",
    "sc.pp.normalize_total(adata)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5975e2d8-c183-431d-952b-90590520a501",
   "metadata": {},
   "source": [
    "### 2.2 Get NicheCompass GPs for Use as Prior Knowledge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "45979bc8-9e57-48e7-a94c-8cfce8209bf7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/ictstr01/home/aih/sebastian.birk/workspace/projects/nichecompass/src/nichecompass/utils/gene_programs.py:506: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Combining ACE2_ligand_receptor_GP and Ace2_ligand_receptor_target_gene_GP.\n",
      "Combining ACE2_ligand_receptor_GP and L-Phenylalanine_metabolite_enzyme_sensor_GP.\n",
      "Combining ADCYAP1_ligand_receptor_GP and Adcyap1_ligand_receptor_target_gene_GP.\n",
      "Combining ADGRE5_ligand_receptor_GP and Adgre5_ligand_receptor_target_gene_GP.\n",
      "Combining ADIPOQ_ligand_receptor_GP and Adipoq_ligand_receptor_target_gene_GP.\n",
      "Combining ADM_ligand_receptor_GP and Adm_ligand_receptor_target_gene_GP.\n",
      "Combining ADM2_ligand_receptor_GP and Adm2_ligand_receptor_target_gene_GP.\n",
      "Combining AGRP_ligand_receptor_GP and Agrp_ligand_receptor_target_gene_GP.\n",
      "Combining AGT_ligand_receptor_GP and Agt_ligand_receptor_target_gene_GP.\n",
      "Combining ALCAM_ligand_receptor_GP and Alcam_ligand_receptor_target_gene_GP.\n",
      "Combining ALKAL1_ligand_receptor_GP and Alkal1_ligand_receptor_target_gene_GP.\n",
      "Combining ALKAL2_ligand_receptor_GP and Alkal2_ligand_receptor_target_gene_GP.\n",
      "Combining AMBN_ligand_receptor_GP and Ambn_ligand_receptor_target_gene_GP.\n",
      "Combining AMELX_ligand_receptor_GP and Amelx_ligand_receptor_target_gene_GP.\n",
      "Combining AMH_ligand_receptor_GP and Amh_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPT1_ligand_receptor_GP and Angpt1_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPT2_ligand_receptor_GP and Angpt2_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPT4_ligand_receptor_GP and Angpt4_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPTL1_ligand_receptor_GP and Angptl1_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPTL2_ligand_receptor_GP and Angptl2_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPTL3_ligand_receptor_GP and Angptl3_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPTL4_ligand_receptor_GP and Angptl4_ligand_receptor_target_gene_GP.\n",
      "Combining ANXA1_ligand_receptor_GP and Anxa1_ligand_receptor_target_gene_GP.\n",
      "Combining ANXA2_ligand_receptor_GP and Anxa2_ligand_receptor_target_gene_GP.\n",
      "Combining APLN_ligand_receptor_GP and Apln_ligand_receptor_target_gene_GP.\n",
      "Combining APOB_ligand_receptor_GP and Apob_ligand_receptor_target_gene_GP.\n",
      "Combining APOD_ligand_receptor_GP and Apod_ligand_receptor_target_gene_GP.\n",
      "Combining APOE_ligand_receptor_GP and Apoe_ligand_receptor_target_gene_GP.\n",
      "Combining AREG_ligand_receptor_GP and Areg_ligand_receptor_target_gene_GP.\n",
      "Combining ARTN_ligand_receptor_GP and Artn_ligand_receptor_target_gene_GP.\n",
      "Combining AVP_ligand_receptor_GP and Avp_ligand_receptor_target_gene_GP.\n",
      "Combining AZGP1_ligand_receptor_GP and Azgp1_ligand_receptor_target_gene_GP.\n",
      "Combining BDNF_ligand_receptor_GP and Bdnf_ligand_receptor_target_gene_GP.\n",
      "Combining BMP10_ligand_receptor_GP and Bmp10_ligand_receptor_target_gene_GP.\n",
      "Combining BMP15_ligand_receptor_GP and Bmp15_ligand_receptor_target_gene_GP.\n",
      "Combining BMP2_ligand_receptor_GP and Bmp2_ligand_receptor_target_gene_GP.\n",
      "Combining BMP3_ligand_receptor_GP and Bmp3_ligand_receptor_target_gene_GP.\n",
      "Combining BMP4_ligand_receptor_GP and Bmp4_ligand_receptor_target_gene_GP.\n",
      "Combining BMP5_ligand_receptor_GP and Bmp5_ligand_receptor_target_gene_GP.\n",
      "Combining BMP6_ligand_receptor_GP and Bmp6_ligand_receptor_target_gene_GP.\n",
      "Combining BMP7_ligand_receptor_GP and Bmp7_ligand_receptor_target_gene_GP.\n",
      "Combining BMP8A_ligand_receptor_GP and Bmp8a_ligand_receptor_target_gene_GP.\n",
      "Combining BST2_ligand_receptor_GP and Bst2_ligand_receptor_target_gene_GP.\n",
      "Combining BTC_ligand_receptor_GP and Btc_ligand_receptor_target_gene_GP.\n",
      "Combining C1QTNF1_ligand_receptor_GP and C1qtnf1_ligand_receptor_target_gene_GP.\n",
      "Combining C1QTNF5_ligand_receptor_GP and C1qtnf5_ligand_receptor_target_gene_GP.\n",
      "Combining C4B_ligand_receptor_GP and C4b_ligand_receptor_target_gene_GP.\n",
      "Combining CALCA_ligand_receptor_GP and Calca_ligand_receptor_target_gene_GP.\n",
      "Combining CALCB_ligand_receptor_GP and Calcb_ligand_receptor_target_gene_GP.\n",
      "Combining CALR_ligand_receptor_GP and Calr_ligand_receptor_target_gene_GP.\n",
      "Combining CCK_ligand_receptor_GP and Cck_ligand_receptor_target_gene_GP.\n",
      "Combining CCL1_ligand_receptor_GP and Ccl1_ligand_receptor_target_gene_GP.\n",
      "Combining CCL11_ligand_receptor_GP and Ccl11_ligand_receptor_target_gene_GP.\n",
      "Combining CCL17_ligand_receptor_GP and Ccl17_ligand_receptor_target_gene_GP.\n",
      "Combining CCL18_ligand_receptor_GP and CCL3_ligand_receptor_GP.\n",
      "Combining CCL18_ligand_receptor_GP and Ccl3_ligand_receptor_target_gene_GP.\n",
      "Combining CCL19_ligand_receptor_GP and Ccl19_ligand_receptor_target_gene_GP.\n",
      "Combining CCL2_ligand_receptor_GP and Ccl2_ligand_receptor_target_gene_GP.\n",
      "Combining CCL20_ligand_receptor_GP and Ccl20_ligand_receptor_target_gene_GP.\n",
      "Combining CCL22_ligand_receptor_GP and Ccl22_ligand_receptor_target_gene_GP.\n",
      "Combining CCL24_ligand_receptor_GP and Ccl24_ligand_receptor_target_gene_GP.\n",
      "Combining CCL25_ligand_receptor_GP and Ccl25_ligand_receptor_target_gene_GP.\n",
      "Combining CCL26_ligand_receptor_GP and Ccl26_ligand_receptor_target_gene_GP.\n",
      "Combining CCL28_ligand_receptor_GP and Ccl28_ligand_receptor_target_gene_GP.\n",
      "Combining CCL3_ligand_receptor_GP and CCL18_ligand_receptor_GP.\n",
      "Combining CCL3_ligand_receptor_GP and Ccl3_ligand_receptor_target_gene_GP.\n",
      "Combining CCL4_ligand_receptor_GP and CCL4L2_ligand_receptor_GP.\n",
      "Combining CCL4_ligand_receptor_GP and Ccl4_ligand_receptor_target_gene_GP.\n",
      "Combining CCL4L2_ligand_receptor_GP and CCL4_ligand_receptor_GP.\n",
      "Combining CCL4L2_ligand_receptor_GP and Ccl4_ligand_receptor_target_gene_GP.\n",
      "Combining CCL5_ligand_receptor_GP and Ccl5_ligand_receptor_target_gene_GP.\n",
      "Combining CCL7_ligand_receptor_GP and Ccl7_ligand_receptor_target_gene_GP.\n",
      "Combining CCL8_ligand_receptor_GP and Ccl8_ligand_receptor_target_gene_GP.\n",
      "Combining CCN1_ligand_receptor_GP and Ccn1_ligand_receptor_target_gene_GP.\n",
      "Combining CCN2_ligand_receptor_GP and Ccn2_ligand_receptor_target_gene_GP.\n",
      "Combining CCN3_ligand_receptor_GP and Ccn3_ligand_receptor_target_gene_GP.\n",
      "Combining CD160_ligand_receptor_GP and Cd160_ligand_receptor_target_gene_GP.\n",
      "Combining CD200_ligand_receptor_GP and Cd200_ligand_receptor_target_gene_GP.\n",
      "Combining CD274_ligand_receptor_GP and Cd274_ligand_receptor_target_gene_GP.\n",
      "Combining CD34_ligand_receptor_GP and Cd34_ligand_receptor_target_gene_GP.\n",
      "Combining CD40LG_ligand_receptor_GP and Cd40lg_ligand_receptor_target_gene_GP.\n",
      "Combining CD48_ligand_receptor_GP and Cd48_ligand_receptor_target_gene_GP.\n",
      "Combining CD52_ligand_receptor_GP and Cd52_ligand_receptor_target_gene_GP.\n",
      "Combining CD5L_ligand_receptor_GP and Cd5l_ligand_receptor_target_gene_GP.\n",
      "Combining CD70_ligand_receptor_GP and Cd70_ligand_receptor_target_gene_GP.\n",
      "Combining CD72_ligand_receptor_GP and Cd72_ligand_receptor_target_gene_GP.\n",
      "Combining CD80_ligand_receptor_GP and Cd80_ligand_receptor_target_gene_GP.\n",
      "Combining CD99_ligand_receptor_GP and Cd99_ligand_receptor_target_gene_GP.\n",
      "Combining CEACAM5_ligand_receptor_GP and Ceacam5_ligand_receptor_target_gene_GP.\n",
      "Combining CGA_ligand_receptor_GP and Cga_ligand_receptor_target_gene_GP.\n",
      "Combining CHAD_ligand_receptor_GP and Chad_ligand_receptor_target_gene_GP.\n",
      "Combining CKLF_ligand_receptor_GP and Cklf_ligand_receptor_target_gene_GP.\n",
      "Combining CLCF1_ligand_receptor_GP and Clcf1_ligand_receptor_target_gene_GP.\n",
      "Combining CLDN1_ligand_receptor_GP and Cldn1_ligand_receptor_target_gene_GP.\n",
      "Combining CLDN16_ligand_receptor_GP and Cldn16_ligand_receptor_target_gene_GP.\n",
      "Combining CLEC4G_ligand_receptor_GP and Clec4g_ligand_receptor_target_gene_GP.\n",
      "Combining CLU_ligand_receptor_GP and Clu_ligand_receptor_target_gene_GP.\n",
      "Combining CMTM8_ligand_receptor_GP and Cmtm8_ligand_receptor_target_gene_GP.\n",
      "Combining CNTN1_ligand_receptor_GP and Cntn1_ligand_receptor_target_gene_GP.\n",
      "Combining CNTN2_ligand_receptor_GP and Cntn2_ligand_receptor_target_gene_GP.\n",
      "Combining CNTN3_ligand_receptor_GP and Cntn3_ligand_receptor_target_gene_GP.\n",
      "Combining CNTN4_ligand_receptor_GP and Cntn4_ligand_receptor_target_gene_GP.\n",
      "Combining CNTN6_ligand_receptor_GP and Cntn6_ligand_receptor_target_gene_GP.\n",
      "Combining COL10A1_ligand_receptor_GP and Col10a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL11A2_ligand_receptor_GP and Col11a2_ligand_receptor_target_gene_GP.\n",
      "Combining COL12A1_ligand_receptor_GP and Col12a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL13A1_ligand_receptor_GP and Col13a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL15A1_ligand_receptor_GP and Col15a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL16A1_ligand_receptor_GP and Col16a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL17A1_ligand_receptor_GP and Col17a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL19A1_ligand_receptor_GP and Col19a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL20A1_ligand_receptor_GP and Col20a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL22A1_ligand_receptor_GP and Col22a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL24A1_ligand_receptor_GP and Col24a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL26A1_ligand_receptor_GP and Col26a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL27A1_ligand_receptor_GP and Col27a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL28A1_ligand_receptor_GP and Col28a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL6A3_ligand_receptor_GP and Col6a3_ligand_receptor_target_gene_GP.\n",
      "Combining COL6A5_ligand_receptor_GP and Col6a5_ligand_receptor_target_gene_GP.\n",
      "Combining COL6A6_ligand_receptor_GP and Col6a6_ligand_receptor_target_gene_GP.\n",
      "Combining COL8A2_ligand_receptor_GP and Col8a2_ligand_receptor_target_gene_GP.\n",
      "Combining COMP_ligand_receptor_GP and Comp_ligand_receptor_target_gene_GP.\n",
      "Combining CORT_ligand_receptor_GP and Cort_ligand_receptor_target_gene_GP.\n",
      "Combining CRH_ligand_receptor_GP and Crh_ligand_receptor_target_gene_GP.\n",
      "Combining CRLF1_ligand_receptor_GP and Crlf1_ligand_receptor_target_gene_GP.\n",
      "Combining CSF1_ligand_receptor_GP and Csf1_ligand_receptor_target_gene_GP.\n",
      "Combining CSF2_ligand_receptor_GP and Csf2_ligand_receptor_target_gene_GP.\n",
      "Combining CSF3_ligand_receptor_GP and Csf3_ligand_receptor_target_gene_GP.\n",
      "Combining CSH1_ligand_receptor_GP and GH1_ligand_receptor_GP.\n",
      "Combining CSH1_ligand_receptor_GP and GH2_ligand_receptor_GP.\n",
      "Combining CSH1_ligand_receptor_GP and Gh_ligand_receptor_target_gene_GP.\n",
      "Combining CTF1_ligand_receptor_GP and Ctf1_ligand_receptor_target_gene_GP.\n",
      "Combining CTSG_ligand_receptor_GP and Ctsg_ligand_receptor_target_gene_GP.\n",
      "Combining CX3CL1_ligand_receptor_GP and Cx3cl1_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL1_ligand_receptor_GP and Cxcl1_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL10_ligand_receptor_GP and Cxcl10_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL11_ligand_receptor_GP and Cxcl11_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL12_ligand_receptor_GP and Cxcl12_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL13_ligand_receptor_GP and Cxcl13_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL14_ligand_receptor_GP and Cxcl14_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL17_ligand_receptor_GP and Cxcl17_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL2_ligand_receptor_GP and Cxcl2_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL5_ligand_receptor_GP and CXCL6_ligand_receptor_GP.\n",
      "Combining CXCL5_ligand_receptor_GP and Cxcl5_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL6_ligand_receptor_GP and CXCL5_ligand_receptor_GP.\n",
      "Combining CXCL6_ligand_receptor_GP and Cxcl5_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL9_ligand_receptor_GP and Cxcl9_ligand_receptor_target_gene_GP.\n",
      "Combining DCN_ligand_receptor_GP and Dcn_ligand_receptor_target_gene_GP.\n",
      "Combining DEFB103A_ligand_receptor_GP and Defb14_ligand_receptor_target_gene_GP.\n",
      "Combining DHH_ligand_receptor_GP and Dhh_ligand_receptor_target_gene_GP.\n",
      "Combining DKK1_ligand_receptor_GP and Dkk1_ligand_receptor_target_gene_GP.\n",
      "Combining DKK2_ligand_receptor_GP and Dkk2_ligand_receptor_target_gene_GP.\n",
      "Combining DKK4_ligand_receptor_GP and Dkk4_ligand_receptor_target_gene_GP.\n",
      "Combining DLK1_ligand_receptor_GP and Dlk1_ligand_receptor_target_gene_GP.\n",
      "Combining DLK2_ligand_receptor_GP and Dlk2_ligand_receptor_target_gene_GP.\n",
      "Combining DLL1_ligand_receptor_GP and Dll1_ligand_receptor_target_gene_GP.\n",
      "Combining DLL3_ligand_receptor_GP and Dll3_ligand_receptor_target_gene_GP.\n",
      "Combining DLL4_ligand_receptor_GP and Dll4_ligand_receptor_target_gene_GP.\n",
      "Combining DRAXIN_ligand_receptor_GP and Draxin_ligand_receptor_target_gene_GP.\n",
      "Combining DSC1_ligand_receptor_GP and Dsc1_ligand_receptor_target_gene_GP.\n",
      "Combining DSC2_ligand_receptor_GP and Dsc2_ligand_receptor_target_gene_GP.\n",
      "Combining DSC3_ligand_receptor_GP and Dsc3_ligand_receptor_target_gene_GP.\n",
      "Combining EBI3_ligand_receptor_GP and Ebi3_ligand_receptor_target_gene_GP.\n",
      "Combining EDA_ligand_receptor_GP and Eda_ligand_receptor_target_gene_GP.\n",
      "Combining EDIL3_ligand_receptor_GP and Edil3_ligand_receptor_target_gene_GP.\n",
      "Combining EDN1_ligand_receptor_GP and Edn1_ligand_receptor_target_gene_GP.\n",
      "Combining EDN2_ligand_receptor_GP and Edn2_ligand_receptor_target_gene_GP.\n",
      "Combining EDN3_ligand_receptor_GP and Edn3_ligand_receptor_target_gene_GP.\n",
      "Combining EFEMP1_ligand_receptor_GP and Efemp1_ligand_receptor_target_gene_GP.\n",
      "Combining EFEMP2_ligand_receptor_GP and Efemp2_ligand_receptor_target_gene_GP.\n",
      "Combining EFNA1_ligand_receptor_GP and Efna1_ligand_receptor_target_gene_GP.\n",
      "Combining EFNA2_ligand_receptor_GP and Efna2_ligand_receptor_target_gene_GP.\n",
      "Combining EFNA3_ligand_receptor_GP and Efna3_ligand_receptor_target_gene_GP.\n",
      "Combining EFNA4_ligand_receptor_GP and Efna4_ligand_receptor_target_gene_GP.\n",
      "Combining EFNA5_ligand_receptor_GP and Efna5_ligand_receptor_target_gene_GP.\n",
      "Combining EFNB1_ligand_receptor_GP and Efnb1_ligand_receptor_target_gene_GP.\n",
      "Combining EFNB2_ligand_receptor_GP and Efnb2_ligand_receptor_target_gene_GP.\n",
      "Combining EFNB3_ligand_receptor_GP and Efnb3_ligand_receptor_target_gene_GP.\n",
      "Combining EGF_ligand_receptor_GP and Egf_ligand_receptor_target_gene_GP.\n",
      "Combining ENHO_ligand_receptor_GP and Enho_ligand_receptor_target_gene_GP.\n",
      "Combining EPGN_ligand_receptor_GP and Epgn_ligand_receptor_target_gene_GP.\n",
      "Combining EPO_ligand_receptor_GP and Epo_ligand_receptor_target_gene_GP.\n",
      "Combining EREG_ligand_receptor_GP and Ereg_ligand_receptor_target_gene_GP.\n",
      "Combining FASLG_ligand_receptor_GP and Fasl_ligand_receptor_target_gene_GP.\n",
      "Combining FBLN1_ligand_receptor_GP and Fbln1_ligand_receptor_target_gene_GP.\n",
      "Combining FBLN2_ligand_receptor_GP and Fbln2_ligand_receptor_target_gene_GP.\n",
      "Combining FGF1_ligand_receptor_GP and Fgf1_ligand_receptor_target_gene_GP.\n",
      "Combining FGF10_ligand_receptor_GP and Fgf10_ligand_receptor_target_gene_GP.\n",
      "Combining FGF11_ligand_receptor_GP and Fgf11_ligand_receptor_target_gene_GP.\n",
      "Combining FGF12_ligand_receptor_GP and Fgf12_ligand_receptor_target_gene_GP.\n",
      "Combining FGF13_ligand_receptor_GP and Fgf13_ligand_receptor_target_gene_GP.\n",
      "Combining FGF14_ligand_receptor_GP and Fgf14_ligand_receptor_target_gene_GP.\n",
      "Combining FGF16_ligand_receptor_GP and Fgf16_ligand_receptor_target_gene_GP.\n",
      "Combining FGF17_ligand_receptor_GP and Fgf17_ligand_receptor_target_gene_GP.\n",
      "Combining FGF18_ligand_receptor_GP and Fgf18_ligand_receptor_target_gene_GP.\n",
      "Combining FGF19_ligand_receptor_GP and Fgf15_ligand_receptor_target_gene_GP.\n",
      "Combining FGF2_ligand_receptor_GP and Fgf2_ligand_receptor_target_gene_GP.\n",
      "Combining FGF20_ligand_receptor_GP and Fgf20_ligand_receptor_target_gene_GP.\n",
      "Combining FGF21_ligand_receptor_GP and Fgf21_ligand_receptor_target_gene_GP.\n",
      "Combining FGF22_ligand_receptor_GP and Fgf22_ligand_receptor_target_gene_GP.\n",
      "Combining FGF23_ligand_receptor_GP and Fgf23_ligand_receptor_target_gene_GP.\n",
      "Combining FGF3_ligand_receptor_GP and Fgf3_ligand_receptor_target_gene_GP.\n",
      "Combining FGF4_ligand_receptor_GP and Fgf4_ligand_receptor_target_gene_GP.\n",
      "Combining FGF5_ligand_receptor_GP and Fgf5_ligand_receptor_target_gene_GP.\n",
      "Combining FGF6_ligand_receptor_GP and Fgf6_ligand_receptor_target_gene_GP.\n",
      "Combining FGF7_ligand_receptor_GP and Fgf7_ligand_receptor_target_gene_GP.\n",
      "Combining FGF8_ligand_receptor_GP and Fgf8_ligand_receptor_target_gene_GP.\n",
      "Combining FGF9_ligand_receptor_GP and Fgf9_ligand_receptor_target_gene_GP.\n",
      "Combining FGL1_ligand_receptor_GP and Fgl1_ligand_receptor_target_gene_GP.\n",
      "Combining FLRT3_ligand_receptor_GP and Flrt3_ligand_receptor_target_gene_GP.\n",
      "Combining FLT3LG_ligand_receptor_GP and Flt3l_ligand_receptor_target_gene_GP.\n",
      "Combining FSHB_ligand_receptor_GP and Fshb_ligand_receptor_target_gene_GP.\n",
      "Combining GAL_ligand_receptor_GP and Gal_ligand_receptor_target_gene_GP.\n",
      "Combining GALP_ligand_receptor_GP and Galp_ligand_receptor_target_gene_GP.\n",
      "Combining GAS6_ligand_receptor_GP and Gas6_ligand_receptor_target_gene_GP.\n",
      "Combining GAST_ligand_receptor_GP and Gast_ligand_receptor_target_gene_GP.\n",
      "Combining GC_ligand_receptor_GP and Gc_ligand_receptor_target_gene_GP.\n",
      "Combining GCG_ligand_receptor_GP and Gcg_ligand_receptor_target_gene_GP.\n",
      "Combining GDF10_ligand_receptor_GP and Gdf10_ligand_receptor_target_gene_GP.\n",
      "Combining GDF11_ligand_receptor_GP and Gdf11_ligand_receptor_target_gene_GP.\n",
      "Combining GDF15_ligand_receptor_GP and Gdf15_ligand_receptor_target_gene_GP.\n",
      "Combining GDF2_ligand_receptor_GP and Gdf2_ligand_receptor_target_gene_GP.\n",
      "Combining GDF3_ligand_receptor_GP and Gdf3_ligand_receptor_target_gene_GP.\n",
      "Combining GDF5_ligand_receptor_GP and Gdf5_ligand_receptor_target_gene_GP.\n",
      "Combining GDF6_ligand_receptor_GP and Gdf6_ligand_receptor_target_gene_GP.\n",
      "Combining GDF7_ligand_receptor_GP and Gdf7_ligand_receptor_target_gene_GP.\n",
      "Combining GDF9_ligand_receptor_GP and Gdf9_ligand_receptor_target_gene_GP.\n",
      "Combining GDNF_ligand_receptor_GP and Gdnf_ligand_receptor_target_gene_GP.\n",
      "Combining GH1_ligand_receptor_GP and CSH1_ligand_receptor_GP.\n",
      "Combining GH1_ligand_receptor_GP and GH2_ligand_receptor_GP.\n",
      "Combining GH1_ligand_receptor_GP and Gh_ligand_receptor_target_gene_GP.\n",
      "Combining GH2_ligand_receptor_GP and CSH1_ligand_receptor_GP.\n",
      "Combining GH2_ligand_receptor_GP and GH1_ligand_receptor_GP.\n",
      "Combining GH2_ligand_receptor_GP and Gh_ligand_receptor_target_gene_GP.\n",
      "Combining GHRH_ligand_receptor_GP and Ghrh_ligand_receptor_target_gene_GP.\n",
      "Combining GHRL_ligand_receptor_GP and Ghrl_ligand_receptor_target_gene_GP.\n",
      "Combining GIP_ligand_receptor_GP and Gip_ligand_receptor_target_gene_GP.\n",
      "Combining GNRH1_ligand_receptor_GP and Gnrh1_ligand_receptor_target_gene_GP.\n",
      "Combining GPHA2_ligand_receptor_GP and Gpha2_ligand_receptor_target_gene_GP.\n",
      "Combining GPR15LG_ligand_receptor_GP and 2610528A11Rik_ligand_receptor_target_gene_GP.\n",
      "Combining GREM1_ligand_receptor_GP and Grem1_ligand_receptor_target_gene_GP.\n",
      "Combining GRN_ligand_receptor_GP and Grn_ligand_receptor_target_gene_GP.\n",
      "Combining GRP_ligand_receptor_GP and Grp_ligand_receptor_target_gene_GP.\n",
      "Combining GUCA2A_ligand_receptor_GP and Guca2a_ligand_receptor_target_gene_GP.\n",
      "Combining GUCA2B_ligand_receptor_GP and Guca2b_ligand_receptor_target_gene_GP.\n",
      "Combining HBEGF_ligand_receptor_GP and Hbegf_ligand_receptor_target_gene_GP.\n",
      "Combining HCRT_ligand_receptor_GP and Hcrt_ligand_receptor_target_gene_GP.\n",
      "Combining HGF_ligand_receptor_GP and Hgf_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DMA_ligand_receptor_GP and H2-DMa_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DQB1_ligand_receptor_GP and HLA-DQB2_ligand_receptor_GP.\n",
      "Combining HLA-DQB1_ligand_receptor_GP and H2-Ab1_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DQB2_ligand_receptor_GP and HLA-DQB1_ligand_receptor_GP.\n",
      "Combining HLA-DQB2_ligand_receptor_GP and H2-Ab1_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DRA_ligand_receptor_GP and H2-Ea_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DRB1_ligand_receptor_GP and HLA-DRB5_ligand_receptor_GP.\n",
      "Combining HLA-DRB1_ligand_receptor_GP and H2-Eb2_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DRB5_ligand_receptor_GP and HLA-DRB1_ligand_receptor_GP.\n",
      "Combining HLA-DRB5_ligand_receptor_GP and H2-Eb2_ligand_receptor_target_gene_GP.\n",
      "Combining HP_ligand_receptor_GP and Hp_ligand_receptor_target_gene_GP.\n",
      "Combining IAPP_ligand_receptor_GP and Iapp_ligand_receptor_target_gene_GP.\n",
      "Combining ICAM1_ligand_receptor_GP and Icam1_ligand_receptor_target_gene_GP.\n",
      "Combining ICAM2_ligand_receptor_GP and Icam2_ligand_receptor_target_gene_GP.\n",
      "Combining ICAM4_ligand_receptor_GP and Icam4_ligand_receptor_target_gene_GP.\n",
      "Combining ICAM5_ligand_receptor_GP and Icam5_ligand_receptor_target_gene_GP.\n",
      "Combining ICOSLG_ligand_receptor_GP and Icosl_ligand_receptor_target_gene_GP.\n",
      "Combining IFNB1_ligand_receptor_GP and Ifnb1_ligand_receptor_target_gene_GP.\n",
      "Combining IFNE_ligand_receptor_GP and Ifne_ligand_receptor_target_gene_GP.\n",
      "Combining IFNG_ligand_receptor_GP and Ifng_ligand_receptor_target_gene_GP.\n",
      "Combining IFNK_ligand_receptor_GP and Ifnk_ligand_receptor_target_gene_GP.\n",
      "Combining IGF1_ligand_receptor_GP and Igf1_ligand_receptor_target_gene_GP.\n",
      "Combining IGF2_ligand_receptor_GP and Igf2_ligand_receptor_target_gene_GP.\n",
      "Combining IGFL2_ligand_receptor_GP and IGFL3_ligand_receptor_GP.\n",
      "Combining IGFL2_ligand_receptor_GP and Igfl3_ligand_receptor_target_gene_GP.\n",
      "Combining IGFL3_ligand_receptor_GP and IGFL2_ligand_receptor_GP.\n",
      "Combining IGFL3_ligand_receptor_GP and Igfl3_ligand_receptor_target_gene_GP.\n",
      "Combining IHH_ligand_receptor_GP and Ihh_ligand_receptor_target_gene_GP.\n",
      "Combining IL10_ligand_receptor_GP and Il10_ligand_receptor_target_gene_GP.\n",
      "Combining IL11_ligand_receptor_GP and Il11_ligand_receptor_target_gene_GP.\n",
      "Combining IL12A_ligand_receptor_GP and Il12a_ligand_receptor_target_gene_GP.\n",
      "Combining IL12B_ligand_receptor_GP and Il12b_ligand_receptor_target_gene_GP.\n",
      "Combining IL13_ligand_receptor_GP and Il13_ligand_receptor_target_gene_GP.\n",
      "Combining IL15_ligand_receptor_GP and Il15_ligand_receptor_target_gene_GP.\n",
      "Combining IL16_ligand_receptor_GP and Il16_ligand_receptor_target_gene_GP.\n",
      "Combining IL17A_ligand_receptor_GP and Il17a_ligand_receptor_target_gene_GP.\n",
      "Combining IL17B_ligand_receptor_GP and Il17b_ligand_receptor_target_gene_GP.\n",
      "Combining IL17C_ligand_receptor_GP and Il17c_ligand_receptor_target_gene_GP.\n",
      "Combining IL17F_ligand_receptor_GP and Il17f_ligand_receptor_target_gene_GP.\n",
      "Combining IL18_ligand_receptor_GP and Il18_ligand_receptor_target_gene_GP.\n",
      "Combining IL19_ligand_receptor_GP and Il19_ligand_receptor_target_gene_GP.\n",
      "Combining IL1A_ligand_receptor_GP and Il1a_ligand_receptor_target_gene_GP.\n",
      "Combining IL1B_ligand_receptor_GP and Il1b_ligand_receptor_target_gene_GP.\n",
      "Combining IL1F10_ligand_receptor_GP and Il1f10_ligand_receptor_target_gene_GP.\n",
      "Combining IL1RN_ligand_receptor_GP and Il1rn_ligand_receptor_target_gene_GP.\n",
      "Combining IL2_ligand_receptor_GP and Il2_ligand_receptor_target_gene_GP.\n",
      "Combining IL20_ligand_receptor_GP and Il20_ligand_receptor_target_gene_GP.\n",
      "Combining IL21_ligand_receptor_GP and Il21_ligand_receptor_target_gene_GP.\n",
      "Combining IL23A_ligand_receptor_GP and Il23a_ligand_receptor_target_gene_GP.\n",
      "Combining IL24_ligand_receptor_GP and Il24_ligand_receptor_target_gene_GP.\n",
      "Combining IL25_ligand_receptor_GP and Il25_ligand_receptor_target_gene_GP.\n",
      "Combining IL27_ligand_receptor_GP and Il27_ligand_receptor_target_gene_GP.\n",
      "Combining IL3_ligand_receptor_GP and Il3_ligand_receptor_target_gene_GP.\n",
      "Combining IL31_ligand_receptor_GP and Il31_ligand_receptor_target_gene_GP.\n",
      "Combining IL33_ligand_receptor_GP and Il33_ligand_receptor_target_gene_GP.\n",
      "Combining IL34_ligand_receptor_GP and Il34_ligand_receptor_target_gene_GP.\n",
      "Combining IL36A_ligand_receptor_GP and Il36a_ligand_receptor_target_gene_GP.\n",
      "Combining IL36B_ligand_receptor_GP and Il36b_ligand_receptor_target_gene_GP.\n",
      "Combining IL36G_ligand_receptor_GP and Il36g_ligand_receptor_target_gene_GP.\n",
      "Combining IL36RN_ligand_receptor_GP and Il36rn_ligand_receptor_target_gene_GP.\n",
      "Combining IL4_ligand_receptor_GP and Il4_ligand_receptor_target_gene_GP.\n",
      "Combining IL5_ligand_receptor_GP and Il5_ligand_receptor_target_gene_GP.\n",
      "Combining IL6_ligand_receptor_GP and Il6_ligand_receptor_target_gene_GP.\n",
      "Combining IL7_ligand_receptor_GP and Il7_ligand_receptor_target_gene_GP.\n",
      "Combining IL9_ligand_receptor_GP and Il9_ligand_receptor_target_gene_GP.\n",
      "Combining INHA_ligand_receptor_GP and Inha_ligand_receptor_target_gene_GP.\n",
      "Combining INHBA_ligand_receptor_GP and Inhba_ligand_receptor_target_gene_GP.\n",
      "Combining INHBB_ligand_receptor_GP and Inhbb_ligand_receptor_target_gene_GP.\n",
      "Combining INHBC_ligand_receptor_GP and Inhbc_ligand_receptor_target_gene_GP.\n",
      "Combining INHBE_ligand_receptor_GP and Inhbe_ligand_receptor_target_gene_GP.\n",
      "Combining INSL3_ligand_receptor_GP and Insl3_ligand_receptor_target_gene_GP.\n",
      "Combining INSL5_ligand_receptor_GP and Insl5_ligand_receptor_target_gene_GP.\n",
      "Combining IZUMO1_ligand_receptor_GP and Izumo1_ligand_receptor_target_gene_GP.\n",
      "Combining JAG1_ligand_receptor_GP and Jag1_ligand_receptor_target_gene_GP.\n",
      "Combining JAG2_ligand_receptor_GP and Jag2_ligand_receptor_target_gene_GP.\n",
      "Combining JAM2_ligand_receptor_GP and Jam2_ligand_receptor_target_gene_GP.\n",
      "Combining JAM3_ligand_receptor_GP and Jam3_ligand_receptor_target_gene_GP.\n",
      "Combining JAML_ligand_receptor_GP and Jaml_ligand_receptor_target_gene_GP.\n",
      "Combining KITLG_ligand_receptor_GP and Kitl_ligand_receptor_target_gene_GP.\n",
      "Combining KL_ligand_receptor_GP and Kl_ligand_receptor_target_gene_GP.\n",
      "Combining L1CAM_ligand_receptor_GP and L1cam_ligand_receptor_target_gene_GP.\n",
      "Combining LAMA1_ligand_receptor_GP and Lama1_ligand_receptor_target_gene_GP.\n",
      "Combining LAMA2_ligand_receptor_GP and Lama2_ligand_receptor_target_gene_GP.\n",
      "Combining LAMA3_ligand_receptor_GP and Lama3_ligand_receptor_target_gene_GP.\n",
      "Combining LAMA5_ligand_receptor_GP and Lama5_ligand_receptor_target_gene_GP.\n",
      "Combining LAMB1_ligand_receptor_GP and Lamb1_ligand_receptor_target_gene_GP.\n",
      "Combining LAMB2_ligand_receptor_GP and Lamb2_ligand_receptor_target_gene_GP.\n",
      "Combining LAMB3_ligand_receptor_GP and Lamb3_ligand_receptor_target_gene_GP.\n",
      "Combining LAMC1_ligand_receptor_GP and Lamc1_ligand_receptor_target_gene_GP.\n",
      "Combining LAMC2_ligand_receptor_GP and Lamc2_ligand_receptor_target_gene_GP.\n",
      "Combining LEAP2_ligand_receptor_GP and Leap2_ligand_receptor_target_gene_GP.\n",
      "Combining LEP_ligand_receptor_GP and Lep_ligand_receptor_target_gene_GP.\n",
      "Combining LFNG_ligand_receptor_GP and Lfng_ligand_receptor_target_gene_GP.\n",
      "Combining LGALS1_ligand_receptor_GP and Lgals1_ligand_receptor_target_gene_GP.\n",
      "Combining LGALS3_ligand_receptor_GP and Lgals3_ligand_receptor_target_gene_GP.\n",
      "Combining LGALS3BP_ligand_receptor_GP and Lgals3bp_ligand_receptor_target_gene_GP.\n",
      "Combining LGI3_ligand_receptor_GP and Lgi3_ligand_receptor_target_gene_GP.\n",
      "Combining LGI4_ligand_receptor_GP and Lgi4_ligand_receptor_target_gene_GP.\n",
      "Combining LHB_ligand_receptor_GP and Lhb_ligand_receptor_target_gene_GP.\n",
      "Combining LIF_ligand_receptor_GP and Lif_ligand_receptor_target_gene_GP.\n",
      "Combining LRIG1_ligand_receptor_GP and Lrig1_ligand_receptor_target_gene_GP.\n",
      "Combining LRRC4B_ligand_receptor_GP and Lrrc4b_ligand_receptor_target_gene_GP.\n",
      "Combining LTA_ligand_receptor_GP and Lta_ligand_receptor_target_gene_GP.\n",
      "Combining LTB_ligand_receptor_GP and Ltb_ligand_receptor_target_gene_GP.\n",
      "Combining LTBP3_ligand_receptor_GP and Ltbp3_ligand_receptor_target_gene_GP.\n",
      "Combining MDK_ligand_receptor_GP and Mdk_ligand_receptor_target_gene_GP.\n",
      "Combining MFAP2_ligand_receptor_GP and Mfap2_ligand_receptor_target_gene_GP.\n",
      "Combining MFGE8_ligand_receptor_GP and Mfge8_ligand_receptor_target_gene_GP.\n",
      "Combining MIA_ligand_receptor_GP and Mia_ligand_receptor_target_gene_GP.\n",
      "Combining MIF_ligand_receptor_GP and Mif_ligand_receptor_target_gene_GP.\n",
      "Combining MST1_ligand_receptor_GP and Mst1_ligand_receptor_target_gene_GP.\n",
      "Combining MSTN_ligand_receptor_GP and Mstn_ligand_receptor_target_gene_GP.\n",
      "Combining MUC1_ligand_receptor_GP and Muc1_ligand_receptor_target_gene_GP.\n",
      "Combining MYOC_ligand_receptor_GP and Myoc_ligand_receptor_target_gene_GP.\n",
      "Combining NDP_ligand_receptor_GP and Ndp_ligand_receptor_target_gene_GP.\n",
      "Combining NECTIN2_ligand_receptor_GP and Nectin2_ligand_receptor_target_gene_GP.\n",
      "Combining NECTIN3_ligand_receptor_GP and Nectin3_ligand_receptor_target_gene_GP.\n",
      "Combining NECTIN4_ligand_receptor_GP and Nectin4_ligand_receptor_target_gene_GP.\n",
      "Combining NELL2_ligand_receptor_GP and Nell2_ligand_receptor_target_gene_GP.\n",
      "Combining NGF_ligand_receptor_GP and Ngf_ligand_receptor_target_gene_GP.\n",
      "Combining NID1_ligand_receptor_GP and Nid1_ligand_receptor_target_gene_GP.\n",
      "Combining NLGN1_ligand_receptor_GP and Nlgn1_ligand_receptor_target_gene_GP.\n",
      "Combining NLGN2_ligand_receptor_GP and Nlgn2_ligand_receptor_target_gene_GP.\n",
      "Combining NLGN3_ligand_receptor_GP and Nlgn3_ligand_receptor_target_gene_GP.\n",
      "Combining NMB_ligand_receptor_GP and Nmb_ligand_receptor_target_gene_GP.\n",
      "Combining NMS_ligand_receptor_GP and Nms_ligand_receptor_target_gene_GP.\n",
      "Combining NMU_ligand_receptor_GP and Nmu_ligand_receptor_target_gene_GP.\n",
      "Combining NODAL_ligand_receptor_GP and Nodal_ligand_receptor_target_gene_GP.\n",
      "Combining NOG_ligand_receptor_GP and Nog_ligand_receptor_target_gene_GP.\n",
      "Combining NPB_ligand_receptor_GP and Npb_ligand_receptor_target_gene_GP.\n",
      "Combining NPFF_ligand_receptor_GP and Npff_ligand_receptor_target_gene_GP.\n",
      "Combining NPNT_ligand_receptor_GP and Npnt_ligand_receptor_target_gene_GP.\n",
      "Combining NPPA_ligand_receptor_GP and Nppa_ligand_receptor_target_gene_GP.\n",
      "Combining NPPB_ligand_receptor_GP and Nppb_ligand_receptor_target_gene_GP.\n",
      "Combining NPPC_ligand_receptor_GP and Nppc_ligand_receptor_target_gene_GP.\n",
      "Combining NPS_ligand_receptor_GP and Nps_ligand_receptor_target_gene_GP.\n",
      "Combining NPTX2_ligand_receptor_GP and Nptx2_ligand_receptor_target_gene_GP.\n",
      "Combining NPVF_ligand_receptor_GP and Npvf_ligand_receptor_target_gene_GP.\n",
      "Combining NPW_ligand_receptor_GP and Npw_ligand_receptor_target_gene_GP.\n",
      "Combining NPY_ligand_receptor_GP and Npy_ligand_receptor_target_gene_GP.\n",
      "Combining NRG1_ligand_receptor_GP and Nrg1_ligand_receptor_target_gene_GP.\n",
      "Combining NRG2_ligand_receptor_GP and Nrg2_ligand_receptor_target_gene_GP.\n",
      "Combining NRG3_ligand_receptor_GP and Nrg3_ligand_receptor_target_gene_GP.\n",
      "Combining NRG4_ligand_receptor_GP and Nrg4_ligand_receptor_target_gene_GP.\n",
      "Combining NRP1_ligand_receptor_GP and Nrp1_ligand_receptor_target_gene_GP.\n",
      "Combining NRTN_ligand_receptor_GP and Nrtn_ligand_receptor_target_gene_GP.\n",
      "Combining NTF3_ligand_receptor_GP and Ntf3_ligand_receptor_target_gene_GP.\n",
      "Combining NTF4_ligand_receptor_GP and Ntf5_ligand_receptor_target_gene_GP.\n",
      "Combining NTN1_ligand_receptor_GP and Ntn1_ligand_receptor_target_gene_GP.\n",
      "Combining NTN3_ligand_receptor_GP and Ntn3_ligand_receptor_target_gene_GP.\n",
      "Combining NTN4_ligand_receptor_GP and Ntn4_ligand_receptor_target_gene_GP.\n",
      "Combining NTNG2_ligand_receptor_GP and Ntng2_ligand_receptor_target_gene_GP.\n",
      "Combining NTS_ligand_receptor_GP and Nts_ligand_receptor_target_gene_GP.\n",
      "Combining NXPH1_ligand_receptor_GP and Nxph1_ligand_receptor_target_gene_GP.\n",
      "Combining NXPH2_ligand_receptor_GP and Nxph2_ligand_receptor_target_gene_GP.\n",
      "Combining NXPH3_ligand_receptor_GP and Nxph3_ligand_receptor_target_gene_GP.\n",
      "Combining OCLN_ligand_receptor_GP and Ocln_ligand_receptor_target_gene_GP.\n",
      "Combining OLFM2_ligand_receptor_GP and Olfm2_ligand_receptor_target_gene_GP.\n",
      "Combining OSM_ligand_receptor_GP and Osm_ligand_receptor_target_gene_GP.\n",
      "Combining OSTN_ligand_receptor_GP and Ostn_ligand_receptor_target_gene_GP.\n",
      "Combining OXT_ligand_receptor_GP and Oxt_ligand_receptor_target_gene_GP.\n",
      "Combining PAM_ligand_receptor_GP and Pam_ligand_receptor_target_gene_GP.\n",
      "Combining PCSK9_ligand_receptor_GP and Pcsk9_ligand_receptor_target_gene_GP.\n",
      "Combining PDCD1LG2_ligand_receptor_GP and Pdcd1lg2_ligand_receptor_target_gene_GP.\n",
      "Combining PDGFA_ligand_receptor_GP and Pdgfa_ligand_receptor_target_gene_GP.\n",
      "Combining PDGFB_ligand_receptor_GP and Pdgfb_ligand_receptor_target_gene_GP.\n",
      "Combining PDGFC_ligand_receptor_GP and Pdgfc_ligand_receptor_target_gene_GP.\n",
      "Combining PDGFD_ligand_receptor_GP and Pdgfd_ligand_receptor_target_gene_GP.\n",
      "Combining PDYN_ligand_receptor_GP and Pdyn_ligand_receptor_target_gene_GP.\n",
      "Combining PECAM1_ligand_receptor_GP and Pecam1_ligand_receptor_target_gene_GP.\n",
      "Combining PENK_ligand_receptor_GP and Penk_ligand_receptor_target_gene_GP.\n",
      "Combining PF4_ligand_receptor_GP and Pf4_ligand_receptor_target_gene_GP.\n",
      "Combining PGF_ligand_receptor_GP and Pgf_ligand_receptor_target_gene_GP.\n",
      "Combining PGLYRP1_ligand_receptor_GP and Pglyrp1_ligand_receptor_target_gene_GP.\n",
      "Combining PIP_ligand_receptor_GP and Pip_ligand_receptor_target_gene_GP.\n",
      "Combining PLG_ligand_receptor_GP and Plg_ligand_receptor_target_gene_GP.\n",
      "Combining PMCH_ligand_receptor_GP and Pmch_ligand_receptor_target_gene_GP.\n",
      "Combining PNOC_ligand_receptor_GP and Pnoc_ligand_receptor_target_gene_GP.\n",
      "Combining PODXL2_ligand_receptor_GP and Podxl2_ligand_receptor_target_gene_GP.\n",
      "Combining POMC_ligand_receptor_GP and Pomc_ligand_receptor_target_gene_GP.\n",
      "Combining POSTN_ligand_receptor_GP and Postn_ligand_receptor_target_gene_GP.\n",
      "Combining PPBP_ligand_receptor_GP and Ppbp_ligand_receptor_target_gene_GP.\n",
      "Combining PPY_ligand_receptor_GP and Ppy_ligand_receptor_target_gene_GP.\n",
      "Combining PRLH_ligand_receptor_GP and Prlh_ligand_receptor_target_gene_GP.\n",
      "Combining PRND_ligand_receptor_GP and Prnd_ligand_receptor_target_gene_GP.\n",
      "Combining PROK1_ligand_receptor_GP and Prok1_ligand_receptor_target_gene_GP.\n",
      "Combining PROK2_ligand_receptor_GP and Prok2_ligand_receptor_target_gene_GP.\n",
      "Combining PSPN_ligand_receptor_GP and Pspn_ligand_receptor_target_gene_GP.\n",
      "Combining PTH_ligand_receptor_GP and Pth_ligand_receptor_target_gene_GP.\n",
      "Combining PTH2_ligand_receptor_GP and Pth2_ligand_receptor_target_gene_GP.\n",
      "Combining PTHLH_ligand_receptor_GP and Pthlh_ligand_receptor_target_gene_GP.\n",
      "Combining PTN_ligand_receptor_GP and Ptn_ligand_receptor_target_gene_GP.\n",
      "Combining PYY_ligand_receptor_GP and Pyy_ligand_receptor_target_gene_GP.\n",
      "Combining QRFP_ligand_receptor_GP and Qrfp_ligand_receptor_target_gene_GP.\n",
      "Combining RAET1E_ligand_receptor_GP and Raet1e_ligand_receptor_target_gene_GP.\n",
      "Combining RARRES2_ligand_receptor_GP and Rarres2_ligand_receptor_target_gene_GP.\n",
      "Combining RBP3_ligand_receptor_GP and Rbp3_ligand_receptor_target_gene_GP.\n",
      "Combining RELN_ligand_receptor_GP and Reln_ligand_receptor_target_gene_GP.\n",
      "Combining RETN_ligand_receptor_GP and Retn_ligand_receptor_target_gene_GP.\n",
      "Combining RGMA_ligand_receptor_GP and Rgma_ligand_receptor_target_gene_GP.\n",
      "Combining RLN1_ligand_receptor_GP and RLN2_ligand_receptor_GP.\n",
      "Combining RLN1_ligand_receptor_GP and Rln1_ligand_receptor_target_gene_GP.\n",
      "Combining RLN2_ligand_receptor_GP and RLN1_ligand_receptor_GP.\n",
      "Combining RLN2_ligand_receptor_GP and Rln1_ligand_receptor_target_gene_GP.\n",
      "Combining RLN3_ligand_receptor_GP and Rln3_ligand_receptor_target_gene_GP.\n",
      "Combining RSPO1_ligand_receptor_GP and Rspo1_ligand_receptor_target_gene_GP.\n",
      "Combining RSPO2_ligand_receptor_GP and Rspo2_ligand_receptor_target_gene_GP.\n",
      "Combining RSPO3_ligand_receptor_GP and Rspo3_ligand_receptor_target_gene_GP.\n",
      "Combining RSPO4_ligand_receptor_GP and Rspo4_ligand_receptor_target_gene_GP.\n",
      "Combining S100A10_ligand_receptor_GP and S100a10_ligand_receptor_target_gene_GP.\n",
      "Combining S100A4_ligand_receptor_GP and S100a4_ligand_receptor_target_gene_GP.\n",
      "Combining S100A8_ligand_receptor_GP and S100a8_ligand_receptor_target_gene_GP.\n",
      "Combining S100B_ligand_receptor_GP and S100b_ligand_receptor_target_gene_GP.\n",
      "Combining SCGB1A1_ligand_receptor_GP and Scgb1a1_ligand_receptor_target_gene_GP.\n",
      "Combining SCGB3A1_ligand_receptor_GP and Scgb3a1_ligand_receptor_target_gene_GP.\n",
      "Combining SCGB3A2_ligand_receptor_GP and Scgb3a2_ligand_receptor_target_gene_GP.\n",
      "Combining SCT_ligand_receptor_GP and Sct_ligand_receptor_target_gene_GP.\n",
      "Combining SDC2_ligand_receptor_GP and Sdc2_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3A_ligand_receptor_GP and Sema3a_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3B_ligand_receptor_GP and Sema3b_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3C_ligand_receptor_GP and Sema3c_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3D_ligand_receptor_GP and Sema3d_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3E_ligand_receptor_GP and Sema3e_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3F_ligand_receptor_GP and Sema3f_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3G_ligand_receptor_GP and Sema3g_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA4B_ligand_receptor_GP and Sema4b_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA5A_ligand_receptor_GP and Sema5a_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA5B_ligand_receptor_GP and Sema5b_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA6B_ligand_receptor_GP and Sema6b_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA6D_ligand_receptor_GP and Sema6d_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA7A_ligand_receptor_GP and Sema7a_ligand_receptor_target_gene_GP.\n",
      "Combining SFRP2_ligand_receptor_GP and Sfrp2_ligand_receptor_target_gene_GP.\n",
      "Combining SFTPA1_ligand_receptor_GP and Sftpa1_ligand_receptor_target_gene_GP.\n",
      "Combining SFTPD_ligand_receptor_GP and Sftpd_ligand_receptor_target_gene_GP.\n",
      "Combining SHH_ligand_receptor_GP and Shh_ligand_receptor_target_gene_GP.\n",
      "Combining SIGLEC9_ligand_receptor_GP and Siglece_ligand_receptor_target_gene_GP.\n",
      "Combining SLIT1_ligand_receptor_GP and Slit1_ligand_receptor_target_gene_GP.\n",
      "Combining SLIT2_ligand_receptor_GP and Slit2_ligand_receptor_target_gene_GP.\n",
      "Combining SLIT3_ligand_receptor_GP and Slit3_ligand_receptor_target_gene_GP.\n",
      "Combining SLITRK1_ligand_receptor_GP and Slitrk1_ligand_receptor_target_gene_GP.\n",
      "Combining SLURP1_ligand_receptor_GP and Slurp1_ligand_receptor_target_gene_GP.\n",
      "Combining SORL1_ligand_receptor_GP and Sorl1_ligand_receptor_target_gene_GP.\n",
      "Combining SOST_ligand_receptor_GP and Sost_ligand_receptor_target_gene_GP.\n",
      "Combining SPON1_ligand_receptor_GP and Spon1_ligand_receptor_target_gene_GP.\n",
      "Combining SPON2_ligand_receptor_GP and Spon2_ligand_receptor_target_gene_GP.\n",
      "Combining SPP1_ligand_receptor_GP and Spp1_ligand_receptor_target_gene_GP.\n",
      "Combining SPX_ligand_receptor_GP and Spx_ligand_receptor_target_gene_GP.\n",
      "Combining SST_ligand_receptor_GP and Sst_ligand_receptor_target_gene_GP.\n",
      "Combining TAC1_ligand_receptor_GP and Tac1_ligand_receptor_target_gene_GP.\n",
      "Combining TAC3_ligand_receptor_GP and Tac2_ligand_receptor_target_gene_GP.\n",
      "Combining TAC4_ligand_receptor_GP and Tac4_ligand_receptor_target_gene_GP.\n",
      "Combining TF_ligand_receptor_GP and Trf_ligand_receptor_target_gene_GP.\n",
      "Combining TFF1_ligand_receptor_GP and Tff1_ligand_receptor_target_gene_GP.\n",
      "Combining TFF2_ligand_receptor_GP and Tff2_ligand_receptor_target_gene_GP.\n",
      "Combining TG_ligand_receptor_GP and Tg_ligand_receptor_target_gene_GP.\n",
      "Combining TGFA_ligand_receptor_GP and Tgfa_ligand_receptor_target_gene_GP.\n",
      "Combining TGFB1_ligand_receptor_GP and Tgfb1_ligand_receptor_target_gene_GP.\n",
      "Combining TGFB2_ligand_receptor_GP and Tgfb2_ligand_receptor_target_gene_GP.\n",
      "Combining TGFB3_ligand_receptor_GP and Tgfb3_ligand_receptor_target_gene_GP.\n",
      "Combining THBS1_ligand_receptor_GP and Thbs1_ligand_receptor_target_gene_GP.\n",
      "Combining THBS2_ligand_receptor_GP and Thbs2_ligand_receptor_target_gene_GP.\n",
      "Combining THPO_ligand_receptor_GP and Thpo_ligand_receptor_target_gene_GP.\n",
      "Combining TNC_ligand_receptor_GP and Tnc_ligand_receptor_target_gene_GP.\n",
      "Combining TNF_ligand_receptor_GP and Tnf_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF10_ligand_receptor_GP and Tnfsf10_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF11_ligand_receptor_GP and Tnfsf11_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF12_ligand_receptor_GP and Tnfsf12_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF13_ligand_receptor_GP and Tnfsf13_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF13B_ligand_receptor_GP and Tnfsf13b_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF14_ligand_receptor_GP and Tnfsf14_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF15_ligand_receptor_GP and Tnfsf15_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF18_ligand_receptor_GP and Tnfsf18_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF4_ligand_receptor_GP and Tnfsf4_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF8_ligand_receptor_GP and Tnfsf8_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF9_ligand_receptor_GP and Tnfsf9_ligand_receptor_target_gene_GP.\n",
      "Combining TRH_ligand_receptor_GP and Trh_ligand_receptor_target_gene_GP.\n",
      "Combining TSHB_ligand_receptor_GP and Tshb_ligand_receptor_target_gene_GP.\n",
      "Combining TSLP_ligand_receptor_GP and Tslp_ligand_receptor_target_gene_GP.\n",
      "Combining UCN_ligand_receptor_GP and Ucn_ligand_receptor_target_gene_GP.\n",
      "Combining UCN2_ligand_receptor_GP and Ucn2_ligand_receptor_target_gene_GP.\n",
      "Combining UCN3_ligand_receptor_GP and Ucn3_ligand_receptor_target_gene_GP.\n",
      "Combining UTS2_ligand_receptor_GP and Uts2_ligand_receptor_target_gene_GP.\n",
      "Combining UTS2B_ligand_receptor_GP and Uts2b_ligand_receptor_target_gene_GP.\n",
      "Combining VCAM1_ligand_receptor_GP and Vcam1_ligand_receptor_target_gene_GP.\n",
      "Combining VCAN_ligand_receptor_GP and Vcan_ligand_receptor_target_gene_GP.\n",
      "Combining VEGFA_ligand_receptor_GP and Vegfa_ligand_receptor_target_gene_GP.\n",
      "Combining VEGFB_ligand_receptor_GP and Vegfb_ligand_receptor_target_gene_GP.\n",
      "Combining VEGFC_ligand_receptor_GP and Vegfc_ligand_receptor_target_gene_GP.\n",
      "Combining VEGFD_ligand_receptor_GP and Vegfd_ligand_receptor_target_gene_GP.\n",
      "Combining VIP_ligand_receptor_GP and Vip_ligand_receptor_target_gene_GP.\n",
      "Combining VTN_ligand_receptor_GP and Vtn_ligand_receptor_target_gene_GP.\n",
      "Combining VWF_ligand_receptor_GP and Vwf_ligand_receptor_target_gene_GP.\n",
      "Combining WNT1_ligand_receptor_GP and Wnt1_ligand_receptor_target_gene_GP.\n",
      "Combining WNT10B_ligand_receptor_GP and Wnt10b_ligand_receptor_target_gene_GP.\n",
      "Combining WNT11_ligand_receptor_GP and Wnt11_ligand_receptor_target_gene_GP.\n",
      "Combining WNT2_ligand_receptor_GP and Wnt2_ligand_receptor_target_gene_GP.\n",
      "Combining WNT2B_ligand_receptor_GP and Wnt2b_ligand_receptor_target_gene_GP.\n",
      "Combining WNT3_ligand_receptor_GP and Wnt3_ligand_receptor_target_gene_GP.\n",
      "Combining WNT3A_ligand_receptor_GP and Wnt3a_ligand_receptor_target_gene_GP.\n",
      "Combining WNT4_ligand_receptor_GP and Wnt4_ligand_receptor_target_gene_GP.\n",
      "Combining WNT5A_ligand_receptor_GP and Wnt5a_ligand_receptor_target_gene_GP.\n",
      "Combining WNT5B_ligand_receptor_GP and Wnt5b_ligand_receptor_target_gene_GP.\n",
      "Combining WNT6_ligand_receptor_GP and Wnt6_ligand_receptor_target_gene_GP.\n",
      "Combining WNT7A_ligand_receptor_GP and Wnt7a_ligand_receptor_target_gene_GP.\n",
      "Combining WNT7B_ligand_receptor_GP and Wnt7b_ligand_receptor_target_gene_GP.\n",
      "Combining WNT8A_ligand_receptor_GP and Wnt8a_ligand_receptor_target_gene_GP.\n",
      "Combining WNT9A_ligand_receptor_GP and Wnt9a_ligand_receptor_target_gene_GP.\n",
      "Combining WNT9B_ligand_receptor_GP and Wnt9b_ligand_receptor_target_gene_GP.\n",
      "Combining XCL1_ligand_receptor_GP and XCL2_ligand_receptor_GP.\n",
      "Combining XCL1_ligand_receptor_GP and Xcl1_ligand_receptor_target_gene_GP.\n",
      "Combining XCL2_ligand_receptor_GP and XCL1_ligand_receptor_GP.\n",
      "Combining XCL2_ligand_receptor_GP and Xcl1_ligand_receptor_target_gene_GP.\n",
      "Combining ZP3_ligand_receptor_GP and Zp3_ligand_receptor_target_gene_GP.\n",
      "Combining 2610528A11Rik_ligand_receptor_target_gene_GP and GPR15LG_ligand_receptor_GP.\n",
      "Combining Ace2_ligand_receptor_target_gene_GP and ACE2_ligand_receptor_GP.\n",
      "Combining Ace2_ligand_receptor_target_gene_GP and L-Phenylalanine_metabolite_enzyme_sensor_GP.\n",
      "Combining Adcyap1_ligand_receptor_target_gene_GP and ADCYAP1_ligand_receptor_GP.\n",
      "Combining Adgre5_ligand_receptor_target_gene_GP and ADGRE5_ligand_receptor_GP.\n",
      "Combining Adipoq_ligand_receptor_target_gene_GP and ADIPOQ_ligand_receptor_GP.\n",
      "Combining Adm_ligand_receptor_target_gene_GP and ADM_ligand_receptor_GP.\n",
      "Combining Adm2_ligand_receptor_target_gene_GP and ADM2_ligand_receptor_GP.\n",
      "Combining Agrp_ligand_receptor_target_gene_GP and AGRP_ligand_receptor_GP.\n",
      "Combining Agt_ligand_receptor_target_gene_GP and AGT_ligand_receptor_GP.\n",
      "Combining Alcam_ligand_receptor_target_gene_GP and ALCAM_ligand_receptor_GP.\n",
      "Combining Alkal1_ligand_receptor_target_gene_GP and ALKAL1_ligand_receptor_GP.\n",
      "Combining Alkal2_ligand_receptor_target_gene_GP and ALKAL2_ligand_receptor_GP.\n",
      "Combining Ambn_ligand_receptor_target_gene_GP and AMBN_ligand_receptor_GP.\n",
      "Combining Amelx_ligand_receptor_target_gene_GP and AMELX_ligand_receptor_GP.\n",
      "Combining Amh_ligand_receptor_target_gene_GP and AMH_ligand_receptor_GP.\n",
      "Combining Angpt1_ligand_receptor_target_gene_GP and ANGPT1_ligand_receptor_GP.\n",
      "Combining Angpt2_ligand_receptor_target_gene_GP and ANGPT2_ligand_receptor_GP.\n",
      "Combining Angpt4_ligand_receptor_target_gene_GP and ANGPT4_ligand_receptor_GP.\n",
      "Combining Angptl1_ligand_receptor_target_gene_GP and ANGPTL1_ligand_receptor_GP.\n",
      "Combining Angptl2_ligand_receptor_target_gene_GP and ANGPTL2_ligand_receptor_GP.\n",
      "Combining Angptl3_ligand_receptor_target_gene_GP and ANGPTL3_ligand_receptor_GP.\n",
      "Combining Angptl4_ligand_receptor_target_gene_GP and ANGPTL4_ligand_receptor_GP.\n",
      "Combining Anxa1_ligand_receptor_target_gene_GP and ANXA1_ligand_receptor_GP.\n",
      "Combining Anxa2_ligand_receptor_target_gene_GP and ANXA2_ligand_receptor_GP.\n",
      "Combining Apln_ligand_receptor_target_gene_GP and APLN_ligand_receptor_GP.\n",
      "Combining Apob_ligand_receptor_target_gene_GP and APOB_ligand_receptor_GP.\n",
      "Combining Apod_ligand_receptor_target_gene_GP and APOD_ligand_receptor_GP.\n",
      "Combining Apoe_ligand_receptor_target_gene_GP and APOE_ligand_receptor_GP.\n",
      "Combining Areg_ligand_receptor_target_gene_GP and AREG_ligand_receptor_GP.\n",
      "Combining Artn_ligand_receptor_target_gene_GP and ARTN_ligand_receptor_GP.\n",
      "Combining Avp_ligand_receptor_target_gene_GP and AVP_ligand_receptor_GP.\n",
      "Combining Azgp1_ligand_receptor_target_gene_GP and AZGP1_ligand_receptor_GP.\n",
      "Combining Bdnf_ligand_receptor_target_gene_GP and BDNF_ligand_receptor_GP.\n",
      "Combining Bmp10_ligand_receptor_target_gene_GP and BMP10_ligand_receptor_GP.\n",
      "Combining Bmp15_ligand_receptor_target_gene_GP and BMP15_ligand_receptor_GP.\n",
      "Combining Bmp2_ligand_receptor_target_gene_GP and BMP2_ligand_receptor_GP.\n",
      "Combining Bmp3_ligand_receptor_target_gene_GP and BMP3_ligand_receptor_GP.\n",
      "Combining Bmp4_ligand_receptor_target_gene_GP and BMP4_ligand_receptor_GP.\n",
      "Combining Bmp5_ligand_receptor_target_gene_GP and BMP5_ligand_receptor_GP.\n",
      "Combining Bmp6_ligand_receptor_target_gene_GP and BMP6_ligand_receptor_GP.\n",
      "Combining Bmp7_ligand_receptor_target_gene_GP and BMP7_ligand_receptor_GP.\n",
      "Combining Bmp8a_ligand_receptor_target_gene_GP and BMP8A_ligand_receptor_GP.\n",
      "Combining Bst2_ligand_receptor_target_gene_GP and BST2_ligand_receptor_GP.\n",
      "Combining Btc_ligand_receptor_target_gene_GP and BTC_ligand_receptor_GP.\n",
      "Combining C1qtnf1_ligand_receptor_target_gene_GP and C1QTNF1_ligand_receptor_GP.\n",
      "Combining C1qtnf5_ligand_receptor_target_gene_GP and C1QTNF5_ligand_receptor_GP.\n",
      "Combining C4b_ligand_receptor_target_gene_GP and C4B_ligand_receptor_GP.\n",
      "Combining Calca_ligand_receptor_target_gene_GP and CALCA_ligand_receptor_GP.\n",
      "Combining Calcb_ligand_receptor_target_gene_GP and CALCB_ligand_receptor_GP.\n",
      "Combining Calr_ligand_receptor_target_gene_GP and CALR_ligand_receptor_GP.\n",
      "Combining Cck_ligand_receptor_target_gene_GP and CCK_ligand_receptor_GP.\n",
      "Combining Ccl1_ligand_receptor_target_gene_GP and CCL1_ligand_receptor_GP.\n",
      "Combining Ccl11_ligand_receptor_target_gene_GP and CCL11_ligand_receptor_GP.\n",
      "Combining Ccl17_ligand_receptor_target_gene_GP and CCL17_ligand_receptor_GP.\n",
      "Combining Ccl19_ligand_receptor_target_gene_GP and CCL19_ligand_receptor_GP.\n",
      "Combining Ccl2_ligand_receptor_target_gene_GP and CCL2_ligand_receptor_GP.\n",
      "Combining Ccl20_ligand_receptor_target_gene_GP and CCL20_ligand_receptor_GP.\n",
      "Combining Ccl22_ligand_receptor_target_gene_GP and CCL22_ligand_receptor_GP.\n",
      "Combining Ccl24_ligand_receptor_target_gene_GP and CCL24_ligand_receptor_GP.\n",
      "Combining Ccl25_ligand_receptor_target_gene_GP and CCL25_ligand_receptor_GP.\n",
      "Combining Ccl26_ligand_receptor_target_gene_GP and CCL26_ligand_receptor_GP.\n",
      "Combining Ccl28_ligand_receptor_target_gene_GP and CCL28_ligand_receptor_GP.\n",
      "Combining Ccl3_ligand_receptor_target_gene_GP and CCL18_ligand_receptor_GP.\n",
      "Combining Ccl3_ligand_receptor_target_gene_GP and CCL3_ligand_receptor_GP.\n",
      "Combining Ccl4_ligand_receptor_target_gene_GP and CCL4_ligand_receptor_GP.\n",
      "Combining Ccl4_ligand_receptor_target_gene_GP and CCL4L2_ligand_receptor_GP.\n",
      "Combining Ccl5_ligand_receptor_target_gene_GP and CCL5_ligand_receptor_GP.\n",
      "Combining Ccl7_ligand_receptor_target_gene_GP and CCL7_ligand_receptor_GP.\n",
      "Combining Ccl8_ligand_receptor_target_gene_GP and CCL8_ligand_receptor_GP.\n",
      "Combining Ccn1_ligand_receptor_target_gene_GP and CCN1_ligand_receptor_GP.\n",
      "Combining Ccn2_ligand_receptor_target_gene_GP and CCN2_ligand_receptor_GP.\n",
      "Combining Ccn3_ligand_receptor_target_gene_GP and CCN3_ligand_receptor_GP.\n",
      "Combining Cd160_ligand_receptor_target_gene_GP and CD160_ligand_receptor_GP.\n",
      "Combining Cd200_ligand_receptor_target_gene_GP and CD200_ligand_receptor_GP.\n",
      "Combining Cd274_ligand_receptor_target_gene_GP and CD274_ligand_receptor_GP.\n",
      "Combining Cd34_ligand_receptor_target_gene_GP and CD34_ligand_receptor_GP.\n",
      "Combining Cd40lg_ligand_receptor_target_gene_GP and CD40LG_ligand_receptor_GP.\n",
      "Combining Cd48_ligand_receptor_target_gene_GP and CD48_ligand_receptor_GP.\n",
      "Combining Cd52_ligand_receptor_target_gene_GP and CD52_ligand_receptor_GP.\n",
      "Combining Cd5l_ligand_receptor_target_gene_GP and CD5L_ligand_receptor_GP.\n",
      "Combining Cd70_ligand_receptor_target_gene_GP and CD70_ligand_receptor_GP.\n",
      "Combining Cd72_ligand_receptor_target_gene_GP and CD72_ligand_receptor_GP.\n",
      "Combining Cd80_ligand_receptor_target_gene_GP and CD80_ligand_receptor_GP.\n",
      "Combining Cd99_ligand_receptor_target_gene_GP and CD99_ligand_receptor_GP.\n",
      "Combining Ceacam5_ligand_receptor_target_gene_GP and CEACAM5_ligand_receptor_GP.\n",
      "Combining Cga_ligand_receptor_target_gene_GP and CGA_ligand_receptor_GP.\n",
      "Combining Chad_ligand_receptor_target_gene_GP and CHAD_ligand_receptor_GP.\n",
      "Combining Cklf_ligand_receptor_target_gene_GP and CKLF_ligand_receptor_GP.\n",
      "Combining Clcf1_ligand_receptor_target_gene_GP and CLCF1_ligand_receptor_GP.\n",
      "Combining Cldn1_ligand_receptor_target_gene_GP and CLDN1_ligand_receptor_GP.\n",
      "Combining Cldn16_ligand_receptor_target_gene_GP and CLDN16_ligand_receptor_GP.\n",
      "Combining Clec4g_ligand_receptor_target_gene_GP and CLEC4G_ligand_receptor_GP.\n",
      "Combining Clu_ligand_receptor_target_gene_GP and CLU_ligand_receptor_GP.\n",
      "Combining Cmtm8_ligand_receptor_target_gene_GP and CMTM8_ligand_receptor_GP.\n",
      "Combining Cntn1_ligand_receptor_target_gene_GP and CNTN1_ligand_receptor_GP.\n",
      "Combining Cntn2_ligand_receptor_target_gene_GP and CNTN2_ligand_receptor_GP.\n",
      "Combining Cntn3_ligand_receptor_target_gene_GP and CNTN3_ligand_receptor_GP.\n",
      "Combining Cntn4_ligand_receptor_target_gene_GP and CNTN4_ligand_receptor_GP.\n",
      "Combining Cntn6_ligand_receptor_target_gene_GP and CNTN6_ligand_receptor_GP.\n",
      "Combining Col10a1_ligand_receptor_target_gene_GP and COL10A1_ligand_receptor_GP.\n",
      "Combining Col11a2_ligand_receptor_target_gene_GP and COL11A2_ligand_receptor_GP.\n",
      "Combining Col12a1_ligand_receptor_target_gene_GP and COL12A1_ligand_receptor_GP.\n",
      "Combining Col13a1_ligand_receptor_target_gene_GP and COL13A1_ligand_receptor_GP.\n",
      "Combining Col15a1_ligand_receptor_target_gene_GP and COL15A1_ligand_receptor_GP.\n",
      "Combining Col16a1_ligand_receptor_target_gene_GP and COL16A1_ligand_receptor_GP.\n",
      "Combining Col17a1_ligand_receptor_target_gene_GP and COL17A1_ligand_receptor_GP.\n",
      "Combining Col19a1_ligand_receptor_target_gene_GP and COL19A1_ligand_receptor_GP.\n",
      "Combining Col20a1_ligand_receptor_target_gene_GP and COL20A1_ligand_receptor_GP.\n",
      "Combining Col22a1_ligand_receptor_target_gene_GP and COL22A1_ligand_receptor_GP.\n",
      "Combining Col24a1_ligand_receptor_target_gene_GP and COL24A1_ligand_receptor_GP.\n",
      "Combining Col26a1_ligand_receptor_target_gene_GP and COL26A1_ligand_receptor_GP.\n",
      "Combining Col27a1_ligand_receptor_target_gene_GP and COL27A1_ligand_receptor_GP.\n",
      "Combining Col28a1_ligand_receptor_target_gene_GP and COL28A1_ligand_receptor_GP.\n",
      "Combining Col6a3_ligand_receptor_target_gene_GP and COL6A3_ligand_receptor_GP.\n",
      "Combining Col6a5_ligand_receptor_target_gene_GP and COL6A5_ligand_receptor_GP.\n",
      "Combining Col6a6_ligand_receptor_target_gene_GP and COL6A6_ligand_receptor_GP.\n",
      "Combining Col8a2_ligand_receptor_target_gene_GP and COL8A2_ligand_receptor_GP.\n",
      "Combining Comp_ligand_receptor_target_gene_GP and COMP_ligand_receptor_GP.\n",
      "Combining Cort_ligand_receptor_target_gene_GP and CORT_ligand_receptor_GP.\n",
      "Combining Crh_ligand_receptor_target_gene_GP and CRH_ligand_receptor_GP.\n",
      "Combining Crlf1_ligand_receptor_target_gene_GP and CRLF1_ligand_receptor_GP.\n",
      "Combining Csf1_ligand_receptor_target_gene_GP and CSF1_ligand_receptor_GP.\n",
      "Combining Csf2_ligand_receptor_target_gene_GP and CSF2_ligand_receptor_GP.\n",
      "Combining Csf3_ligand_receptor_target_gene_GP and CSF3_ligand_receptor_GP.\n",
      "Combining Ctf1_ligand_receptor_target_gene_GP and CTF1_ligand_receptor_GP.\n",
      "Combining Ctsg_ligand_receptor_target_gene_GP and CTSG_ligand_receptor_GP.\n",
      "Combining Cx3cl1_ligand_receptor_target_gene_GP and CX3CL1_ligand_receptor_GP.\n",
      "Combining Cxcl1_ligand_receptor_target_gene_GP and CXCL1_ligand_receptor_GP.\n",
      "Combining Cxcl10_ligand_receptor_target_gene_GP and CXCL10_ligand_receptor_GP.\n",
      "Combining Cxcl11_ligand_receptor_target_gene_GP and CXCL11_ligand_receptor_GP.\n",
      "Combining Cxcl12_ligand_receptor_target_gene_GP and CXCL12_ligand_receptor_GP.\n",
      "Combining Cxcl13_ligand_receptor_target_gene_GP and CXCL13_ligand_receptor_GP.\n",
      "Combining Cxcl14_ligand_receptor_target_gene_GP and CXCL14_ligand_receptor_GP.\n",
      "Combining Cxcl17_ligand_receptor_target_gene_GP and CXCL17_ligand_receptor_GP.\n",
      "Combining Cxcl2_ligand_receptor_target_gene_GP and CXCL2_ligand_receptor_GP.\n",
      "Combining Cxcl5_ligand_receptor_target_gene_GP and CXCL5_ligand_receptor_GP.\n",
      "Combining Cxcl5_ligand_receptor_target_gene_GP and CXCL6_ligand_receptor_GP.\n",
      "Combining Cxcl9_ligand_receptor_target_gene_GP and CXCL9_ligand_receptor_GP.\n",
      "Combining Dbh_ligand_receptor_target_gene_GP and Norepinephrine_metabolite_enzyme_sensor_GP.\n",
      "Combining Dcn_ligand_receptor_target_gene_GP and DCN_ligand_receptor_GP.\n",
      "Combining Defb14_ligand_receptor_target_gene_GP and DEFB103A_ligand_receptor_GP.\n",
      "Combining Dhh_ligand_receptor_target_gene_GP and DHH_ligand_receptor_GP.\n",
      "Combining Dkk1_ligand_receptor_target_gene_GP and DKK1_ligand_receptor_GP.\n",
      "Combining Dkk2_ligand_receptor_target_gene_GP and DKK2_ligand_receptor_GP.\n",
      "Combining Dkk4_ligand_receptor_target_gene_GP and DKK4_ligand_receptor_GP.\n",
      "Combining Dlk1_ligand_receptor_target_gene_GP and DLK1_ligand_receptor_GP.\n",
      "Combining Dlk2_ligand_receptor_target_gene_GP and DLK2_ligand_receptor_GP.\n",
      "Combining Dll1_ligand_receptor_target_gene_GP and DLL1_ligand_receptor_GP.\n",
      "Combining Dll3_ligand_receptor_target_gene_GP and DLL3_ligand_receptor_GP.\n",
      "Combining Dll4_ligand_receptor_target_gene_GP and DLL4_ligand_receptor_GP.\n",
      "Combining Draxin_ligand_receptor_target_gene_GP and DRAXIN_ligand_receptor_GP.\n",
      "Combining Dsc1_ligand_receptor_target_gene_GP and DSC1_ligand_receptor_GP.\n",
      "Combining Dsc2_ligand_receptor_target_gene_GP and DSC2_ligand_receptor_GP.\n",
      "Combining Dsc3_ligand_receptor_target_gene_GP and DSC3_ligand_receptor_GP.\n",
      "Combining Ebi3_ligand_receptor_target_gene_GP and EBI3_ligand_receptor_GP.\n",
      "Combining Eda_ligand_receptor_target_gene_GP and EDA_ligand_receptor_GP.\n",
      "Combining Edil3_ligand_receptor_target_gene_GP and EDIL3_ligand_receptor_GP.\n",
      "Combining Edn1_ligand_receptor_target_gene_GP and EDN1_ligand_receptor_GP.\n",
      "Combining Edn2_ligand_receptor_target_gene_GP and EDN2_ligand_receptor_GP.\n",
      "Combining Edn3_ligand_receptor_target_gene_GP and EDN3_ligand_receptor_GP.\n",
      "Combining Efemp1_ligand_receptor_target_gene_GP and EFEMP1_ligand_receptor_GP.\n",
      "Combining Efemp2_ligand_receptor_target_gene_GP and EFEMP2_ligand_receptor_GP.\n",
      "Combining Efna1_ligand_receptor_target_gene_GP and EFNA1_ligand_receptor_GP.\n",
      "Combining Efna2_ligand_receptor_target_gene_GP and EFNA2_ligand_receptor_GP.\n",
      "Combining Efna3_ligand_receptor_target_gene_GP and EFNA3_ligand_receptor_GP.\n",
      "Combining Efna4_ligand_receptor_target_gene_GP and EFNA4_ligand_receptor_GP.\n",
      "Combining Efna5_ligand_receptor_target_gene_GP and EFNA5_ligand_receptor_GP.\n",
      "Combining Efnb1_ligand_receptor_target_gene_GP and EFNB1_ligand_receptor_GP.\n",
      "Combining Efnb2_ligand_receptor_target_gene_GP and EFNB2_ligand_receptor_GP.\n",
      "Combining Efnb3_ligand_receptor_target_gene_GP and EFNB3_ligand_receptor_GP.\n",
      "Combining Egf_ligand_receptor_target_gene_GP and EGF_ligand_receptor_GP.\n",
      "Combining Enho_ligand_receptor_target_gene_GP and ENHO_ligand_receptor_GP.\n",
      "Combining Epgn_ligand_receptor_target_gene_GP and EPGN_ligand_receptor_GP.\n",
      "Combining Epo_ligand_receptor_target_gene_GP and EPO_ligand_receptor_GP.\n",
      "Combining Ereg_ligand_receptor_target_gene_GP and EREG_ligand_receptor_GP.\n",
      "Combining Fasl_ligand_receptor_target_gene_GP and FASLG_ligand_receptor_GP.\n",
      "Combining Fbln1_ligand_receptor_target_gene_GP and FBLN1_ligand_receptor_GP.\n",
      "Combining Fbln2_ligand_receptor_target_gene_GP and FBLN2_ligand_receptor_GP.\n",
      "Combining Fgf1_ligand_receptor_target_gene_GP and FGF1_ligand_receptor_GP.\n",
      "Combining Fgf10_ligand_receptor_target_gene_GP and FGF10_ligand_receptor_GP.\n",
      "Combining Fgf11_ligand_receptor_target_gene_GP and FGF11_ligand_receptor_GP.\n",
      "Combining Fgf12_ligand_receptor_target_gene_GP and FGF12_ligand_receptor_GP.\n",
      "Combining Fgf13_ligand_receptor_target_gene_GP and FGF13_ligand_receptor_GP.\n",
      "Combining Fgf14_ligand_receptor_target_gene_GP and FGF14_ligand_receptor_GP.\n",
      "Combining Fgf15_ligand_receptor_target_gene_GP and FGF19_ligand_receptor_GP.\n",
      "Combining Fgf16_ligand_receptor_target_gene_GP and FGF16_ligand_receptor_GP.\n",
      "Combining Fgf17_ligand_receptor_target_gene_GP and FGF17_ligand_receptor_GP.\n",
      "Combining Fgf18_ligand_receptor_target_gene_GP and FGF18_ligand_receptor_GP.\n",
      "Combining Fgf2_ligand_receptor_target_gene_GP and FGF2_ligand_receptor_GP.\n",
      "Combining Fgf20_ligand_receptor_target_gene_GP and FGF20_ligand_receptor_GP.\n",
      "Combining Fgf21_ligand_receptor_target_gene_GP and FGF21_ligand_receptor_GP.\n",
      "Combining Fgf22_ligand_receptor_target_gene_GP and FGF22_ligand_receptor_GP.\n",
      "Combining Fgf23_ligand_receptor_target_gene_GP and FGF23_ligand_receptor_GP.\n",
      "Combining Fgf3_ligand_receptor_target_gene_GP and FGF3_ligand_receptor_GP.\n",
      "Combining Fgf4_ligand_receptor_target_gene_GP and FGF4_ligand_receptor_GP.\n",
      "Combining Fgf5_ligand_receptor_target_gene_GP and FGF5_ligand_receptor_GP.\n",
      "Combining Fgf6_ligand_receptor_target_gene_GP and FGF6_ligand_receptor_GP.\n",
      "Combining Fgf7_ligand_receptor_target_gene_GP and FGF7_ligand_receptor_GP.\n",
      "Combining Fgf8_ligand_receptor_target_gene_GP and FGF8_ligand_receptor_GP.\n",
      "Combining Fgf9_ligand_receptor_target_gene_GP and FGF9_ligand_receptor_GP.\n",
      "Combining Fgl1_ligand_receptor_target_gene_GP and FGL1_ligand_receptor_GP.\n",
      "Combining Flrt3_ligand_receptor_target_gene_GP and FLRT3_ligand_receptor_GP.\n",
      "Combining Flt3l_ligand_receptor_target_gene_GP and FLT3LG_ligand_receptor_GP.\n",
      "Combining Fshb_ligand_receptor_target_gene_GP and FSHB_ligand_receptor_GP.\n",
      "Combining Gal_ligand_receptor_target_gene_GP and GAL_ligand_receptor_GP.\n",
      "Combining Galp_ligand_receptor_target_gene_GP and GALP_ligand_receptor_GP.\n",
      "Combining Gas6_ligand_receptor_target_gene_GP and GAS6_ligand_receptor_GP.\n",
      "Combining Gast_ligand_receptor_target_gene_GP and GAST_ligand_receptor_GP.\n",
      "Combining Gc_ligand_receptor_target_gene_GP and GC_ligand_receptor_GP.\n",
      "Combining Gcg_ligand_receptor_target_gene_GP and GCG_ligand_receptor_GP.\n",
      "Combining Gdf10_ligand_receptor_target_gene_GP and GDF10_ligand_receptor_GP.\n",
      "Combining Gdf11_ligand_receptor_target_gene_GP and GDF11_ligand_receptor_GP.\n",
      "Combining Gdf15_ligand_receptor_target_gene_GP and GDF15_ligand_receptor_GP.\n",
      "Combining Gdf2_ligand_receptor_target_gene_GP and GDF2_ligand_receptor_GP.\n",
      "Combining Gdf3_ligand_receptor_target_gene_GP and GDF3_ligand_receptor_GP.\n",
      "Combining Gdf5_ligand_receptor_target_gene_GP and GDF5_ligand_receptor_GP.\n",
      "Combining Gdf6_ligand_receptor_target_gene_GP and GDF6_ligand_receptor_GP.\n",
      "Combining Gdf7_ligand_receptor_target_gene_GP and GDF7_ligand_receptor_GP.\n",
      "Combining Gdf9_ligand_receptor_target_gene_GP and GDF9_ligand_receptor_GP.\n",
      "Combining Gdnf_ligand_receptor_target_gene_GP and GDNF_ligand_receptor_GP.\n",
      "Combining Gh_ligand_receptor_target_gene_GP and CSH1_ligand_receptor_GP.\n",
      "Combining Gh_ligand_receptor_target_gene_GP and GH1_ligand_receptor_GP.\n",
      "Combining Gh_ligand_receptor_target_gene_GP and GH2_ligand_receptor_GP.\n",
      "Combining Ghrh_ligand_receptor_target_gene_GP and GHRH_ligand_receptor_GP.\n",
      "Combining Ghrl_ligand_receptor_target_gene_GP and GHRL_ligand_receptor_GP.\n",
      "Combining Gip_ligand_receptor_target_gene_GP and GIP_ligand_receptor_GP.\n",
      "Combining Gnrh1_ligand_receptor_target_gene_GP and GNRH1_ligand_receptor_GP.\n",
      "Combining Gpha2_ligand_receptor_target_gene_GP and GPHA2_ligand_receptor_GP.\n",
      "Combining Grem1_ligand_receptor_target_gene_GP and GREM1_ligand_receptor_GP.\n",
      "Combining Grn_ligand_receptor_target_gene_GP and GRN_ligand_receptor_GP.\n",
      "Combining Grp_ligand_receptor_target_gene_GP and GRP_ligand_receptor_GP.\n",
      "Combining Guca2a_ligand_receptor_target_gene_GP and GUCA2A_ligand_receptor_GP.\n",
      "Combining Guca2b_ligand_receptor_target_gene_GP and GUCA2B_ligand_receptor_GP.\n",
      "Combining H2-Ab1_ligand_receptor_target_gene_GP and HLA-DQB1_ligand_receptor_GP.\n",
      "Combining H2-Ab1_ligand_receptor_target_gene_GP and HLA-DQB2_ligand_receptor_GP.\n",
      "Combining H2-DMa_ligand_receptor_target_gene_GP and HLA-DMA_ligand_receptor_GP.\n",
      "Combining H2-Ea_ligand_receptor_target_gene_GP and HLA-DRA_ligand_receptor_GP.\n",
      "Combining H2-Eb2_ligand_receptor_target_gene_GP and HLA-DRB1_ligand_receptor_GP.\n",
      "Combining H2-Eb2_ligand_receptor_target_gene_GP and HLA-DRB5_ligand_receptor_GP.\n",
      "Combining Hbegf_ligand_receptor_target_gene_GP and HBEGF_ligand_receptor_GP.\n",
      "Combining Hcrt_ligand_receptor_target_gene_GP and HCRT_ligand_receptor_GP.\n",
      "Combining Hgf_ligand_receptor_target_gene_GP and HGF_ligand_receptor_GP.\n",
      "Combining Hp_ligand_receptor_target_gene_GP and HP_ligand_receptor_GP.\n",
      "Combining Iapp_ligand_receptor_target_gene_GP and IAPP_ligand_receptor_GP.\n",
      "Combining Icam1_ligand_receptor_target_gene_GP and ICAM1_ligand_receptor_GP.\n",
      "Combining Icam2_ligand_receptor_target_gene_GP and ICAM2_ligand_receptor_GP.\n",
      "Combining Icam4_ligand_receptor_target_gene_GP and ICAM4_ligand_receptor_GP.\n",
      "Combining Icam5_ligand_receptor_target_gene_GP and ICAM5_ligand_receptor_GP.\n",
      "Combining Icosl_ligand_receptor_target_gene_GP and ICOSLG_ligand_receptor_GP.\n",
      "Combining Ifnb1_ligand_receptor_target_gene_GP and IFNB1_ligand_receptor_GP.\n",
      "Combining Ifne_ligand_receptor_target_gene_GP and IFNE_ligand_receptor_GP.\n",
      "Combining Ifng_ligand_receptor_target_gene_GP and IFNG_ligand_receptor_GP.\n",
      "Combining Ifnk_ligand_receptor_target_gene_GP and IFNK_ligand_receptor_GP.\n",
      "Combining Igf1_ligand_receptor_target_gene_GP and IGF1_ligand_receptor_GP.\n",
      "Combining Igf2_ligand_receptor_target_gene_GP and IGF2_ligand_receptor_GP.\n",
      "Combining Igfl3_ligand_receptor_target_gene_GP and IGFL2_ligand_receptor_GP.\n",
      "Combining Igfl3_ligand_receptor_target_gene_GP and IGFL3_ligand_receptor_GP.\n",
      "Combining Ihh_ligand_receptor_target_gene_GP and IHH_ligand_receptor_GP.\n",
      "Combining Il10_ligand_receptor_target_gene_GP and IL10_ligand_receptor_GP.\n",
      "Combining Il11_ligand_receptor_target_gene_GP and IL11_ligand_receptor_GP.\n",
      "Combining Il12a_ligand_receptor_target_gene_GP and IL12A_ligand_receptor_GP.\n",
      "Combining Il12b_ligand_receptor_target_gene_GP and IL12B_ligand_receptor_GP.\n",
      "Combining Il13_ligand_receptor_target_gene_GP and IL13_ligand_receptor_GP.\n",
      "Combining Il15_ligand_receptor_target_gene_GP and IL15_ligand_receptor_GP.\n",
      "Combining Il16_ligand_receptor_target_gene_GP and IL16_ligand_receptor_GP.\n",
      "Combining Il17a_ligand_receptor_target_gene_GP and IL17A_ligand_receptor_GP.\n",
      "Combining Il17b_ligand_receptor_target_gene_GP and IL17B_ligand_receptor_GP.\n",
      "Combining Il17c_ligand_receptor_target_gene_GP and IL17C_ligand_receptor_GP.\n",
      "Combining Il17f_ligand_receptor_target_gene_GP and IL17F_ligand_receptor_GP.\n",
      "Combining Il18_ligand_receptor_target_gene_GP and IL18_ligand_receptor_GP.\n",
      "Combining Il19_ligand_receptor_target_gene_GP and IL19_ligand_receptor_GP.\n",
      "Combining Il1a_ligand_receptor_target_gene_GP and IL1A_ligand_receptor_GP.\n",
      "Combining Il1b_ligand_receptor_target_gene_GP and IL1B_ligand_receptor_GP.\n",
      "Combining Il1f10_ligand_receptor_target_gene_GP and IL1F10_ligand_receptor_GP.\n",
      "Combining Il1rn_ligand_receptor_target_gene_GP and IL1RN_ligand_receptor_GP.\n",
      "Combining Il2_ligand_receptor_target_gene_GP and IL2_ligand_receptor_GP.\n",
      "Combining Il20_ligand_receptor_target_gene_GP and IL20_ligand_receptor_GP.\n",
      "Combining Il21_ligand_receptor_target_gene_GP and IL21_ligand_receptor_GP.\n",
      "Combining Il23a_ligand_receptor_target_gene_GP and IL23A_ligand_receptor_GP.\n",
      "Combining Il24_ligand_receptor_target_gene_GP and IL24_ligand_receptor_GP.\n",
      "Combining Il25_ligand_receptor_target_gene_GP and IL25_ligand_receptor_GP.\n",
      "Combining Il27_ligand_receptor_target_gene_GP and IL27_ligand_receptor_GP.\n",
      "Combining Il3_ligand_receptor_target_gene_GP and IL3_ligand_receptor_GP.\n",
      "Combining Il31_ligand_receptor_target_gene_GP and IL31_ligand_receptor_GP.\n",
      "Combining Il33_ligand_receptor_target_gene_GP and IL33_ligand_receptor_GP.\n",
      "Combining Il34_ligand_receptor_target_gene_GP and IL34_ligand_receptor_GP.\n",
      "Combining Il36a_ligand_receptor_target_gene_GP and IL36A_ligand_receptor_GP.\n",
      "Combining Il36b_ligand_receptor_target_gene_GP and IL36B_ligand_receptor_GP.\n",
      "Combining Il36g_ligand_receptor_target_gene_GP and IL36G_ligand_receptor_GP.\n",
      "Combining Il36rn_ligand_receptor_target_gene_GP and IL36RN_ligand_receptor_GP.\n",
      "Combining Il4_ligand_receptor_target_gene_GP and IL4_ligand_receptor_GP.\n",
      "Combining Il5_ligand_receptor_target_gene_GP and IL5_ligand_receptor_GP.\n",
      "Combining Il6_ligand_receptor_target_gene_GP and IL6_ligand_receptor_GP.\n",
      "Combining Il7_ligand_receptor_target_gene_GP and IL7_ligand_receptor_GP.\n",
      "Combining Il9_ligand_receptor_target_gene_GP and IL9_ligand_receptor_GP.\n",
      "Combining Inha_ligand_receptor_target_gene_GP and INHA_ligand_receptor_GP.\n",
      "Combining Inhba_ligand_receptor_target_gene_GP and INHBA_ligand_receptor_GP.\n",
      "Combining Inhbb_ligand_receptor_target_gene_GP and INHBB_ligand_receptor_GP.\n",
      "Combining Inhbc_ligand_receptor_target_gene_GP and INHBC_ligand_receptor_GP.\n",
      "Combining Inhbe_ligand_receptor_target_gene_GP and INHBE_ligand_receptor_GP.\n",
      "Combining Insl3_ligand_receptor_target_gene_GP and INSL3_ligand_receptor_GP.\n",
      "Combining Insl5_ligand_receptor_target_gene_GP and INSL5_ligand_receptor_GP.\n",
      "Combining Izumo1_ligand_receptor_target_gene_GP and IZUMO1_ligand_receptor_GP.\n",
      "Combining Jag1_ligand_receptor_target_gene_GP and JAG1_ligand_receptor_GP.\n",
      "Combining Jag2_ligand_receptor_target_gene_GP and JAG2_ligand_receptor_GP.\n",
      "Combining Jam2_ligand_receptor_target_gene_GP and JAM2_ligand_receptor_GP.\n",
      "Combining Jam3_ligand_receptor_target_gene_GP and JAM3_ligand_receptor_GP.\n",
      "Combining Jaml_ligand_receptor_target_gene_GP and JAML_ligand_receptor_GP.\n",
      "Combining Kitl_ligand_receptor_target_gene_GP and KITLG_ligand_receptor_GP.\n",
      "Combining Kl_ligand_receptor_target_gene_GP and KL_ligand_receptor_GP.\n",
      "Combining L1cam_ligand_receptor_target_gene_GP and L1CAM_ligand_receptor_GP.\n",
      "Combining Lama1_ligand_receptor_target_gene_GP and LAMA1_ligand_receptor_GP.\n",
      "Combining Lama2_ligand_receptor_target_gene_GP and LAMA2_ligand_receptor_GP.\n",
      "Combining Lama3_ligand_receptor_target_gene_GP and LAMA3_ligand_receptor_GP.\n",
      "Combining Lama5_ligand_receptor_target_gene_GP and LAMA5_ligand_receptor_GP.\n",
      "Combining Lamb1_ligand_receptor_target_gene_GP and LAMB1_ligand_receptor_GP.\n",
      "Combining Lamb2_ligand_receptor_target_gene_GP and LAMB2_ligand_receptor_GP.\n",
      "Combining Lamb3_ligand_receptor_target_gene_GP and LAMB3_ligand_receptor_GP.\n",
      "Combining Lamc1_ligand_receptor_target_gene_GP and LAMC1_ligand_receptor_GP.\n",
      "Combining Lamc2_ligand_receptor_target_gene_GP and LAMC2_ligand_receptor_GP.\n",
      "Combining Leap2_ligand_receptor_target_gene_GP and LEAP2_ligand_receptor_GP.\n",
      "Combining Lep_ligand_receptor_target_gene_GP and LEP_ligand_receptor_GP.\n",
      "Combining Lfng_ligand_receptor_target_gene_GP and LFNG_ligand_receptor_GP.\n",
      "Combining Lgals1_ligand_receptor_target_gene_GP and LGALS1_ligand_receptor_GP.\n",
      "Combining Lgals3_ligand_receptor_target_gene_GP and LGALS3_ligand_receptor_GP.\n",
      "Combining Lgals3bp_ligand_receptor_target_gene_GP and LGALS3BP_ligand_receptor_GP.\n",
      "Combining Lgi3_ligand_receptor_target_gene_GP and LGI3_ligand_receptor_GP.\n",
      "Combining Lgi4_ligand_receptor_target_gene_GP and LGI4_ligand_receptor_GP.\n",
      "Combining Lhb_ligand_receptor_target_gene_GP and LHB_ligand_receptor_GP.\n",
      "Combining Lif_ligand_receptor_target_gene_GP and LIF_ligand_receptor_GP.\n",
      "Combining Lrig1_ligand_receptor_target_gene_GP and LRIG1_ligand_receptor_GP.\n",
      "Combining Lrrc4b_ligand_receptor_target_gene_GP and LRRC4B_ligand_receptor_GP.\n",
      "Combining Lta_ligand_receptor_target_gene_GP and LTA_ligand_receptor_GP.\n",
      "Combining Ltb_ligand_receptor_target_gene_GP and LTB_ligand_receptor_GP.\n",
      "Combining Ltbp3_ligand_receptor_target_gene_GP and LTBP3_ligand_receptor_GP.\n",
      "Combining Mdk_ligand_receptor_target_gene_GP and MDK_ligand_receptor_GP.\n",
      "Combining Mfap2_ligand_receptor_target_gene_GP and MFAP2_ligand_receptor_GP.\n",
      "Combining Mfge8_ligand_receptor_target_gene_GP and MFGE8_ligand_receptor_GP.\n",
      "Combining Mia_ligand_receptor_target_gene_GP and MIA_ligand_receptor_GP.\n",
      "Combining Mif_ligand_receptor_target_gene_GP and MIF_ligand_receptor_GP.\n",
      "Combining Mst1_ligand_receptor_target_gene_GP and MST1_ligand_receptor_GP.\n",
      "Combining Mstn_ligand_receptor_target_gene_GP and MSTN_ligand_receptor_GP.\n",
      "Combining Muc1_ligand_receptor_target_gene_GP and MUC1_ligand_receptor_GP.\n",
      "Combining Myoc_ligand_receptor_target_gene_GP and MYOC_ligand_receptor_GP.\n",
      "Combining Ndp_ligand_receptor_target_gene_GP and NDP_ligand_receptor_GP.\n",
      "Combining Nectin2_ligand_receptor_target_gene_GP and NECTIN2_ligand_receptor_GP.\n",
      "Combining Nectin3_ligand_receptor_target_gene_GP and NECTIN3_ligand_receptor_GP.\n",
      "Combining Nectin4_ligand_receptor_target_gene_GP and NECTIN4_ligand_receptor_GP.\n",
      "Combining Nell2_ligand_receptor_target_gene_GP and NELL2_ligand_receptor_GP.\n",
      "Combining Ngf_ligand_receptor_target_gene_GP and NGF_ligand_receptor_GP.\n",
      "Combining Nid1_ligand_receptor_target_gene_GP and NID1_ligand_receptor_GP.\n",
      "Combining Nlgn1_ligand_receptor_target_gene_GP and NLGN1_ligand_receptor_GP.\n",
      "Combining Nlgn2_ligand_receptor_target_gene_GP and NLGN2_ligand_receptor_GP.\n",
      "Combining Nlgn3_ligand_receptor_target_gene_GP and NLGN3_ligand_receptor_GP.\n",
      "Combining Nmb_ligand_receptor_target_gene_GP and NMB_ligand_receptor_GP.\n",
      "Combining Nms_ligand_receptor_target_gene_GP and NMS_ligand_receptor_GP.\n",
      "Combining Nmu_ligand_receptor_target_gene_GP and NMU_ligand_receptor_GP.\n",
      "Combining Nodal_ligand_receptor_target_gene_GP and NODAL_ligand_receptor_GP.\n",
      "Combining Nog_ligand_receptor_target_gene_GP and NOG_ligand_receptor_GP.\n",
      "Combining Npb_ligand_receptor_target_gene_GP and NPB_ligand_receptor_GP.\n",
      "Combining Npff_ligand_receptor_target_gene_GP and NPFF_ligand_receptor_GP.\n",
      "Combining Npnt_ligand_receptor_target_gene_GP and NPNT_ligand_receptor_GP.\n",
      "Combining Nppa_ligand_receptor_target_gene_GP and NPPA_ligand_receptor_GP.\n",
      "Combining Nppb_ligand_receptor_target_gene_GP and NPPB_ligand_receptor_GP.\n",
      "Combining Nppc_ligand_receptor_target_gene_GP and NPPC_ligand_receptor_GP.\n",
      "Combining Nps_ligand_receptor_target_gene_GP and NPS_ligand_receptor_GP.\n",
      "Combining Nptx2_ligand_receptor_target_gene_GP and NPTX2_ligand_receptor_GP.\n",
      "Combining Npvf_ligand_receptor_target_gene_GP and NPVF_ligand_receptor_GP.\n",
      "Combining Npw_ligand_receptor_target_gene_GP and NPW_ligand_receptor_GP.\n",
      "Combining Npy_ligand_receptor_target_gene_GP and NPY_ligand_receptor_GP.\n",
      "Combining Nrg1_ligand_receptor_target_gene_GP and NRG1_ligand_receptor_GP.\n",
      "Combining Nrg2_ligand_receptor_target_gene_GP and NRG2_ligand_receptor_GP.\n",
      "Combining Nrg3_ligand_receptor_target_gene_GP and NRG3_ligand_receptor_GP.\n",
      "Combining Nrg4_ligand_receptor_target_gene_GP and NRG4_ligand_receptor_GP.\n",
      "Combining Nrp1_ligand_receptor_target_gene_GP and NRP1_ligand_receptor_GP.\n",
      "Combining Nrtn_ligand_receptor_target_gene_GP and NRTN_ligand_receptor_GP.\n",
      "Combining Ntf3_ligand_receptor_target_gene_GP and NTF3_ligand_receptor_GP.\n",
      "Combining Ntf5_ligand_receptor_target_gene_GP and NTF4_ligand_receptor_GP.\n",
      "Combining Ntn1_ligand_receptor_target_gene_GP and NTN1_ligand_receptor_GP.\n",
      "Combining Ntn3_ligand_receptor_target_gene_GP and NTN3_ligand_receptor_GP.\n",
      "Combining Ntn4_ligand_receptor_target_gene_GP and NTN4_ligand_receptor_GP.\n",
      "Combining Ntng2_ligand_receptor_target_gene_GP and NTNG2_ligand_receptor_GP.\n",
      "Combining Nts_ligand_receptor_target_gene_GP and NTS_ligand_receptor_GP.\n",
      "Combining Nxph1_ligand_receptor_target_gene_GP and NXPH1_ligand_receptor_GP.\n",
      "Combining Nxph2_ligand_receptor_target_gene_GP and NXPH2_ligand_receptor_GP.\n",
      "Combining Nxph3_ligand_receptor_target_gene_GP and NXPH3_ligand_receptor_GP.\n",
      "Combining Ocln_ligand_receptor_target_gene_GP and OCLN_ligand_receptor_GP.\n",
      "Combining Olfm2_ligand_receptor_target_gene_GP and OLFM2_ligand_receptor_GP.\n",
      "Combining Osm_ligand_receptor_target_gene_GP and OSM_ligand_receptor_GP.\n",
      "Combining Ostn_ligand_receptor_target_gene_GP and OSTN_ligand_receptor_GP.\n",
      "Combining Oxt_ligand_receptor_target_gene_GP and OXT_ligand_receptor_GP.\n",
      "Combining Pam_ligand_receptor_target_gene_GP and PAM_ligand_receptor_GP.\n",
      "Combining Pcsk9_ligand_receptor_target_gene_GP and PCSK9_ligand_receptor_GP.\n",
      "Combining Pdcd1lg2_ligand_receptor_target_gene_GP and PDCD1LG2_ligand_receptor_GP.\n",
      "Combining Pdgfa_ligand_receptor_target_gene_GP and PDGFA_ligand_receptor_GP.\n",
      "Combining Pdgfb_ligand_receptor_target_gene_GP and PDGFB_ligand_receptor_GP.\n",
      "Combining Pdgfc_ligand_receptor_target_gene_GP and PDGFC_ligand_receptor_GP.\n",
      "Combining Pdgfd_ligand_receptor_target_gene_GP and PDGFD_ligand_receptor_GP.\n",
      "Combining Pdyn_ligand_receptor_target_gene_GP and PDYN_ligand_receptor_GP.\n",
      "Combining Pecam1_ligand_receptor_target_gene_GP and PECAM1_ligand_receptor_GP.\n",
      "Combining Penk_ligand_receptor_target_gene_GP and PENK_ligand_receptor_GP.\n",
      "Combining Pf4_ligand_receptor_target_gene_GP and PF4_ligand_receptor_GP.\n",
      "Combining Pgf_ligand_receptor_target_gene_GP and PGF_ligand_receptor_GP.\n",
      "Combining Pglyrp1_ligand_receptor_target_gene_GP and PGLYRP1_ligand_receptor_GP.\n",
      "Combining Pip_ligand_receptor_target_gene_GP and PIP_ligand_receptor_GP.\n",
      "Combining Plg_ligand_receptor_target_gene_GP and PLG_ligand_receptor_GP.\n",
      "Combining Pmch_ligand_receptor_target_gene_GP and PMCH_ligand_receptor_GP.\n",
      "Combining Pnoc_ligand_receptor_target_gene_GP and PNOC_ligand_receptor_GP.\n",
      "Combining Podxl2_ligand_receptor_target_gene_GP and PODXL2_ligand_receptor_GP.\n",
      "Combining Pomc_ligand_receptor_target_gene_GP and POMC_ligand_receptor_GP.\n",
      "Combining Postn_ligand_receptor_target_gene_GP and POSTN_ligand_receptor_GP.\n",
      "Combining Ppbp_ligand_receptor_target_gene_GP and PPBP_ligand_receptor_GP.\n",
      "Combining Ppy_ligand_receptor_target_gene_GP and PPY_ligand_receptor_GP.\n",
      "Combining Prlh_ligand_receptor_target_gene_GP and PRLH_ligand_receptor_GP.\n",
      "Combining Prnd_ligand_receptor_target_gene_GP and PRND_ligand_receptor_GP.\n",
      "Combining Prok1_ligand_receptor_target_gene_GP and PROK1_ligand_receptor_GP.\n",
      "Combining Prok2_ligand_receptor_target_gene_GP and PROK2_ligand_receptor_GP.\n",
      "Combining Pspn_ligand_receptor_target_gene_GP and PSPN_ligand_receptor_GP.\n",
      "Combining Pth_ligand_receptor_target_gene_GP and PTH_ligand_receptor_GP.\n",
      "Combining Pth2_ligand_receptor_target_gene_GP and PTH2_ligand_receptor_GP.\n",
      "Combining Pthlh_ligand_receptor_target_gene_GP and PTHLH_ligand_receptor_GP.\n",
      "Combining Ptn_ligand_receptor_target_gene_GP and PTN_ligand_receptor_GP.\n",
      "Combining Pyy_ligand_receptor_target_gene_GP and PYY_ligand_receptor_GP.\n",
      "Combining Qrfp_ligand_receptor_target_gene_GP and QRFP_ligand_receptor_GP.\n",
      "Combining Raet1e_ligand_receptor_target_gene_GP and RAET1E_ligand_receptor_GP.\n",
      "Combining Rarres2_ligand_receptor_target_gene_GP and RARRES2_ligand_receptor_GP.\n",
      "Combining Rbp3_ligand_receptor_target_gene_GP and RBP3_ligand_receptor_GP.\n",
      "Combining Reln_ligand_receptor_target_gene_GP and RELN_ligand_receptor_GP.\n",
      "Combining Retn_ligand_receptor_target_gene_GP and RETN_ligand_receptor_GP.\n",
      "Combining Rgma_ligand_receptor_target_gene_GP and RGMA_ligand_receptor_GP.\n",
      "Combining Rln1_ligand_receptor_target_gene_GP and RLN1_ligand_receptor_GP.\n",
      "Combining Rln1_ligand_receptor_target_gene_GP and RLN2_ligand_receptor_GP.\n",
      "Combining Rln3_ligand_receptor_target_gene_GP and RLN3_ligand_receptor_GP.\n",
      "Combining Rspo1_ligand_receptor_target_gene_GP and RSPO1_ligand_receptor_GP.\n",
      "Combining Rspo2_ligand_receptor_target_gene_GP and RSPO2_ligand_receptor_GP.\n",
      "Combining Rspo3_ligand_receptor_target_gene_GP and RSPO3_ligand_receptor_GP.\n",
      "Combining Rspo4_ligand_receptor_target_gene_GP and RSPO4_ligand_receptor_GP.\n",
      "Combining S100a10_ligand_receptor_target_gene_GP and S100A10_ligand_receptor_GP.\n",
      "Combining S100a4_ligand_receptor_target_gene_GP and S100A4_ligand_receptor_GP.\n",
      "Combining S100a8_ligand_receptor_target_gene_GP and S100A8_ligand_receptor_GP.\n",
      "Combining S100b_ligand_receptor_target_gene_GP and S100B_ligand_receptor_GP.\n",
      "Combining Scgb1a1_ligand_receptor_target_gene_GP and SCGB1A1_ligand_receptor_GP.\n",
      "Combining Scgb3a1_ligand_receptor_target_gene_GP and SCGB3A1_ligand_receptor_GP.\n",
      "Combining Scgb3a2_ligand_receptor_target_gene_GP and SCGB3A2_ligand_receptor_GP.\n",
      "Combining Sct_ligand_receptor_target_gene_GP and SCT_ligand_receptor_GP.\n",
      "Combining Sdc2_ligand_receptor_target_gene_GP and SDC2_ligand_receptor_GP.\n",
      "Combining Sema3a_ligand_receptor_target_gene_GP and SEMA3A_ligand_receptor_GP.\n",
      "Combining Sema3b_ligand_receptor_target_gene_GP and SEMA3B_ligand_receptor_GP.\n",
      "Combining Sema3c_ligand_receptor_target_gene_GP and SEMA3C_ligand_receptor_GP.\n",
      "Combining Sema3d_ligand_receptor_target_gene_GP and SEMA3D_ligand_receptor_GP.\n",
      "Combining Sema3e_ligand_receptor_target_gene_GP and SEMA3E_ligand_receptor_GP.\n",
      "Combining Sema3f_ligand_receptor_target_gene_GP and SEMA3F_ligand_receptor_GP.\n",
      "Combining Sema3g_ligand_receptor_target_gene_GP and SEMA3G_ligand_receptor_GP.\n",
      "Combining Sema4b_ligand_receptor_target_gene_GP and SEMA4B_ligand_receptor_GP.\n",
      "Combining Sema5a_ligand_receptor_target_gene_GP and SEMA5A_ligand_receptor_GP.\n",
      "Combining Sema5b_ligand_receptor_target_gene_GP and SEMA5B_ligand_receptor_GP.\n",
      "Combining Sema6b_ligand_receptor_target_gene_GP and SEMA6B_ligand_receptor_GP.\n",
      "Combining Sema6d_ligand_receptor_target_gene_GP and SEMA6D_ligand_receptor_GP.\n",
      "Combining Sema7a_ligand_receptor_target_gene_GP and SEMA7A_ligand_receptor_GP.\n",
      "Combining Sfrp2_ligand_receptor_target_gene_GP and SFRP2_ligand_receptor_GP.\n",
      "Combining Sftpa1_ligand_receptor_target_gene_GP and SFTPA1_ligand_receptor_GP.\n",
      "Combining Sftpd_ligand_receptor_target_gene_GP and SFTPD_ligand_receptor_GP.\n",
      "Combining Shh_ligand_receptor_target_gene_GP and SHH_ligand_receptor_GP.\n",
      "Combining Siglece_ligand_receptor_target_gene_GP and SIGLEC9_ligand_receptor_GP.\n",
      "Combining Slit1_ligand_receptor_target_gene_GP and SLIT1_ligand_receptor_GP.\n",
      "Combining Slit2_ligand_receptor_target_gene_GP and SLIT2_ligand_receptor_GP.\n",
      "Combining Slit3_ligand_receptor_target_gene_GP and SLIT3_ligand_receptor_GP.\n",
      "Combining Slitrk1_ligand_receptor_target_gene_GP and SLITRK1_ligand_receptor_GP.\n",
      "Combining Slurp1_ligand_receptor_target_gene_GP and SLURP1_ligand_receptor_GP.\n",
      "Combining Sorl1_ligand_receptor_target_gene_GP and SORL1_ligand_receptor_GP.\n",
      "Combining Sost_ligand_receptor_target_gene_GP and SOST_ligand_receptor_GP.\n",
      "Combining Spon1_ligand_receptor_target_gene_GP and SPON1_ligand_receptor_GP.\n",
      "Combining Spon2_ligand_receptor_target_gene_GP and SPON2_ligand_receptor_GP.\n",
      "Combining Spp1_ligand_receptor_target_gene_GP and SPP1_ligand_receptor_GP.\n",
      "Combining Spx_ligand_receptor_target_gene_GP and SPX_ligand_receptor_GP.\n",
      "Combining Sst_ligand_receptor_target_gene_GP and SST_ligand_receptor_GP.\n",
      "Combining Tac1_ligand_receptor_target_gene_GP and TAC1_ligand_receptor_GP.\n",
      "Combining Tac2_ligand_receptor_target_gene_GP and TAC3_ligand_receptor_GP.\n",
      "Combining Tac4_ligand_receptor_target_gene_GP and TAC4_ligand_receptor_GP.\n",
      "Combining Tff1_ligand_receptor_target_gene_GP and TFF1_ligand_receptor_GP.\n",
      "Combining Tff2_ligand_receptor_target_gene_GP and TFF2_ligand_receptor_GP.\n",
      "Combining Tg_ligand_receptor_target_gene_GP and TG_ligand_receptor_GP.\n",
      "Combining Tgfa_ligand_receptor_target_gene_GP and TGFA_ligand_receptor_GP.\n",
      "Combining Tgfb1_ligand_receptor_target_gene_GP and TGFB1_ligand_receptor_GP.\n",
      "Combining Tgfb2_ligand_receptor_target_gene_GP and TGFB2_ligand_receptor_GP.\n",
      "Combining Tgfb3_ligand_receptor_target_gene_GP and TGFB3_ligand_receptor_GP.\n",
      "Combining Thbs1_ligand_receptor_target_gene_GP and THBS1_ligand_receptor_GP.\n",
      "Combining Thbs2_ligand_receptor_target_gene_GP and THBS2_ligand_receptor_GP.\n",
      "Combining Thpo_ligand_receptor_target_gene_GP and THPO_ligand_receptor_GP.\n",
      "Combining Tnc_ligand_receptor_target_gene_GP and TNC_ligand_receptor_GP.\n",
      "Combining Tnf_ligand_receptor_target_gene_GP and TNF_ligand_receptor_GP.\n",
      "Combining Tnfsf10_ligand_receptor_target_gene_GP and TNFSF10_ligand_receptor_GP.\n",
      "Combining Tnfsf11_ligand_receptor_target_gene_GP and TNFSF11_ligand_receptor_GP.\n",
      "Combining Tnfsf12_ligand_receptor_target_gene_GP and TNFSF12_ligand_receptor_GP.\n",
      "Combining Tnfsf13_ligand_receptor_target_gene_GP and TNFSF13_ligand_receptor_GP.\n",
      "Combining Tnfsf13b_ligand_receptor_target_gene_GP and TNFSF13B_ligand_receptor_GP.\n",
      "Combining Tnfsf14_ligand_receptor_target_gene_GP and TNFSF14_ligand_receptor_GP.\n",
      "Combining Tnfsf15_ligand_receptor_target_gene_GP and TNFSF15_ligand_receptor_GP.\n",
      "Combining Tnfsf18_ligand_receptor_target_gene_GP and TNFSF18_ligand_receptor_GP.\n",
      "Combining Tnfsf4_ligand_receptor_target_gene_GP and TNFSF4_ligand_receptor_GP.\n",
      "Combining Tnfsf8_ligand_receptor_target_gene_GP and TNFSF8_ligand_receptor_GP.\n",
      "Combining Tnfsf9_ligand_receptor_target_gene_GP and TNFSF9_ligand_receptor_GP.\n",
      "Combining Tpo_ligand_receptor_target_gene_GP and Liothyronine_metabolite_enzyme_sensor_GP.\n",
      "Combining Trf_ligand_receptor_target_gene_GP and TF_ligand_receptor_GP.\n",
      "Combining Trh_ligand_receptor_target_gene_GP and TRH_ligand_receptor_GP.\n",
      "Combining Tshb_ligand_receptor_target_gene_GP and TSHB_ligand_receptor_GP.\n",
      "Combining Tslp_ligand_receptor_target_gene_GP and TSLP_ligand_receptor_GP.\n",
      "Combining Ucn_ligand_receptor_target_gene_GP and UCN_ligand_receptor_GP.\n",
      "Combining Ucn2_ligand_receptor_target_gene_GP and UCN2_ligand_receptor_GP.\n",
      "Combining Ucn3_ligand_receptor_target_gene_GP and UCN3_ligand_receptor_GP.\n",
      "Combining Uts2_ligand_receptor_target_gene_GP and UTS2_ligand_receptor_GP.\n",
      "Combining Uts2b_ligand_receptor_target_gene_GP and UTS2B_ligand_receptor_GP.\n",
      "Combining Vcam1_ligand_receptor_target_gene_GP and VCAM1_ligand_receptor_GP.\n",
      "Combining Vcan_ligand_receptor_target_gene_GP and VCAN_ligand_receptor_GP.\n",
      "Combining Vegfa_ligand_receptor_target_gene_GP and VEGFA_ligand_receptor_GP.\n",
      "Combining Vegfb_ligand_receptor_target_gene_GP and VEGFB_ligand_receptor_GP.\n",
      "Combining Vegfc_ligand_receptor_target_gene_GP and VEGFC_ligand_receptor_GP.\n",
      "Combining Vegfd_ligand_receptor_target_gene_GP and VEGFD_ligand_receptor_GP.\n",
      "Combining Vip_ligand_receptor_target_gene_GP and VIP_ligand_receptor_GP.\n",
      "Combining Vtn_ligand_receptor_target_gene_GP and VTN_ligand_receptor_GP.\n",
      "Combining Vwf_ligand_receptor_target_gene_GP and VWF_ligand_receptor_GP.\n",
      "Combining Wnt1_ligand_receptor_target_gene_GP and WNT1_ligand_receptor_GP.\n",
      "Combining Wnt10b_ligand_receptor_target_gene_GP and WNT10B_ligand_receptor_GP.\n",
      "Combining Wnt11_ligand_receptor_target_gene_GP and WNT11_ligand_receptor_GP.\n",
      "Combining Wnt2_ligand_receptor_target_gene_GP and WNT2_ligand_receptor_GP.\n",
      "Combining Wnt2b_ligand_receptor_target_gene_GP and WNT2B_ligand_receptor_GP.\n",
      "Combining Wnt3_ligand_receptor_target_gene_GP and WNT3_ligand_receptor_GP.\n",
      "Combining Wnt3a_ligand_receptor_target_gene_GP and WNT3A_ligand_receptor_GP.\n",
      "Combining Wnt4_ligand_receptor_target_gene_GP and WNT4_ligand_receptor_GP.\n",
      "Combining Wnt5a_ligand_receptor_target_gene_GP and WNT5A_ligand_receptor_GP.\n",
      "Combining Wnt5b_ligand_receptor_target_gene_GP and WNT5B_ligand_receptor_GP.\n",
      "Combining Wnt6_ligand_receptor_target_gene_GP and WNT6_ligand_receptor_GP.\n",
      "Combining Wnt7a_ligand_receptor_target_gene_GP and WNT7A_ligand_receptor_GP.\n",
      "Combining Wnt7b_ligand_receptor_target_gene_GP and WNT7B_ligand_receptor_GP.\n",
      "Combining Wnt8a_ligand_receptor_target_gene_GP and WNT8A_ligand_receptor_GP.\n",
      "Combining Wnt9a_ligand_receptor_target_gene_GP and WNT9A_ligand_receptor_GP.\n",
      "Combining Wnt9b_ligand_receptor_target_gene_GP and WNT9B_ligand_receptor_GP.\n",
      "Combining Xcl1_ligand_receptor_target_gene_GP and XCL1_ligand_receptor_GP.\n",
      "Combining Xcl1_ligand_receptor_target_gene_GP and XCL2_ligand_receptor_GP.\n",
      "Combining Zp3_ligand_receptor_target_gene_GP and ZP3_ligand_receptor_GP.\n",
      "Combining L-Phenylalanine_metabolite_enzyme_sensor_GP and ACE2_ligand_receptor_GP.\n",
      "Combining L-Phenylalanine_metabolite_enzyme_sensor_GP and Ace2_ligand_receptor_target_gene_GP.\n",
      "Combining Norepinephrine_metabolite_enzyme_sensor_GP and Dbh_ligand_receptor_target_gene_GP.\n",
      "Combining Liothyronine_metabolite_enzyme_sensor_GP and Tpo_ligand_receptor_target_gene_GP.\n",
      "Combining Vitamin D3_metabolite_enzyme_sensor_GP and Calcidiol_metabolite_enzyme_sensor_GP.\n",
      "Combining all-trans-Retinoic acid_metabolite_enzyme_sensor_GP and 9-cis-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining all-trans-Retinoic acid_metabolite_enzyme_sensor_GP and 13-cis-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining 9-cis-Retinoic acid_metabolite_enzyme_sensor_GP and all-trans-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining 9-cis-Retinoic acid_metabolite_enzyme_sensor_GP and 13-cis-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining Calcidiol_metabolite_enzyme_sensor_GP and Vitamin D3_metabolite_enzyme_sensor_GP.\n",
      "Combining 13-cis-Retinoic acid_metabolite_enzyme_sensor_GP and all-trans-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining 13-cis-Retinoic acid_metabolite_enzyme_sensor_GP and 9-cis-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining ACE2_combined_GP and Ace2_combined_GP.\n",
      "Combining CCL18_combined_GP and CCL3_combined_GP.\n",
      "Combining CCL3_combined_GP and CCL18_combined_GP.\n",
      "Combining CCL4_combined_GP and CCL4L2_combined_GP.\n",
      "Combining CCL4L2_combined_GP and CCL4_combined_GP.\n",
      "Combining CSH1_combined_GP and GH1_combined_GP.\n",
      "Combining CSH1_combined_GP and GH2_combined_GP.\n",
      "Combining CXCL5_combined_GP and CXCL6_combined_GP.\n",
      "Combining CXCL6_combined_GP and CXCL5_combined_GP.\n",
      "Combining GH1_combined_GP and CSH1_combined_GP.\n",
      "Combining GH1_combined_GP and GH2_combined_GP.\n",
      "Combining GH2_combined_GP and CSH1_combined_GP.\n",
      "Combining GH2_combined_GP and GH1_combined_GP.\n",
      "Combining HLA-DQB1_combined_GP and HLA-DQB2_combined_GP.\n",
      "Combining HLA-DQB2_combined_GP and HLA-DQB1_combined_GP.\n",
      "Combining HLA-DRB1_combined_GP and HLA-DRB5_combined_GP.\n",
      "Combining HLA-DRB5_combined_GP and HLA-DRB1_combined_GP.\n",
      "Combining IGFL2_combined_GP and IGFL3_combined_GP.\n",
      "Combining IGFL3_combined_GP and IGFL2_combined_GP.\n",
      "Combining RLN1_combined_GP and RLN2_combined_GP.\n",
      "Combining RLN2_combined_GP and RLN1_combined_GP.\n",
      "Combining XCL1_combined_GP and XCL2_combined_GP.\n",
      "Combining XCL2_combined_GP and XCL1_combined_GP.\n",
      "Combining Ace2_combined_GP and ACE2_combined_GP.\n",
      "Combining all-trans-Retinoic acid_combined_GP and 9-cis-Retinoic acid_combined_GP.\n",
      "Combining 9-cis-Retinoic acid_combined_GP and all-trans-Retinoic acid_combined_GP.\n"
     ]
    }
   ],
   "source": [
    "# Retrieve OmniPath GPs (source: ligand genes; target: receptor genes)\n",
    "omnipath_gp_dict = extract_gp_dict_from_omnipath_lr_interactions(\n",
    "    species=species,\n",
    "    load_from_disk=True,\n",
    "    save_to_disk=False,\n",
    "    lr_network_file_path=omnipath_lr_network_file_path,\n",
    "    gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path,\n",
    "    plot_gp_gene_count_distributions=False)\n",
    "\n",
    "# Retrieve NicheNet GPs (source: ligand genes; target: receptor genes, target genes)\n",
    "nichenet_gp_dict = extract_gp_dict_from_nichenet_lrt_interactions(\n",
    "    species=species,\n",
    "    version=\"v2\",\n",
    "    load_from_disk=True,\n",
    "    save_to_disk=False,\n",
    "    lr_network_file_path=nichenet_lr_network_file_path,\n",
    "    ligand_target_matrix_file_path=nichenet_ligand_target_matrix_file_path,\n",
    "    gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path,\n",
    "    plot_gp_gene_count_distributions=False)\n",
    "\n",
    "# Retrieve MEBOCOST GPs (source: enzyme genes; target: sensor genes)\n",
    "mebocost_gp_dict = extract_gp_dict_from_mebocost_ms_interactions(\n",
    "    dir_path=mebocost_enzyme_sensor_interactions_folder_path,\n",
    "    species=species,\n",
    "    plot_gp_gene_count_distributions=False)\n",
    "\n",
    "# Filter and combine GPs\n",
    "gp_dicts = [omnipath_gp_dict, nichenet_gp_dict, mebocost_gp_dict]\n",
    "combined_gp_dict = filter_and_combine_gp_dict_gps_v2(\n",
    "    gp_dicts,\n",
    "    verbose=True)\n",
    "\n",
    "combined_gp_df = pd.DataFrame(combined_gp_dict.values(), index=combined_gp_dict.keys())\n",
    "\n",
    "# Filter genes to keep only genes in adata\n",
    "adata_genes = adata.var.index.tolist()\n",
    "combined_gp_df[\"sources\"] = combined_gp_df[\"sources\"].apply(lambda x: [gene for gene in x if gene in adata_genes])\n",
    "combined_gp_df[\"targets\"] = combined_gp_df[\"targets\"].apply(lambda x: [gene for gene in x if gene in adata_genes])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba9a6edd-94ce-4142-b6a3-6a4273afd83c",
   "metadata": {},
   "source": [
    "### 2.3 Get Simulation Ground Truth GPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5934d15c-2deb-40f1-ac8d-d4da5c244230",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1582691/3182646810.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n"
     ]
    }
   ],
   "source": [
    "# Get GP ground truth stats\n",
    "increment_mode = \"strong\"\n",
    "sim_gps_stats_file_path = f\"sim_gps_filtered_stats_{increment_mode}increments.csv\"\n",
    "sim_gps_stats = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/{sim_gps_stats_file_path}\")\n",
    "sim_gps_stats = sim_gps_stats.applymap(lambda x: \"\" if pd.isna(x) else x)\n",
    "sim_gps_stats[\"prior_gps\"] = sim_gps_stats[\"prior_gps\"].apply(\n",
    "    lambda x: [gp for gp in x.split(\",\")])\n",
    "sim_gps_stats = sim_gps_stats[sim_gps_stats[\"increment_param\"] != 1]\n",
    "sim_gps_stats = sim_gps_stats.groupby(\"regionA\")[\"prior_gps\"].sum().reset_index()\n",
    "sim_gps_stats[\"prior_gps\"] = sim_gps_stats[\"prior_gps\"].apply(lambda x: {item for item in x if item not in [None, '']})\n",
    "sim_gps_stats.rename(columns={'regionA': 'niche'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "562b0ed8-2a6c-4d1f-a7d1-2642683a8c91",
   "metadata": {},
   "source": [
    "### 2.4 NicheCompass Metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "de0c4a96-942b-4530-8358-8aaa3632b741",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 1...\n",
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, rna_recon_loss: nb\n",
      "NODE LABEL METHOD -> one-hop-norm\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "LOG VARIATIONAL -> True\n",
      "ONE HOP GCN NORM RNA NODE LABEL AGGREGATOR\n",
      "ENCODER -> n_input: 1105, n_cat_covariates_embed_input: 0, n_hidden: 1105, n_latent: 1452, n_addon_latent: 10, n_fc_layers: 1, n_layers: 1, conv_layer: gatv2conv, n_attention_heads: 4, dropout_rate: 0.0, \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COSINE SIM GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED TARGET RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "MASKED SOURCE RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing enriched GPs...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/torch_geometric/sampler/neighbor_sampler.py:55: UserWarning: The usage of the 'directed' argument in 'NeighborSampler' is deprecated. Use `subgraph_type='induced'` instead.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, rna_recon_loss: nb\n",
      "NODE LABEL METHOD -> one-hop-norm\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "LOG VARIATIONAL -> True\n",
      "ONE HOP GCN NORM RNA NODE LABEL AGGREGATOR\n",
      "ENCODER -> n_input: 1105, n_cat_covariates_embed_input: 0, n_hidden: 1105, n_latent: 1452, n_addon_latent: 10, n_fc_layers: 1, n_layers: 1, conv_layer: gatv2conv, n_attention_heads: 4, dropout_rate: 0.0, \n",
      "COSINE SIM GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED TARGET RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "MASKED SOURCE RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing enriched GPs...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/torch_geometric/sampler/neighbor_sampler.py:55: UserWarning: The usage of the 'directed' argument in 'NeighborSampler' is deprecated. Use `subgraph_type='induced'` instead.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 3...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, rna_recon_loss: nb\n",
      "NODE LABEL METHOD -> one-hop-norm\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "LOG VARIATIONAL -> True\n",
      "ONE HOP GCN NORM RNA NODE LABEL AGGREGATOR\n",
      "ENCODER -> n_input: 1105, n_cat_covariates_embed_input: 0, n_hidden: 1105, n_latent: 1452, n_addon_latent: 10, n_fc_layers: 1, n_layers: 1, conv_layer: gatv2conv, n_attention_heads: 4, dropout_rate: 0.0, \n",
      "COSINE SIM GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED TARGET RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "MASKED SOURCE RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing enriched GPs...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/torch_geometric/sampler/neighbor_sampler.py:55: UserWarning: The usage of the 'directed' argument in 'NeighborSampler' is deprecated. Use `subgraph_type='induced'` instead.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 4...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, rna_recon_loss: nb\n",
      "NODE LABEL METHOD -> one-hop-norm\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "LOG VARIATIONAL -> True\n",
      "ONE HOP GCN NORM RNA NODE LABEL AGGREGATOR\n",
      "ENCODER -> n_input: 1105, n_cat_covariates_embed_input: 0, n_hidden: 1105, n_latent: 1452, n_addon_latent: 10, n_fc_layers: 1, n_layers: 1, conv_layer: gatv2conv, n_attention_heads: 4, dropout_rate: 0.0, \n",
      "COSINE SIM GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED TARGET RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "MASKED SOURCE RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing enriched GPs...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/torch_geometric/sampler/neighbor_sampler.py:55: UserWarning: The usage of the 'directed' argument in 'NeighborSampler' is deprecated. Use `subgraph_type='induced'` instead.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 5...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, rna_recon_loss: nb\n",
      "NODE LABEL METHOD -> one-hop-norm\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "LOG VARIATIONAL -> True\n",
      "ONE HOP GCN NORM RNA NODE LABEL AGGREGATOR\n",
      "ENCODER -> n_input: 1105, n_cat_covariates_embed_input: 0, n_hidden: 1105, n_latent: 1452, n_addon_latent: 10, n_fc_layers: 1, n_layers: 1, conv_layer: gatv2conv, n_attention_heads: 4, dropout_rate: 0.0, \n",
      "COSINE SIM GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED TARGET RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "MASKED SOURCE RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing enriched GPs...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/torch_geometric/sampler/neighbor_sampler.py:55: UserWarning: The usage of the 'directed' argument in 'NeighborSampler' is deprecated. Use `subgraph_type='induced'` instead.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 6...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, rna_recon_loss: nb\n",
      "NODE LABEL METHOD -> one-hop-norm\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "LOG VARIATIONAL -> True\n",
      "ONE HOP GCN NORM RNA NODE LABEL AGGREGATOR\n",
      "ENCODER -> n_input: 1105, n_cat_covariates_embed_input: 0, n_hidden: 1105, n_latent: 1452, n_addon_latent: 10, n_fc_layers: 1, n_layers: 1, conv_layer: gatv2conv, n_attention_heads: 4, dropout_rate: 0.0, \n",
      "COSINE SIM GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED TARGET RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "MASKED SOURCE RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing enriched GPs...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/torch_geometric/sampler/neighbor_sampler.py:55: UserWarning: The usage of the 'directed' argument in 'NeighborSampler' is deprecated. Use `subgraph_type='induced'` instead.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 7...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, rna_recon_loss: nb\n",
      "NODE LABEL METHOD -> one-hop-norm\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "LOG VARIATIONAL -> True\n",
      "ONE HOP GCN NORM RNA NODE LABEL AGGREGATOR\n",
      "ENCODER -> n_input: 1105, n_cat_covariates_embed_input: 0, n_hidden: 1105, n_latent: 1452, n_addon_latent: 10, n_fc_layers: 1, n_layers: 1, conv_layer: gatv2conv, n_attention_heads: 4, dropout_rate: 0.0, \n",
      "COSINE SIM GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED TARGET RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "MASKED SOURCE RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing enriched GPs...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/torch_geometric/sampler/neighbor_sampler.py:55: UserWarning: The usage of the 'directed' argument in 'NeighborSampler' is deprecated. Use `subgraph_type='induced'` instead.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 8...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, rna_recon_loss: nb\n",
      "NODE LABEL METHOD -> one-hop-norm\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "LOG VARIATIONAL -> True\n",
      "ONE HOP GCN NORM RNA NODE LABEL AGGREGATOR\n",
      "ENCODER -> n_input: 1105, n_cat_covariates_embed_input: 0, n_hidden: 1105, n_latent: 1452, n_addon_latent: 10, n_fc_layers: 1, n_layers: 1, conv_layer: gatv2conv, n_attention_heads: 4, dropout_rate: 0.0, \n",
      "COSINE SIM GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED TARGET RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "MASKED SOURCE RNA DECODER -> n_prior_gp_input: 1452, n_addon_gp_input: 10, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing enriched GPs...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/torch_geometric/sampler/neighbor_sampler.py:55: UserWarning: The usage of the 'directed' argument in 'NeighborSampler' is deprecated. Use `subgraph_type='induced'` instead.\n"
     ]
    }
   ],
   "source": [
    "# Get differentially expressed genes\n",
    "f1_scores = []\n",
    "\n",
    "for run_number in range(1,9):\n",
    "    print(f\"Starting run {run_number}...\")\n",
    "    latent_key = f\"nichecompass_latent\"\n",
    "    \n",
    "    load_timestamp = \"26072024_145319_1\"\n",
    "    \n",
    "    model_label = \"gatv2conv_single_sample_method_benchmarking\"\n",
    "    model_folder_path = f\"{artifacts_folder_path}/sim1_1105genes_10000locs_strongincrements/models/{model_label}/{load_timestamp}/run{run_number}\"\n",
    "\n",
    "    model = NicheCompass.load(dir_path=model_folder_path,\n",
    "                              adata=None,\n",
    "                              adata_file_name=f\"sim1_1105genes_10000locs_strongincrements_{model_label}.h5ad\",\n",
    "                              gp_names_key=gp_names_key) \n",
    "    \n",
    "    sc.pp.neighbors(model.adata,\n",
    "                    use_rep=latent_key,\n",
    "                    key_added=latent_key)\n",
    "    sc.tl.umap(model.adata,\n",
    "               neighbors_key=latent_key)\n",
    "    \n",
    "    # Compute Leiden clustering of latent space until 8 niches are obtained (to match ground truth number)\n",
    "    print(f\"Computing Leiden clusters...\")\n",
    "    latent_leiden_resolution = 0.3\n",
    "    leiden_resolution_increments = 0.1\n",
    "    latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\"\n",
    "    counter = 0\n",
    "    while True:\n",
    "        sc.tl.leiden(adata=model.adata,\n",
    "                     resolution=latent_leiden_resolution,\n",
    "                     key_added=\"pred_niche_types\",\n",
    "                     neighbors_key=latent_key)\n",
    "        \n",
    "        niche_counts = adata.obs[\"pred_niche_types\"].value_counts()\n",
    "        valid_niches = niche_counts[niche_counts >= 100].index\n",
    "        n_niches = adata.obs[adata.obs[\"pred_niche_types\"].isin(valid_niches)][\"pred_niche_types\"].nunique()\n",
    "        print(f\"Current number of niches: {n_niches}\")\n",
    "        print(f\"Cluster counter: {counter}\")\n",
    "        if n_niches == 8:\n",
    "            break\n",
    "        elif n_niches < 7 and counter < 30:\n",
    "            print(\"Big increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments\n",
    "        elif n_niches < 8 and counter < 60:\n",
    "            print(\"Slight increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments/10\n",
    "        elif n_niches > 9 and counter < 30:\n",
    "            print(\"Big decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments\n",
    "        elif n_niches > 8 and counter < 60:\n",
    "            print(\"Slight decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments/10\n",
    "        elif counter > 60:\n",
    "            break\n",
    "        counter += 1\n",
    "        \n",
    "    cross_tab = pd.crosstab(model.adata.obs[\"pred_niche_types\"], model.adata.obs[\"niche_types\"])\n",
    "    majority_map = cross_tab.idxmax(axis=1)\n",
    "    model.adata.obs[\"mapped_pred_niche_types\"] = model.adata.obs[\"pred_niche_types\"].map(majority_map)\n",
    "    \n",
    "    print(f\"Computing enriched GPs...\")\n",
    "    selected_cats = None\n",
    "    comparison_cats = \"rest\"\n",
    "    log_bayes_factor_thresh = 2.3\n",
    "    enriched_gps = model.run_differential_gp_tests(\n",
    "        cat_key=\"mapped_pred_niche_types\",\n",
    "        selected_cats=selected_cats,\n",
    "        comparison_cats=comparison_cats,\n",
    "        log_bayes_factor_thresh=log_bayes_factor_thresh)\n",
    "\n",
    "    # Get enriched prior and de novo GPs\n",
    "    gp_summary_df = model.get_gp_summary() \n",
    "    enriched_gps = model.adata.uns[differential_gp_test_results_key]\n",
    "    prior_enriched_gps = enriched_gps[~enriched_gps[\"gene_program\"].str.contains(\"Add-on\")]  \n",
    "    enriched_prior_gps_per_niche = prior_enriched_gps.groupby(\n",
    "                \"category\")[[\"gene_program\", \"log_bayes_factor\"]].agg(list)\n",
    "\n",
    "    enriched_prior_gps_per_niche[\"niche\"] = enriched_prior_gps_per_niche.index.tolist()\n",
    "    niche_enriched_df = enriched_prior_gps_per_niche[[\"niche\", \"gene_program\"]]\n",
    "    niche_enriched_df[\"enriched_gps\"] = niche_enriched_df[\"gene_program\"].apply(lambda x: set(x))\n",
    "\n",
    "    # Compute F1 scores\n",
    "    merged_df = pd.merge(niche_enriched_df, sim_gps_stats, on='niche', suffixes=('_df1', '_df2'))\n",
    "    merged_df['f1_score'] = merged_df.apply(lambda row: compute_f1(row['enriched_gps'], row['prior_gps']), axis=1)\n",
    "    f1_scores.append(np.mean(merged_df[\"f1_score\"]))\n",
    "    \n",
    "f1_score_df = pd.DataFrame(f1_scores, columns=[\"f1_score\"])\n",
    "f1_score_df[\"run_number\"] = range(1,9)\n",
    "f1_score_df.to_csv(\"nichecompass_f1_scores.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "a4c64ab0-51c7-4cb8-94a4-194ad44583ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>f1_score</th>\n",
       "      <th>run_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.064958</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.056861</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.097866</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.100066</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.122538</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.122261</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.139038</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.135510</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   f1_score  run_number\n",
       "0  0.064958           1\n",
       "1  0.056861           2\n",
       "2  0.097866           3\n",
       "3  0.100066           4\n",
       "4  0.122538           5\n",
       "5  0.122261           6\n",
       "6  0.139038           7\n",
       "7  0.135510           8"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f1_score_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88ab0e06-c248-46f6-9ac8-52e556832462",
   "metadata": {},
   "source": [
    "### 2.4 Decoupler GSEA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c95c82a6-9e54-4dc3-96ef-769461153691",
   "metadata": {},
   "source": [
    "#### 2.4.1 Perform Gene Set Enrichment Analysis with Decoupler (Based on Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "29c70c7f-717d-4bed-9f57-ef4653f0785d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Format GPs for GSEA\n",
    "gp_df_gsea = combined_gp_df.copy()\n",
    "gp_df_gsea[\"geneset\"] = gp_df_gsea.index\n",
    "gp_df_gsea[\"genesymbol\"] = gp_df_gsea[\"targets\"]\n",
    "del(gp_df_gsea[\"targets\"])\n",
    "gp_df_gsea = gp_df_gsea[[\"geneset\", \"genesymbol\"]]\n",
    "gp_df_gsea.reset_index(drop=True, inplace=True)\n",
    "gp_df_gsea = gp_df_gsea.explode(\"genesymbol\").reset_index(drop=True)\n",
    "gp_df_gsea = gp_df_gsea.dropna(subset=['genesymbol'])\n",
    "gp_df_gsea = gp_df_gsea.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "809c1e9a-03f8-4af0-a038-ae8222db8f01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 1...\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing GSEA...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 2...\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing GSEA...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 3...\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing GSEA...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 4...\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing GSEA...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 5...\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing GSEA...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 6...\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing GSEA...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 7...\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 0\n",
      "Computing GSEA...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting run 8...\n",
      "Computing Leiden clusters...\n",
      "Current number of niches: 9\n",
      "Cluster counter: 0\n",
      "Slight decrease of clustering resolution...\n",
      "Current number of niches: 9\n",
      "Cluster counter: 1\n",
      "Slight decrease of clustering resolution...\n",
      "Current number of niches: 9\n",
      "Cluster counter: 2\n",
      "Slight decrease of clustering resolution...\n",
      "Current number of niches: 9\n",
      "Cluster counter: 3\n",
      "Slight decrease of clustering resolution...\n",
      "Current number of niches: 8\n",
      "Cluster counter: 4\n",
      "Computing GSEA...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/scanpy/get/get.py:69: FutureWarning: The previous implementation of stack is deprecated and will be removed in a future version of pandas. See the What's New notes for pandas 2.1.0 for details. Specify future_stack=True to adopt the new implementation and silence this warning.\n",
      "/tmp/ipykernel_1582691/1574043303.py:88: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n"
     ]
    }
   ],
   "source": [
    "# Get differentially expressed genes\n",
    "f1_scores = []\n",
    "\n",
    "for run_number in range(1,9):\n",
    "    print(f\"Starting run {run_number}...\")\n",
    "    latent_key = f\"banksy_latent_run{run_number}\"\n",
    "    sc.pp.neighbors(adata,\n",
    "                    use_rep=latent_key,\n",
    "                    key_added=latent_key)\n",
    "    sc.tl.umap(adata,\n",
    "               neighbors_key=latent_key)\n",
    "    \n",
    "    # Compute Leiden clustering of latent space until 8 niches are obtained (to match ground truth number)\n",
    "    print(f\"Computing Leiden clusters...\")\n",
    "    latent_leiden_resolution = 0.3\n",
    "    leiden_resolution_increments = 0.1\n",
    "    latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\"\n",
    "    counter = 0\n",
    "    while True:\n",
    "        sc.tl.leiden(adata=adata,\n",
    "                     resolution=latent_leiden_resolution,\n",
    "                     key_added=\"pred_niche_types\",\n",
    "                     neighbors_key=latent_key)\n",
    "        \n",
    "        niche_counts = adata.obs[\"pred_niche_types\"].value_counts()\n",
    "        valid_niches = niche_counts[niche_counts >= 100].index\n",
    "        n_niches = adata.obs[adata.obs[\"pred_niche_types\"].isin(valid_niches)][\"pred_niche_types\"].nunique()\n",
    "        print(f\"Current number of niches: {n_niches}\")\n",
    "        print(f\"Cluster counter: {counter}\")\n",
    "        if n_niches == 8:\n",
    "            break\n",
    "        elif n_niches < 7 and counter < 30:\n",
    "            print(\"Big increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments\n",
    "        elif n_niches < 8 and counter < 60:\n",
    "            print(\"Slight increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments/10\n",
    "        elif n_niches > 9 and counter < 30:\n",
    "            print(\"Big decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments\n",
    "        elif n_niches > 8 and counter < 60:\n",
    "            print(\"Slight decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments/10\n",
    "        elif counter > 60:\n",
    "            break\n",
    "        counter += 1\n",
    "        \n",
    "    cross_tab = pd.crosstab(adata.obs[\"pred_niche_types\"], adata.obs[\"niche_types\"])\n",
    "    majority_map = cross_tab.idxmax(axis=1)\n",
    "    adata.obs[\"mapped_pred_niche_types\"] = adata.obs[\"pred_niche_types\"].map(majority_map)\n",
    "    \n",
    "    print(f\"Computing GSEA...\")\n",
    "    sc.tl.rank_genes_groups(adata, \"mapped_pred_niche_types\", method=\"t-test\", key_added=\"t-test\")\n",
    "    \n",
    "    # Get enriched GPs per niche\n",
    "    enriched_dict = {}\n",
    "\n",
    "    for niche_type in adata.obs[\"mapped_pred_niche_types\"].unique().tolist():\n",
    "\n",
    "        # Extract differential gene expression scores\n",
    "        t_stats = (\n",
    "            # Get dataframe of DE results for condition vs. rest\n",
    "            sc.get.rank_genes_groups_df(adata, niche_type, key=\"t-test\")\n",
    "            .set_index(\"names\")\n",
    "            # Sort by absolute score\n",
    "            .sort_values(\"scores\", key=np.abs, ascending=False)\n",
    "            # Format for decoupler\n",
    "            [[\"scores\"]]\n",
    "            .rename_axis([niche_type], axis=1)\n",
    "        )\n",
    "\n",
    "        # Get gene set enrichment analysis results\n",
    "        scores, norm, pvals = decoupler.run_gsea(\n",
    "            t_stats.T,\n",
    "            gp_df_gsea,\n",
    "            min_n=0,\n",
    "            source=\"geneset\",\n",
    "            target=\"genesymbol\",\n",
    "        )\n",
    "\n",
    "        gsea_results = (\n",
    "            pd.concat({\"score\": scores.T, \"norm\": norm.T, \"pval\": pvals.T}, axis=1)\n",
    "            .droplevel(level=1, axis=1)\n",
    "            .sort_values(\"pval\")\n",
    "        )\n",
    "\n",
    "        # Keep upregulated pathways with p-value < 0.05\n",
    "        enriched_gps = gsea_results[gsea_results[\"score\"] > 0][gsea_results[\"pval\"] < 0.05].index.tolist()\n",
    "        enriched_dict[niche_type] = enriched_gps\n",
    "        \n",
    "    enriched_df = pd.DataFrame(list(enriched_dict.items()), columns=[\"niche\", \"enriched_gps\"])\n",
    "    niche_enriched_df = enriched_df.groupby(\"niche\")[\"enriched_gps\"].sum().reset_index()\n",
    "    niche_enriched_df[\"enriched_gps\"] = niche_enriched_df[\"enriched_gps\"].apply(lambda x: set(x))\n",
    "\n",
    "    # Compute F1 scores\n",
    "    merged_df = pd.merge(niche_enriched_df, sim_gps_stats, on='niche', suffixes=('_df1', '_df2'))\n",
    "    merged_df['f1_score'] = merged_df.apply(lambda row: compute_f1(row['enriched_gps'], row['prior_gps']), axis=1)\n",
    "    f1_scores.append(np.mean(merged_df[\"f1_score\"]))\n",
    "    \n",
    "f1_score_df = pd.DataFrame(f1_scores, columns=[\"f1_score\"])\n",
    "f1_score_df[\"run_number\"] = range(1,9)\n",
    "f1_score_df.to_csv(\"banksy_gsea_f1_scores.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "9df90498-e03b-427d-bcae-c634a2049dcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>f1_score</th>\n",
       "      <th>run_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.005083</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.018298</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.005266</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.017535</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.032993</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.015353</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.023901</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.001932</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   f1_score  run_number\n",
       "0  0.005083           1\n",
       "1  0.018298           2\n",
       "2  0.005266           3\n",
       "3  0.017535           4\n",
       "4  0.032993           5\n",
       "5  0.015353           6\n",
       "6  0.023901           7\n",
       "7  0.001932           8"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f1_score_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5454e62b-0482-4d80-90b9-773c642ac8cc",
   "metadata": {},
   "source": [
    "### 2.5 LIANA CCI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "452857d6-7014-40e3-9086-022a71f8ab89",
   "metadata": {},
   "outputs": [],
   "source": [
    "li.rs.show_resources()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ad57ace-3a6c-4aa5-8548-20dc5c3cb781",
   "metadata": {},
   "outputs": [],
   "source": [
    "li.mt.rank_aggregate.by_sample(\n",
    "    adata,\n",
    "    groupby=groupby,\n",
    "    resource_name='mouseconsensus',\n",
    "    sample_key=sample_key,\n",
    "    use_raw=False,\n",
    "    verbose=True,\n",
    "    n_perms=None,\n",
    "    return_all_lrs=True,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b5f1fe-04ac-4731-bbd8-ad50973a088e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93baf5e6-9ea1-44c4-b3d2-819ca87abd48",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns[\"liana_res\"].sort_values(\"magnitude_rank\").head(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
