{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1b1daabf-c8e3-43d4-aa32-9789abcc8619",
   "metadata": {},
   "source": [
    "# Autotalker Data Analysis Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c325d921-2077-42b3-90a8-6f6c7dd928c5",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>).\n",
    "- **Affiliation:** Helmholtz Munich, Institute of Computational Biology (ICB), Talavera-LÃ³pez Lab\n",
    "- **Date of Creation:** 22.01.2023\n",
    "- **Date of Last Modification:** 11.02.2023"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d464ebf-354f-4f61-8040-c3dcbcc893b9",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5631277-d8d2-4194-a376-280f4f149b7d",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e5d3297-0290-47c0-846f-705e74cdb3de",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dbd4925-7505-4d01-bf91-95b54bb8b533",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../../autotalker\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4aa3c6a-5a44-4cf1-84a3-5e514457cf3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import os\n",
    "import random\n",
    "import warnings\n",
    "from copy import deepcopy\n",
    "from datetime import datetime\n",
    "\n",
    "import anndata as ad\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import mlflow\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scipy.sparse as sp\n",
    "import seaborn as sns\n",
    "import squidpy as sq\n",
    "import torch\n",
    "from matplotlib import gridspec\n",
    "from matplotlib.pyplot import rc_context\n",
    "\n",
    "from autotalker.models import Autotalker\n",
    "from autotalker.utils import (add_gps_from_gp_dict_to_adata,\n",
    "                              extract_gp_dict_from_mebocost_es_interactions,\n",
    "                              extract_gp_dict_from_nichenet_ligand_target_mx,\n",
    "                              extract_gp_dict_from_omnipath_lr_interactions,\n",
    "                              filter_and_combine_gp_dict_gps,\n",
    "                              get_unique_genes_from_gp_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f91f345-c297-47cb-8650-c6d9c60a8b7e",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4d4e6d9-239f-47c5-a30b-579b16d4c27c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Dataset\n",
    "dataset = \"seqfish_mouse_organogenesis\"\n",
    "batch1 = \"embryo1_z2\"\n",
    "batch2 = \"embryo1_z5\"\n",
    "batch3 = \"embryo2_z2\"\n",
    "batch4 = \"embryo2_z5\"\n",
    "batch5 = \"embryo3_z2\"\n",
    "batch6 = \"embryo3_z5\"\n",
    "n_neighbors = 20 # 12\n",
    "\n",
    "## Model\n",
    "# AnnData Keys\n",
    "counts_key = \"log_normalized_counts\" # raw counts not available\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "adj_key = \"spatial_connectivities\"\n",
    "spatial_key = \"spatial\"\n",
    "gp_names_key = \"autotalker_gp_names\"\n",
    "active_gp_names_key = \"autotalker_active_gp_names\"\n",
    "gp_targets_mask_key = \"autotalker_gp_targets\"\n",
    "gp_sources_mask_key = \"autotalker_gp_sources\"\n",
    "latent_key = \"autotalker_latent\"\n",
    "condition_key = \"batch\"\n",
    "genes_idx_key = \"autotalker_genes_idx\"\n",
    "query_enriched_cell_type_key = \"autotalker_query_enriched_cell_types\"\n",
    "query_enriched_latent_cluster_key = \"autotalker_query_enriched_latent_clusters\"\n",
    "mapping_entity_key = \"mapping_entity\"\n",
    "\n",
    "# Architecture\n",
    "active_gp_thresh_ratio = 0.03\n",
    "gene_expr_recon_dist = \"nb\"\n",
    "n_cond_embed = 3\n",
    "log_variational = False # log normalized counts as input\n",
    "\n",
    "# Trainer\n",
    "n_epochs = 40\n",
    "n_epochs_all_gps = 20\n",
    "lr = 0.001\n",
    "query_cond_embed_lr = 0.01\n",
    "lambda_edge_recon = 0.01\n",
    "lambda_gene_expr_recon = 0.0033\n",
    "lambda_l1_addon = 0.01\n",
    "\n",
    "# Benchmarking\n",
    "spatial_knng_key = \"autotalker_spatial_knng\"\n",
    "latent_knng_key = \"autotalker_latent_knng\"\n",
    "\n",
    "# Analysis\n",
    "enriched_query_cell_type_prop_thresh = 5\n",
    "enriched_query_latent_cluster_prop_thresh = 2.5\n",
    "latent_leiden_resolution = 0.3\n",
    "latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\"\n",
    "denovo_cell_type_differential_gp_scores_key = \"autotalker_denovo_cell_type_differential_gp_scores\"\n",
    "cell_type_differential_gp_scores_key = \"autotalker_cell_type_differential_gp_scores\"\n",
    "query_enriched_cell_type_differential_gp_scores_key = \"autotalker_query_enriched_latent_cluster_differential_gp_scores\"\n",
    "latent_cluster_differential_gp_scores_key = \"autotalker_latent_cluster_differential_gp_scores\"\n",
    "query_enriched_latent_cluster_differential_gp_scores_key = \"autotalker_query_enriched_latent_cluster_differential_gp_scores\"\n",
    "n_top_up_gps = 3\n",
    "n_top_down_gps = 3\n",
    "n_top_genes_per_gp = 3\n",
    "\n",
    "## Others\n",
    "random_seed = 42\n",
    "load_timestamp = \"08022023_004101\" # \"10022023_064933\" # saved model to be loaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51c95bd5-f794-4f68-ada0-fc51efb9096b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_colors = {\"Epiblast\" : \"#635547\",\n",
    "                    \"Primitive Streak\" : \"#DABE99\",\n",
    "                    \"Caudal epiblast\" : \"#9e6762\",\n",
    "                    \"PGC\" : \"#FACB12\",\n",
    "                    \"Anterior Primitive Streak\" : \"#c19f70\",\n",
    "                    \"Notochord\" : \"#0F4A9C\",\n",
    "                    \"Def. endoderm\" : \"#F397C0\",\n",
    "                    \"Definitive endoderm\" : \"#F397C0\",\n",
    "                    \"Gut\" : \"#EF5A9D\",\n",
    "                    \"Gut tube\" : \"#EF5A9D\",\n",
    "                    \"Nascent mesoderm\" : \"#C594BF\",\n",
    "                    \"Mixed mesoderm\" : \"#DFCDE4\",\n",
    "                    \"Intermediate mesoderm\" : \"#139992\",\n",
    "                    \"Caudal Mesoderm\" : \"#3F84AA\",\n",
    "                    \"Paraxial mesoderm\" : \"#8DB5CE\",\n",
    "                    \"Somitic mesoderm\" : \"#005579\",\n",
    "                    \"Pharyngeal mesoderm\" : \"#C9EBFB\",\n",
    "                    \"Splanchnic mesoderm\" : \"#C9EBFB\",\n",
    "                    \"Cardiomyocytes\" : \"#B51D8D\",\n",
    "                    \"Allantois\" : \"#532C8A\",\n",
    "                    \"ExE mesoderm\" : \"#8870ad\",\n",
    "                    \"Lateral plate mesoderm\" : \"#8870ad\",\n",
    "                    \"Mesenchyme\" : \"#cc7818\",\n",
    "                    \"Mixed mesenchymal mesoderm\" : \"#cc7818\",\n",
    "                    \"Haematoendothelial progenitors\" : \"#FBBE92\",\n",
    "                    \"Endothelium\" : \"#ff891c\",\n",
    "                    \"Blood progenitors 1\" : \"#f9decf\",\n",
    "                    \"Blood progenitors 2\" : \"#c9a997\",\n",
    "                    \"Erythroid1\" : \"#C72228\",\n",
    "                    \"Erythroid2\" : \"#f79083\",\n",
    "                    \"Erythroid3\" : \"#EF4E22\",\n",
    "                    \"Erythroid\" : \"#f79083\",\n",
    "                    \"Blood progenitors\" : \"#f9decf\",\n",
    "                    \"NMP\" : \"#8EC792\",\n",
    "                    \"Rostral neurectoderm\" : \"#65A83E\",\n",
    "                    \"Caudal neurectoderm\" : \"#354E23\",\n",
    "                    \"Neural crest\" : \"#C3C388\",\n",
    "                    \"Forebrain/Midbrain/Hindbrain\" : \"#647a4f\",\n",
    "                    \"Spinal cord\" : \"#CDE088\",\n",
    "                    \"Surface ectoderm\" : \"#f7f79e\",\n",
    "                    \"Visceral endoderm\" : \"#F6BFCB\",\n",
    "                    \"ExE endoderm\" : \"#7F6874\",\n",
    "                    \"ExE ectoderm\" : \"#989898\",\n",
    "                    \"Parietal endoderm\" : \"#1A1A1A\",\n",
    "                    \"Low quality\" : \"#e6e6e6\",\n",
    "                    \"Cranial mesoderm\" : \"#77441B\",\n",
    "                    \"Anterior somitic tissues\" : \"#F90026\",\n",
    "                    \"Sclerotome\" : \"#A10037\",\n",
    "                    \"Dermomyotome\" : \"#DA5921\",\n",
    "                    \"Posterior somitic tissues\" : \"#E1C239\",\n",
    "                    \"Presomitic mesoderm\" : \"#9DD84A\",\n",
    "                    \"None\" : \"#D3D3D3\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "264d245c-fc5c-4a7a-bbb9-cd96b362e5e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "latent_cluster_colors = {\"0\" : \"#635547\",\n",
    "                         \"1\" : \"#DABE99\",\n",
    "                         \"2\" : \"#9e6762\",\n",
    "                         \"3\" : \"#FACB12\",\n",
    "                         \"4\" : \"#c19f70\",\n",
    "                         \"5\" : \"#0F4A9C\",\n",
    "                         \"6\" : \"#F397C0\",\n",
    "                         \"7\" : \"#F397C0\",\n",
    "                         \"8\" : \"#EF5A9D\",\n",
    "                         \"9\" : \"#EF5A9D\",\n",
    "                         \"10\" : \"#C594BF\",\n",
    "                         \"11\" : \"#DFCDE4\",\n",
    "                         \"12\" : \"#139992\",\n",
    "                         \"13\" : \"#3F84AA\",\n",
    "                         \"14\" : \"#8DB5CE\",\n",
    "                         \"15\" : \"#005579\",\n",
    "                         \"16\" : \"#C9EBFB\",\n",
    "                         \"17\" : \"#C9EBFB\",\n",
    "                         \"18\" : \"#B51D8D\",\n",
    "                         \"19\" : \"#532C8A\",\n",
    "                         \"20\" : \"#8870ad\",\n",
    "                         \"21\" : \"#8870ad\",\n",
    "                         \"22\" : \"#cc7818\",\n",
    "                         \"23\" : \"#cc7818\",\n",
    "                         \"24\" : \"#FBBE92\",\n",
    "                         \"25\" : \"#ff891c\",\n",
    "                         \"26\" : \"#f9decf\",\n",
    "                         \"27\" : \"#c9a997\",\n",
    "                         \"28\" : \"#C72228\",\n",
    "                         \"29\" : \"#f79083\",\n",
    "                         \"30\" : \"#EF4E22\",\n",
    "                         \"31\" : \"#f79083\",\n",
    "                         \"32\" : \"#f9decf\",\n",
    "                         \"33\" : \"#8EC792\",\n",
    "                         \"34\" : \"#65A83E\",\n",
    "                         \"35\" : \"#354E23\",\n",
    "                         \"36\" : \"#C3C388\",\n",
    "                         \"37\" : \"#647a4f\",\n",
    "                         \"38\" : \"#CDE088\",\n",
    "                         \"39\" : \"#f7f79e\",\n",
    "                         \"40\" : \"#F6BFCB\",\n",
    "                         \"41\" : \"#7F6874\",\n",
    "                         \"42\" : \"#989898\",\n",
    "                         \"43\" : \"#1A1A1A\",\n",
    "                         \"44\" : \"#e6e6e6\",\n",
    "                         \"45\" : \"#77441B\",\n",
    "                         \"46\" : \"#F90026\",\n",
    "                         \"47\" : \"#A10037\",\n",
    "                         \"48\" : \"#DA5921\",\n",
    "                         \"49\" : \"#E1C239\",\n",
    "                         \"50\" : \"#9DD84A\",\n",
    "                         \"None\" : \"#D3D3D3\"}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6c3b7fa-fa59-43dd-81d2-7e116650915f",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bf229ca-ab66-4e52-a31e-8655878d290d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(6, 6))\n",
    "sns.set_style(\"whitegrid\", {'axes.grid' : False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "924032d4-24a4-45cf-87e1-0eba197da0c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ignore future warnings and user warnings\n",
    "warnings.simplefilter(action=\"ignore\", category=FutureWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=UserWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=RuntimeWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab22fbaf-aeed-464b-8a20-b640b9e055fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get time of notebook execution for timestamping saved artifacts\n",
    "now = datetime.now()\n",
    "if load_timestamp is not None:\n",
    "    current_timestamp = load_timestamp\n",
    "else:\n",
    "    current_timestamp = now.strftime(\"%d%m%Y_%H%M%S\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b886225-6b6d-433c-b117-47b9a8f8091f",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths and Create Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36675337-808d-40a9-93fc-d6cfa9473f66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define paths\n",
    "figure_folder_path = f\"../figures/{dataset}/analysis/{current_timestamp}\"\n",
    "model_artifacts_folder_path = f\"../artifacts/{dataset}/analysis/{current_timestamp}\"\n",
    "gp_data_folder_path = \"../datasets/gp_data\" # gene program data\n",
    "srt_data_folder_path = \"../datasets/srt_data\" # spatially resolved transcriptomics data\n",
    "srt_data_gold_folder_path = f\"{srt_data_folder_path}/gold\"\n",
    "nichenet_ligand_target_mx_file_path = gp_data_folder_path + \"/nichenet_ligand_target_matrix.csv\"\n",
    "omnipath_lr_interactions_file_path = gp_data_folder_path + \"/omnipath_lr_interactions.csv\"\n",
    "\n",
    "# Create required directories\n",
    "os.makedirs(figure_folder_path, exist_ok=True)\n",
    "os.makedirs(model_artifacts_folder_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35bdd00c-4bf6-4968-acee-e4e99e214cb1",
   "metadata": {},
   "source": [
    "### 1.5 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81a18eb4-6c46-4bd9-952c-3cb35d4be394",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cat_gp_score_gene_expr_summary(analysis_label,\n",
    "                                       model,\n",
    "                                       cats,\n",
    "                                       gps,\n",
    "                                       cat_key,\n",
    "                                       adata=None,\n",
    "                                       feature_spaces=[\"latent\", \"physical_embryo1\", \"physical_embryo2\", \"physical_embryo3\"],\n",
    "                                       plot_types=[\"gene_categories\", \"individual_genes\"],\n",
    "                                       n_top_genes_per_gp=5,\n",
    "                                       save_figs=False):\n",
    "    \n",
    "    if adata is None:\n",
    "        adata = model.adata.copy()\n",
    "    \n",
    "    for gp in gps:\n",
    "        gp_gene_importances_df = model.compute_gp_gene_importances(selected_gp=gp)\n",
    "        \n",
    "        pos_sign_target_genes = gp_gene_importances_df.loc[\n",
    "            (gp_gene_importances_df[\"gene_weight_sign_corrected\"] > 0) &\n",
    "            (gp_gene_importances_df[\"gene_entity\"] == \"target\"), \"gene\"].tolist()\n",
    "        pos_sign_source_genes = gp_gene_importances_df.loc[\n",
    "            (gp_gene_importances_df[\"gene_weight_sign_corrected\"] > 0) &\n",
    "            (gp_gene_importances_df[\"gene_entity\"] == \"source\"), \"gene\"].tolist()\n",
    "        neg_sign_target_genes = gp_gene_importances_df.loc[\n",
    "            (gp_gene_importances_df[\"gene_weight_sign_corrected\"] < 0) &\n",
    "            (gp_gene_importances_df[\"gene_entity\"] == \"target\"), \"gene\"].tolist()\n",
    "        neg_sign_source_genes = gp_gene_importances_df.loc[\n",
    "            (gp_gene_importances_df[\"gene_weight_sign_corrected\"] < 0) &\n",
    "            (gp_gene_importances_df[\"gene_entity\"] == \"source\"), \"gene\"].tolist()\n",
    "        \n",
    "        pos_sign_target_gene_importances = gp_gene_importances_df.loc[\n",
    "            (gp_gene_importances_df[\"gene_weight_sign_corrected\"] > 0) &\n",
    "            (gp_gene_importances_df[\"gene_entity\"] == \"target\"), \"gene_importance\"].values.reshape(1, -1)\n",
    "        pos_sign_source_gene_importances = gp_gene_importances_df.loc[\n",
    "            (gp_gene_importances_df[\"gene_weight_sign_corrected\"] > 0) &\n",
    "            (gp_gene_importances_df[\"gene_entity\"] == \"source\"), \"gene_importance\"].values.reshape(1, -1)\n",
    "        neg_sign_target_gene_importances = gp_gene_importances_df.loc[\n",
    "            (gp_gene_importances_df[\"gene_weight_sign_corrected\"] < 0) &\n",
    "            (gp_gene_importances_df[\"gene_entity\"] == \"target\"), \"gene_importance\"].values.reshape(1, -1)\n",
    "        neg_sign_source_gene_importances = gp_gene_importances_df.loc[\n",
    "            (gp_gene_importances_df[\"gene_weight_sign_corrected\"] < 0) &\n",
    "            (gp_gene_importances_df[\"gene_entity\"] == \"source\"), \"gene_importance\"].values.reshape(1, -1)\n",
    "        \n",
    "        pos_sign_target_gene_expr = adata[:, pos_sign_target_genes].X.toarray()\n",
    "        pos_sign_source_gene_expr = adata[:, pos_sign_source_genes].X.toarray()\n",
    "        neg_sign_target_gene_expr = adata[:, neg_sign_target_genes].X.toarray()\n",
    "        neg_sign_source_gene_expr = adata[:, neg_sign_source_genes].X.toarray()\n",
    "        \n",
    "        adata.obs[f\"{gp}_pos_sign_target_gene_weighted_mean_gene_expr\"] = (\n",
    "            np.mean(pos_sign_target_gene_expr * pos_sign_target_gene_importances, axis=1))\n",
    "        adata.obs[f\"{gp}_pos_sign_source_gene_weighted_mean_gene_expr\"] = (\n",
    "            np.mean(pos_sign_source_gene_expr * pos_sign_source_gene_importances, axis=1))\n",
    "        adata.obs[f\"{gp}_neg_sign_target_gene_weighted_mean_gene_expr\"] = (\n",
    "            np.mean(neg_sign_target_gene_expr * neg_sign_target_gene_importances, axis=1))\n",
    "        adata.obs[f\"{gp}_neg_sign_source_gene_weighted_mean_gene_expr\"] = (\n",
    "            np.mean(neg_sign_source_gene_expr * neg_sign_source_gene_importances, axis=1))\n",
    "        \n",
    "        adata.uns[f\"{gp}_gene_category_importances\"] = np.array([pos_sign_target_gene_importances.sum(),\n",
    "                                                                 pos_sign_source_gene_importances.sum(),\n",
    "                                                                 neg_sign_target_gene_importances.sum(),\n",
    "                                                                 neg_sign_source_gene_importances.sum()])\n",
    "        \n",
    "        adata.uns[\"n_top_genes\"] = n_top_genes_per_gp\n",
    "        adata.uns[f\"{gp}_top_genes\"] = gp_gene_importances_df[\"gene\"][:n_top_genes_per_gp]\n",
    "        adata.uns[f\"{gp}_top_gene_importances\"] = gp_gene_importances_df[\"gene_importance\"][:n_top_genes_per_gp]\n",
    "        adata.uns[f\"{gp}_top_gene_signs\"] = np.where(gp_gene_importances_df[\"gene_weight_sign_corrected\"] > 0, \"+\", \"-\")\n",
    "        adata.uns[f\"{gp}_top_gene_entities\"] = gp_gene_importances_df[\"gene_entity\"]\n",
    "        \n",
    "    for feature_space in feature_spaces:\n",
    "        for plot_type in plot_types:\n",
    "            plot_cat_gp_score_gene_expr_summary(adata=adata,\n",
    "                                                cats=cats,\n",
    "                                                gps=gps,\n",
    "                                                plot_type=plot_type,\n",
    "                                                plot_category=cat_key,\n",
    "                                                feature_space=feature_space,\n",
    "                                                suptitle=f\"{analysis_label.replace('_', ' ').title()} Differential GPs: \"\n",
    "                                                         f\"GP Scores and {'Weighted Mean ' if plot_type == 'gene_categories' else ''}\"\n",
    "                                                         f\"Gene Expression of {plot_type.replace('_', ' ').title()} in {feature_space.replace('_', ' ').title()} Space\",\n",
    "                                                cat_title=f\"GP-enriched \\n {analysis_label.replace('_', ' ')}\",\n",
    "                                                save_fig=save_figs,\n",
    "                                                fig_name=f\"{analysis_label}_gp_scores_{'weighted_mean_' if plot_type == 'gene_categories' else ''}gene_expr_\"\n",
    "                                                         f\"{analysis_label}_{feature_space}_space\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fc62806-dd64-4bbd-ab68-2d2babfbd908",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_cat_gp_score_gene_expr_summary(adata,\n",
    "                                        cats,\n",
    "                                        gps,\n",
    "                                        plot_type,\n",
    "                                        plot_category,\n",
    "                                        feature_space,\n",
    "                                        suptitle,\n",
    "                                        cat_title,\n",
    "                                        save_fig,\n",
    "                                        fig_name):\n",
    "        \n",
    "    # Plot selected gene program latent scores\n",
    "    if plot_type == \"gene_categories\":\n",
    "        ncols = 6\n",
    "        fig_width = 36\n",
    "        wspace = 0.155\n",
    "    elif plot_type == \"individual_genes\":\n",
    "        ncols = 2 + adata.uns[\"n_top_genes\"]\n",
    "        fig_width = 12 + (6 * adata.uns[\"n_top_genes\"])\n",
    "        wspace = 0.3\n",
    "    fig, axs = plt.subplots(nrows=len(gps), ncols=ncols, figsize=(fig_width, 6*len(gps)))\n",
    "    if axs.ndim == 1:\n",
    "        axs = axs.reshape(1, -1)\n",
    "\n",
    "    title = fig.suptitle(t=suptitle,\n",
    "                         x=0.55,\n",
    "                         y=(1.1 if len(gps) == 1 else 0.94),\n",
    "                         fontsize=20)\n",
    "    for i, gp in enumerate(gps):\n",
    "        if feature_space == \"latent\":\n",
    "            sc.pl.umap(adata,\n",
    "                       color=plot_category,\n",
    "                       groups=cats[i],\n",
    "                       ax=axs[i, 0],\n",
    "                       title=cat_title,\n",
    "                       legend_loc=\"on data\",\n",
    "                       na_in_legend=False,\n",
    "                       show=False)\n",
    "            sc.pl.umap(adata,\n",
    "                       color=gps[i],\n",
    "                       color_map=\"RdBu\",\n",
    "                       ax=axs[i, 1],\n",
    "                       title=f\"{gp[:gp.index('_')]}\\n{gp[gp.index('_') + 1: gp.rindex('_')].replace('_', ' ')}\\n{gp[gps[i].rindex('_') + 1:]} score\",\n",
    "                       show=False)\n",
    "        elif \"physical\" in feature_space:\n",
    "            sc.pl.spatial(adata=adata[adata.obs[\"sample\"] == feature_space.split(\"_\")[1]],\n",
    "                          color=plot_category,\n",
    "                          groups=cats[i],\n",
    "                          ax=axs[i, 0],\n",
    "                          spot_size=0.03,\n",
    "                          title=cat_title,\n",
    "                          legend_loc=\"on data\",\n",
    "                          na_in_legend=False,\n",
    "                          show=False)\n",
    "            sc.pl.spatial(adata=adata[adata.obs[\"sample\"] == feature_space.split(\"_\")[1]],\n",
    "                          color=gps[i],\n",
    "                          color_map=\"RdBu\",\n",
    "                          spot_size=0.03,\n",
    "                          title=f\"{gps[i].split('_', 1)[0]}\\n{gps[i].split('_', 1)[1]}\",\n",
    "                          legend_loc=None,\n",
    "                          ax=axs[i, 1],\n",
    "                          show=False) \n",
    "        axs[i, 0].xaxis.label.set_visible(False)\n",
    "        axs[i, 0].yaxis.label.set_visible(False)\n",
    "        axs[i, 1].xaxis.label.set_visible(False)\n",
    "        axs[i, 1].yaxis.label.set_visible(False)\n",
    "        if plot_type == \"gene_categories\":\n",
    "            for j, gene_category in enumerate([\"pos_sign_target_gene\",\n",
    "                                               \"pos_sign_source_gene\",\n",
    "                                               \"neg_sign_target_gene\",\n",
    "                                               \"neg_sign_source_gene\"]):\n",
    "                if not adata.obs[f\"{gp}_{gene_category}_weighted_mean_gene_expr\"].isna().any():\n",
    "                    if feature_space == \"latent\":\n",
    "                        sc.pl.umap(adata,\n",
    "                                   color=f\"{gp}_{gene_category}_weighted_mean_gene_expr\",\n",
    "                                   color_map=(\"Blues\" if \"pos_sign\" in gene_category else \"Reds\"),\n",
    "                                   ax=axs[i, j+2],\n",
    "                                   legend_loc=\"on data\",\n",
    "                                   na_in_legend=False,\n",
    "                                   title=f\"Weighted mean gene expression \\n {gene_category.replace('_', ' ')} ({adata.uns[f'{gp}_gene_category_importances'][j]:.2f})\",\n",
    "                                   show=False)\n",
    "                    elif \"physical\" in feature_space:\n",
    "                        sc.pl.spatial(adata=adata[adata.obs[\"sample\"] == feature_space.split(\"_\")[1]],\n",
    "                                      color=f\"{gp}_{gene_category}_weighted_mean_gene_expr\",\n",
    "                                      color_map=(\"Blues\" if \"pos_sign\" in gene_category else \"Reds\"),\n",
    "                                      ax=axs[i, 2+j],\n",
    "                                      legend_loc=\"on data\",\n",
    "                                      na_in_legend=False,\n",
    "                                      groups=cats[i],\n",
    "                                      spot_size=0.03,\n",
    "                                      title=f\"Weighted mean gene expression \\n {gene_category.replace('_', ' ')} ({adata.uns[f'{gp}_gene_category_importances'][j]:.2f})\",\n",
    "                                      show=False)                        \n",
    "                    axs[i, j+2].xaxis.label.set_visible(False)\n",
    "                    axs[i, j+2].yaxis.label.set_visible(False)\n",
    "                else:\n",
    "                    axs[i, j+2].set_visible(False)\n",
    "        elif plot_type == \"individual_genes\":\n",
    "            for j in range(len(adata.uns[f\"{gp}_top_genes\"])):\n",
    "                if feature_space == \"latent\":\n",
    "                    sc.pl.umap(adata,\n",
    "                               color=adata.uns[f\"{gp}_top_genes\"][j],\n",
    "                               color_map=(\"Blues\" if adata.uns[f\"{gp}_top_gene_signs\"][j] == \"+\" else \"Reds\"),\n",
    "                               ax=axs[i, 2+j],\n",
    "                               legend_loc=\"on data\",\n",
    "                               na_in_legend=False,\n",
    "                               title=f\"{adata.uns[f'{gp}_top_genes'][j]}: {adata.uns[f'{gp}_top_gene_importances'][j]:.2f} ({adata.uns[f'{gp}_top_gene_entities'][j][0]}; {adata.uns[f'{gp}_top_gene_signs'][j]})\",\n",
    "                               show=False)\n",
    "                elif \"physical\" in feature_space:\n",
    "                    sc.pl.spatial(adata=adata[adata.obs[\"sample\"] == feature_space.split(\"_\")[1]],\n",
    "                                  color=adata.uns[f\"{gp}_top_genes\"][j],\n",
    "                                  color_map=(\"Blues\" if adata.uns[f\"{gp}_top_gene_signs\"][j] == \"+\" else \"Reds\"),\n",
    "                                  legend_loc=\"on data\",\n",
    "                                  na_in_legend=False,\n",
    "                                  ax=axs[i, 2+j],\n",
    "                                  # groups=cats[i],\n",
    "                                  spot_size=0.03,\n",
    "                                  title=f\"{adata.uns[f'{gp}_top_genes'][j]}: {adata.uns[f'{gp}_top_gene_importances'][j]:.2f} ({adata.uns[f'{gp}_top_gene_entities'][j][0]}; {adata.uns[f'{gp}_top_gene_signs'][j]})\",\n",
    "                                  show=False)\n",
    "                axs[i, 2+j].xaxis.label.set_visible(False)\n",
    "                axs[i, 2+j].yaxis.label.set_visible(False)\n",
    "            for k in range(len(adata.uns[f\"{gp}_top_genes\"]), ncols - 2):\n",
    "                axs[i, 2+k].set_visible(False)            \n",
    "\n",
    "    # Save and display plot\n",
    "    plt.subplots_adjust(wspace=wspace, hspace=0.275)\n",
    "    if save_fig:\n",
    "        fig.savefig(f\"{figure_folder_path}/{fig_name}.png\",\n",
    "                    bbox_extra_artists=(title,),\n",
    "                    bbox_inches=\"tight\")\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cea1a5c-d075-4e51-90e5-90684bee88a6",
   "metadata": {},
   "source": [
    "## 2. Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d6ede1d-2cdf-4c81-9f56-e4745ae59e6c",
   "metadata": {},
   "source": [
    "### 2.1 Load Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db67b57a-f66a-4d06-be6d-79bbb113f6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "if load_timestamp is not None:\n",
    "    model_artifacts_load_folder_path = f\"../artifacts/{dataset}/batch_integration/{load_timestamp}\"\n",
    "else:\n",
    "    model_artifacts_load_folder_path = model_artifacts_folder_path\n",
    "    \n",
    "# Load trained model    \n",
    "model = Autotalker.load(dir_path=f\"{model_artifacts_load_folder_path}/reference_query\",\n",
    "                        adata=None,\n",
    "                        adata_file_name=f\"{dataset}.h5ad\",\n",
    "                        gp_names_key=\"autotalker_gp_names\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d48d33e1-36d5-4246-a947-4662055cea58",
   "metadata": {},
   "source": [
    "### 2.2 Retrieve GP Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fc40859-6a4a-4bfa-8362-d28d4c1c2057",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check number of active gene programs\n",
    "active_gps = model.get_active_gps()\n",
    "print(f\"Number of total gene programs: {len(model.adata.uns[gp_names_key])}.\")\n",
    "print(f\"Number of active gene programs: {len(active_gps)}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8053fc3f-2e47-4530-8e77-9da69bbb24ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp_summary_df = model.get_gp_summary()\n",
    "gp_summary_df[gp_summary_df[\"gp_active\"]].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26b2eb48-9cd0-4fea-8164-21d57804627b",
   "metadata": {},
   "source": [
    "### 2.3 Cell type-based Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21218f85-91c5-4766-9980-e44d8ec49f12",
   "metadata": {},
   "source": [
    "#### 2.3.1 Visualize Cell Types in Physical and Latent Space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a9da653-33f5-4a55-90d4-77ef130dde4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create plot of cell type annotations in physical and latent space\n",
    "fig = plt.figure(figsize=(12, 14))\n",
    "title = fig.suptitle(t=\"Autotalker: Cell Types in Physical and Latent Space\",\n",
    "                     y=0.96,\n",
    "                     x=0.55,\n",
    "                     fontsize=20)\n",
    "spec = gridspec.GridSpec(ncols=1,\n",
    "                         nrows=2,\n",
    "                         width_ratios=[1],\n",
    "                         height_ratios=[1, 5])\n",
    "ax1 = fig.add_subplot(2, 3, 1)\n",
    "ax2 = fig.add_subplot(2, 3, 2)\n",
    "ax3 = fig.add_subplot(2, 3, 3)\n",
    "ax4 = fig.add_subplot(spec[1])\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo1\"],\n",
    "              color=[cell_type_key],\n",
    "              palette=cell_type_colors,\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 1\",\n",
    "              legend_loc=None,\n",
    "              ax=ax1,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "              color=[cell_type_key],\n",
    "              palette=cell_type_colors,\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 2\",\n",
    "              legend_loc=None,\n",
    "              ax=ax2,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo3\"],\n",
    "              color=[cell_type_key],\n",
    "              palette=cell_type_colors,\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 3\",\n",
    "              legend_loc=None,\n",
    "              ax=ax3,\n",
    "              show=False)\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[cell_type_key],\n",
    "           palette=cell_type_colors,\n",
    "           size=240000/len(model.adata),\n",
    "           title=\"Autotalker Latent Space\",\n",
    "           ax=ax4,\n",
    "           show=False)\n",
    "\n",
    "# Create and position shared legend\n",
    "handles, labels = ax4.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles, labels, bbox_to_anchor=(1.25, 0.8))\n",
    "ax4.get_legend().remove()\n",
    "\n",
    "# Adjust, save and display plot\n",
    "plt.subplots_adjust(wspace=0., hspace=0.85)\n",
    "fig.savefig(f\"{figure_folder_path}/cell_types_physical_latent_space.png\",\n",
    "            bbox_extra_artists=(lgd, title),\n",
    "            bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eb2329c-23bf-4589-823e-eeb77cbc1bc8",
   "metadata": {},
   "source": [
    "#### 2.3.2 Visualize Mapping Entities in Physical and Latent Space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f485d42e-9e5a-47d3-9648-571248156cc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.uns[\"mapping_entity_colors\"] = ['#1f77b4', '#ff7f0e']\n",
    "\n",
    "# Create plot of mapping entity annotations in physical and latent space\n",
    "fig = plt.figure(figsize=(12, 14))\n",
    "title = fig.suptitle(t=\"Autotalker: Mapping Entities in Physical and Latent Space\",\n",
    "                     y=0.96,\n",
    "                     x=0.55,\n",
    "                     fontsize=20)\n",
    "spec = gridspec.GridSpec(ncols=1,\n",
    "                         nrows=2,\n",
    "                         width_ratios=[1],\n",
    "                         height_ratios=[1, 5])\n",
    "ax1 = fig.add_subplot(2, 3, 1)\n",
    "ax2 = fig.add_subplot(2, 3, 2)\n",
    "ax3 = fig.add_subplot(2, 3, 3)\n",
    "ax4 = fig.add_subplot(spec[1])\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo1\"],\n",
    "              color=[mapping_entity_key],\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 1\",\n",
    "              legend_loc=None,\n",
    "              ax=ax1,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "              color=[mapping_entity_key],\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 2\",\n",
    "              legend_loc=None,\n",
    "              ax=ax2,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo3\"],\n",
    "              color=[mapping_entity_key],\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 3\",\n",
    "              legend_loc=None,\n",
    "              ax=ax3,\n",
    "              show=False)\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[mapping_entity_key],\n",
    "           size=240000/len(model.adata),\n",
    "           title=\"Autotalker Latent Space\",\n",
    "           ax=ax4,\n",
    "           show=False)\n",
    "\n",
    "# Create and position shared legend\n",
    "handles, labels = ax4.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles, labels, bbox_to_anchor=(1.08, 0.645))\n",
    "ax4.get_legend().remove()\n",
    "\n",
    "# Adjust, save and display plot\n",
    "plt.subplots_adjust(wspace=0., hspace=0.85)\n",
    "fig.savefig(f\"{figure_folder_path}/mapping_entities_physical_latent_space.png\",\n",
    "            bbox_extra_artists=(lgd, title),\n",
    "            bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f46ff90c-cacf-4dcf-b29a-6a024ab03149",
   "metadata": {},
   "source": [
    "#### 2.3.3 Visualize Query-enriched Cell Types in Physical and Latent Space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e63180c5-7a0a-4a04-b1af-c196151e9fba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get query-enriched cell types\n",
    "adata_reference = model.adata[model.adata.obs[mapping_entity_key] == \"reference\"]\n",
    "adata_query = model.adata[model.adata.obs[mapping_entity_key] == \"query\"]\n",
    "cell_type_reference_proportions = adata_reference.obs[cell_type_key].value_counts().sort_index() / len(adata_reference)\n",
    "cell_type_query_proportions = adata_query.obs[cell_type_key].value_counts().sort_index() / len(adata_query)\n",
    "relative_cell_type_query_proportions = cell_type_query_proportions / cell_type_reference_proportions\n",
    "relative_cell_type_query_proportions.sort_values(ascending=False, inplace=True)\n",
    "display(relative_cell_type_query_proportions)\n",
    "\n",
    "query_enriched_cell_types = relative_cell_type_query_proportions[relative_cell_type_query_proportions > enriched_query_cell_type_prop_thresh].index.to_list()\n",
    "model.adata.obs[query_enriched_cell_type_key] = \"Cell types not enriched in query\"\n",
    "for cell_type in query_enriched_cell_types:\n",
    "    model.adata.obs.loc[model.adata.obs[cell_type_key] == cell_type, query_enriched_cell_type_key] = cell_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe2be87-a38e-4a3d-9dfb-921fb9243a00",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create plot of query-enriched cell-type annotations in physical and latent space\n",
    "fig = plt.figure(figsize=(12, 14))\n",
    "title = fig.suptitle(t=\"Autotalker: Query-enriched Cell Types in Physical and Latent Space\",\n",
    "                     y=0.96,\n",
    "                     x=0.55,\n",
    "                     fontsize=20)\n",
    "spec = gridspec.GridSpec(ncols=1,\n",
    "                         nrows=2,\n",
    "                         width_ratios=[1],\n",
    "                         height_ratios=[1, 5])\n",
    "ax1 = fig.add_subplot(2, 3, 1)\n",
    "ax2 = fig.add_subplot(2, 3, 2)\n",
    "ax3 = fig.add_subplot(2, 3, 3)\n",
    "ax4 = fig.add_subplot(spec[1])\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo1\"],\n",
    "              color=[query_enriched_cell_type_key],\n",
    "              palette=\"coolwarm\",\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 1\",\n",
    "              legend_loc=None,\n",
    "              ax=ax1,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "              color=[query_enriched_cell_type_key],\n",
    "              palette=\"coolwarm\",\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 2\",\n",
    "              legend_loc=None,\n",
    "              ax=ax2,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo3\"],\n",
    "              color=[query_enriched_cell_type_key],\n",
    "              palette=\"coolwarm\",\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 3\",\n",
    "              legend_loc=None,\n",
    "              ax=ax3,\n",
    "              show=False)\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[query_enriched_cell_type_key],\n",
    "           palette=\"coolwarm\",\n",
    "           size=240000/len(model.adata),\n",
    "           title=\"Autotalker Latent Space\",\n",
    "           ax=ax4,\n",
    "           show=False)\n",
    "\n",
    "# Create and position shared legend\n",
    "handles, labels = ax4.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles, labels, bbox_to_anchor=(1.25, 0.6525))\n",
    "ax4.get_legend().remove()\n",
    "\n",
    "# Adjust, save and display plot\n",
    "plt.subplots_adjust(wspace=0., hspace=0.85)\n",
    "fig.savefig(f\"{figure_folder_path}/query_enriched_cell_types_physical_latent_space.png\",\n",
    "            bbox_extra_artists=(lgd, title),\n",
    "            bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcf6e96f-5d9d-495f-8b5a-1fdd7af7a66e",
   "metadata": {},
   "source": [
    "#### 2.3.4 Analyze GP Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d58786b5-d11b-4a69-85d1-69d0fe6f3f7d",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### 2.3.4.1 Explore All GPs Across All Cell Types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac0eb5fc-85e0-4406-bb71-722bafceeaad",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_label = \"cell_type\"\n",
    "\n",
    "# Compute gene program enrichments and retrieve top up- and downregulated gene programs\n",
    "top_unique_gps = model.compute_differential_gp_scores(cat_key=cell_type_key,\n",
    "                                                      adata=model.adata,\n",
    "                                                      selected_gps=None,\n",
    "                                                      selected_cats=None,\n",
    "                                                      gp_scores_weight_normalization=False,\n",
    "                                                      comparison_cats=\"rest\",\n",
    "                                                      n_sample=10000,\n",
    "                                                      key_added=cell_type_differential_gp_scores_key,\n",
    "                                                      n_top_up_gps_retrieved=n_top_up_gps,\n",
    "                                                      n_top_down_gps_retrieved=n_top_down_gps,\n",
    "                                                      seed=random_seed)\n",
    "\n",
    "# Display top upregulated gene programs\n",
    "top_up_gp_df = model.adata.uns[cell_type_differential_gp_scores_key][:n_top_up_gps]\n",
    "display(top_up_gp_df)\n",
    "\n",
    "# Display top downregulated gene programs\n",
    "top_down_gp_df = model.adata.uns[cell_type_differential_gp_scores_key][-n_top_down_gps:][::-1]\n",
    "display(top_down_gp_df)\n",
    "\n",
    "fig = sc.pl.dotplot(model.adata,\n",
    "                    top_unique_gps,\n",
    "                    groupby=cell_type_key,\n",
    "                    dendrogram=True, \n",
    "                    title=f\"{analysis_label.replace('_', ' ').title()} Differential GP Scores\",\n",
    "                    swap_axes=True,\n",
    "                    return_fig=True)\n",
    "# Save and display plot\n",
    "fig.savefig(f\"{figure_folder_path}/{analysis_label}_differential_gp_scores.png\")\n",
    "plt.show()\n",
    "\n",
    "# Inspect top up- and downregulated gene programs\n",
    "display(gp_summary_df[gp_summary_df[\"gp_name\"].isin(top_unique_gps)])\n",
    "\n",
    "top_cats = top_up_gp_df[\"category\"].append(top_down_gp_df[\"category\"]).to_list()\n",
    "top_gps = top_up_gp_df[\"gene_program\"].append(top_down_gp_df[\"gene_program\"]).to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31bc3b5f-c1cd-4fde-a4ff-394ae92c3803",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cat_gp_score_gene_expr_summary(analysis_label=analysis_label,\n",
    "                                   model=model,\n",
    "                                   cats=top_cats,\n",
    "                                   gps=top_gps,\n",
    "                                   cat_key=cell_type_key,\n",
    "                                   adata=None,\n",
    "                                   feature_spaces=[\"latent\"], # [\"physical_embryo1\", \"physical_embryo2\", \"physical_embryo3\"],\n",
    "                                   plot_types=[\"gene_categories\", \"individual_genes\"],\n",
    "                                   n_top_genes_per_gp=5,\n",
    "                                   save_figs=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecce1ce8-adbc-4515-88b8-03c92eabb49d",
   "metadata": {},
   "source": [
    "##### 2.3.4.2 Explore All GPs Across Query-enriched Cell Types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d4a6bb4-0143-424c-905e-a4d0f8e4c1b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_label = \"query_enriched_cell_type\"\n",
    "\n",
    "# Compute gene program enrichments and retrieve top up- and downregulated gene programs\n",
    "top_unique_gps = model.compute_differential_gp_scores(cat_key=cell_type_key,\n",
    "                                                      adata=model.adata,\n",
    "                                                      selected_gps=None,\n",
    "                                                      selected_cats=query_enriched_cell_types,\n",
    "                                                      gp_scores_weight_normalization=False,\n",
    "                                                      comparison_cats=\"rest\",\n",
    "                                                      n_sample=10000,\n",
    "                                                      key_added=cell_type_differential_gp_scores_key,\n",
    "                                                      n_top_up_gps_retrieved=n_top_up_gps,\n",
    "                                                      n_top_down_gps_retrieved=n_top_down_gps,\n",
    "                                                      seed=random_seed)\n",
    "\n",
    "# Display top upregulated gene programs\n",
    "top_up_gp_df = model.adata.uns[cell_type_differential_gp_scores_key][:n_top_up_gps]\n",
    "display(top_up_gp_df)\n",
    "\n",
    "# Display top downregulated gene programs\n",
    "top_down_gp_df = model.adata.uns[cell_type_differential_gp_scores_key][-n_top_down_gps:][::-1]\n",
    "display(top_down_gp_df)\n",
    "\n",
    "fig = sc.pl.dotplot(model.adata,\n",
    "                    top_unique_gps,\n",
    "                    groupby=cell_type_key,\n",
    "                    dendrogram=True, \n",
    "                    title=f\"{analysis_label.replace('_', ' ').title()} Differential GP Scores\",\n",
    "                    swap_axes=True,\n",
    "                    return_fig=True)\n",
    "# Save and display plot\n",
    "fig.savefig(f\"{figure_folder_path}/{analysis_label}_differential_gp_scores.png\")\n",
    "plt.show()\n",
    "\n",
    "# Inspect top up- and downregulated gene programs\n",
    "display(gp_summary_df[gp_summary_df[\"gp_name\"].isin(top_unique_gps)])\n",
    "\n",
    "top_cats = top_up_gp_df[\"category\"].append(top_down_gp_df[\"category\"]).to_list()\n",
    "top_gps = top_up_gp_df[\"gene_program\"].append(top_down_gp_df[\"gene_program\"]).to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "408919a0-c7aa-4727-8a97-e6f71b28daf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cat_gp_score_gene_expr_summary(analysis_label=analysis_label,\n",
    "                                   model=model,\n",
    "                                   cats=top_cats,\n",
    "                                   gps=top_gps,\n",
    "                                   cat_key=cell_type_key,\n",
    "                                   adata=None,\n",
    "                                   feature_spaces=[\"latent\"], # [\"physical_embryo1\", \"physical_embryo2\", \"physical_embryo3\"],\n",
    "                                   plot_types=[\"gene_categories\", \"individual_genes\"],\n",
    "                                   n_top_genes_per_gp=5,\n",
    "                                   save_figs=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32a358cf-106e-40ca-a193-c872961cca79",
   "metadata": {},
   "source": [
    "##### 2.3.4.3 Explore Specific GPs Across All Cell Types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a476f5e6-e7a6-4e16-a9fa-134439552d7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_label = \"cell_type_bmp2_gp\"\n",
    "\n",
    "# Compute gene program enrichments and retrieve top up- and downregulated gene programs\n",
    "top_unique_gps = model.compute_differential_gp_scores(cat_key=cell_type_key,\n",
    "                                                      adata=model.adata,\n",
    "                                                      selected_gps=[\"BMP2_ligand_targetgenes_GP\"],\n",
    "                                                      selected_cats=None,\n",
    "                                                      gp_scores_weight_normalization=False,\n",
    "                                                      comparison_cats=\"rest\",\n",
    "                                                      n_sample=10000,\n",
    "                                                      key_added=cell_type_differential_gp_scores_key,\n",
    "                                                      n_top_up_gps_retrieved=n_top_up_gps,\n",
    "                                                      n_top_down_gps_retrieved=n_top_down_gps,\n",
    "                                                      seed=random_seed)\n",
    "\n",
    "# Display top upregulated gene programs\n",
    "top_up_gp_df = model.adata.uns[cell_type_differential_gp_scores_key][:n_top_up_gps]\n",
    "display(top_up_gp_df)\n",
    "\n",
    "# Display top downregulated gene programs\n",
    "top_down_gp_df = model.adata.uns[cell_type_differential_gp_scores_key][-n_top_down_gps:][::-1]\n",
    "display(top_down_gp_df)\n",
    "\n",
    "fig = sc.pl.dotplot(model.adata,\n",
    "                    top_unique_gps,\n",
    "                    groupby=cell_type_key,\n",
    "                    dendrogram=True, \n",
    "                    title=f\"{analysis_label.replace('_', ' ').title()} Differential GP Scores\",\n",
    "                    swap_axes=True,\n",
    "                    return_fig=True)\n",
    "# Save and display plot\n",
    "fig.savefig(f\"{figure_folder_path}/{analysis_label}_differential_gp_scores.png\")\n",
    "plt.show()\n",
    "\n",
    "# Inspect top up- and downregulated gene programs\n",
    "display(gp_summary_df[gp_summary_df[\"gp_name\"].isin(top_unique_gps)])\n",
    "\n",
    "top_cats = top_up_gp_df[\"category\"].append(top_down_gp_df[\"category\"]).to_list()\n",
    "top_gps = top_up_gp_df[\"gene_program\"].append(top_down_gp_df[\"gene_program\"]).to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b2e624a-5b4a-4162-9f0e-dd24d8aabe10",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cat_gp_score_gene_expr_summary(analysis_label=analysis_label,\n",
    "                                   model=model,\n",
    "                                   cats=top_cats,\n",
    "                                   gps=top_gps,\n",
    "                                   cat_key=cell_type_key,\n",
    "                                   adata=None,\n",
    "                                   feature_spaces=[\"latent\"], # [\"physical_embryo1\", \"physical_embryo2\", \"physical_embryo3\"],\n",
    "                                   plot_types=[\"gene_categories\", \"individual_genes\"],\n",
    "                                   n_top_genes_per_gp=5,\n",
    "                                   save_figs=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b838fc35-9ffb-41b4-8afc-45fcfbac858e",
   "metadata": {},
   "source": [
    "### 2.4 Latent Clustering-based Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4707bd34-c7ae-49c0-9cbf-aaf592f732a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute latent Leiden clustering\n",
    "sc.tl.leiden(adata=model.adata,\n",
    "             resolution=latent_leiden_resolution,\n",
    "             random_state=random_seed,\n",
    "             key_added=f\"latent_leiden_{str(latent_leiden_resolution)}\",\n",
    "             adjacency=model.adata.obsp[f\"{latent_key}_knng_connectivities\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9441d5e8-6320-49ba-8e36-8e714da02305",
   "metadata": {},
   "source": [
    "#### 2.4.1 Visualize Latent Leiden Clusters in Latent and Physical Space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe0702f5-be2f-45cc-b73c-2adf10ca0fb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create plot of latent leiden cluster annotations in physical and latent space\n",
    "fig = plt.figure(figsize=(12, 14))\n",
    "title = fig.suptitle(t=\"Autotalker: Latent Leiden Clusters in Physical and Latent Space\",\n",
    "                     y=0.96,\n",
    "                     x=0.55,\n",
    "                     fontsize=20)\n",
    "spec = gridspec.GridSpec(ncols=1,\n",
    "                         nrows=2,\n",
    "                         width_ratios=[1],\n",
    "                         height_ratios=[5, 1])\n",
    "ax1 = fig.add_subplot(spec[0])\n",
    "ax2 = fig.add_subplot(2, 3, 4)\n",
    "ax3 = fig.add_subplot(2, 3, 5)\n",
    "ax4 = fig.add_subplot(2, 3, 6)\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[latent_cluster_key],\n",
    "           palette=latent_cluster_colors,\n",
    "           size=240000/len(model.adata),\n",
    "           title=\"Autotalker Latent Space\",\n",
    "           ax=ax1,\n",
    "           show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo1\"],\n",
    "              color=[latent_cluster_key],\n",
    "              palette=latent_cluster_colors,\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 1\",\n",
    "              legend_loc=None,\n",
    "              ax=ax2,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "              color=[latent_cluster_key],\n",
    "              palette=latent_cluster_colors,\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 2\",\n",
    "              legend_loc=None,\n",
    "              ax=ax3,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo3\"],\n",
    "              color=[latent_cluster_key],\n",
    "              palette=latent_cluster_colors,\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 3\",\n",
    "              legend_loc=None,\n",
    "              ax=ax4,\n",
    "              show=False)\n",
    "\n",
    "# Create and position shared legend\n",
    "handles, labels = ax1.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles, labels, bbox_to_anchor=(1.025, 0.8))\n",
    "ax1.get_legend().remove()\n",
    "\n",
    "# Adjust, save and display plot\n",
    "plt.subplots_adjust(wspace=0., hspace=0.85)\n",
    "fig.savefig(f\"{figure_folder_path}/latent_leiden_clusters_latent_physical_space.png\",\n",
    "            bbox_extra_artists=(lgd, title),\n",
    "            bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d825ec88-02c4-4da8-8158-4aaf02a9722d",
   "metadata": {},
   "source": [
    "#### 2.4.2 Visualize Query-enriched Latent Clusters in Physical and Latent Space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f2dad83-30dc-42c7-a071-26e2be9797cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get query-enriched latent leiden clusters\n",
    "adata_reference = model.adata[model.adata.obs[mapping_entity_key] == \"reference\"]\n",
    "adata_query = model.adata[model.adata.obs[mapping_entity_key] == \"query\"]\n",
    "latent_cluster_reference_proportions = adata_reference.obs[f\"latent_leiden_{str(latent_leiden_resolution)}\"].value_counts().sort_index() / len(adata_reference)\n",
    "latent_cluster_query_proportions = adata_query.obs[f\"latent_leiden_{str(latent_leiden_resolution)}\"].value_counts().sort_index() / len(adata_query)\n",
    "relative_latent_cluster_query_proportions = latent_cluster_query_proportions / latent_cluster_reference_proportions\n",
    "relative_latent_cluster_query_proportions.sort_values(ascending=False, inplace=True)\n",
    "display(relative_latent_cluster_query_proportions)\n",
    "\n",
    "query_enriched_latent_clusters = relative_latent_cluster_query_proportions[relative_latent_cluster_query_proportions > enriched_query_latent_cluster_prop_thresh].index.to_list()\n",
    "model.adata.obs[query_enriched_latent_cluster_key] = \"Latent clusters not enriched in query\"\n",
    "for latent_cluster in query_enriched_latent_clusters:\n",
    "    model.adata.obs.loc[model.adata.obs[f\"latent_leiden_{str(latent_leiden_resolution)}\"] == latent_cluster, query_enriched_latent_cluster_key] = f\"Cluster {latent_cluster}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "429e5ee8-2dbb-457d-a92d-1ae905b55b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subplot of cell-type annotations in physical and latent space\n",
    "fig = plt.figure(figsize=(12, 14))\n",
    "title = fig.suptitle(t=\"Autotalker: Query-enriched Latent Clusters in Physical and Latent Space\",\n",
    "                     y=0.96,\n",
    "                     x=0.55,\n",
    "                     fontsize=20)\n",
    "spec = gridspec.GridSpec(ncols=1,\n",
    "                         nrows=2,\n",
    "                         width_ratios=[1],\n",
    "                         height_ratios=[1, 5])\n",
    "ax1 = fig.add_subplot(2, 3, 1)\n",
    "ax2 = fig.add_subplot(2, 3, 2)\n",
    "ax3 = fig.add_subplot(2, 3, 3)\n",
    "ax4 = fig.add_subplot(spec[1])\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[query_enriched_latent_cluster_key],\n",
    "           size=240000/len(model.adata),\n",
    "           title=\"Autotalker Latent Space\",\n",
    "           ax=ax4,\n",
    "           show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo1\"],\n",
    "              color=[query_enriched_latent_cluster_key],\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 1\",\n",
    "              legend_loc=None,\n",
    "              ax=ax1,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "              color=[query_enriched_latent_cluster_key],\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 2\",\n",
    "              legend_loc=None,\n",
    "              ax=ax2,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo3\"],\n",
    "              color=[query_enriched_latent_cluster_key],\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 3\",\n",
    "              legend_loc=None,\n",
    "              ax=ax3,\n",
    "              show=False)\n",
    "\n",
    "# Create and position shared legend\n",
    "handles, labels = ax4.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles, labels, bbox_to_anchor=(1.285, 0.6575))\n",
    "ax4.get_legend().remove()\n",
    "\n",
    "# Adjust, save and display plot\n",
    "plt.subplots_adjust(wspace=0., hspace=0.85)\n",
    "fig.savefig(f\"{figure_folder_path}/query_enriched_latent_clusters_physical_latent_space.png\",\n",
    "            bbox_extra_artists=(lgd, title),\n",
    "            bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "212d5b6d-f3ec-4a55-9503-6bcaa6c81fd3",
   "metadata": {},
   "source": [
    "#### 2.4.3 Analyze GP Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17fcfc0c-5b52-4628-aefe-018866c6a62a",
   "metadata": {},
   "source": [
    "##### 2.4.3.1 Explore All GPs Across All Latent Clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60bd4aa9-82aa-40d5-954d-9e8354579234",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_label = \"latent_cluster\"\n",
    "\n",
    "# Compute gene program enrichments and retrieve top up- and downregulated gene programs\n",
    "top_unique_gps = model.compute_differential_gp_scores(cat_key=latent_cluster_key,\n",
    "                                                      adata=model.adata,\n",
    "                                                      selected_gps=None,\n",
    "                                                      selected_cats=None,\n",
    "                                                      gp_scores_weight_normalization=False,\n",
    "                                                      comparison_cats=\"rest\",\n",
    "                                                      n_sample=10000,\n",
    "                                                      key_added=latent_cluster_differential_gp_scores_key,\n",
    "                                                      n_top_up_gps_retrieved=n_top_up_gps,\n",
    "                                                      n_top_down_gps_retrieved=n_top_down_gps,\n",
    "                                                      seed=random_seed)\n",
    "\n",
    "# Display top upregulated gene programs\n",
    "top_up_gp_df = model.adata.uns[latent_cluster_differential_gp_scores_key][:n_top_up_gps]\n",
    "display(top_up_gp_df)\n",
    "\n",
    "# Display top downregulated gene programs\n",
    "top_down_gp_df = model.adata.uns[latent_cluster_differential_gp_scores_key][-n_top_down_gps:][::-1]\n",
    "display(top_down_gp_df)\n",
    "\n",
    "fig = sc.pl.dotplot(model.adata,\n",
    "                    top_unique_gps,\n",
    "                    groupby=cell_type_key,\n",
    "                    dendrogram=True, \n",
    "                    title=f\"{analysis_label.replace('_', ' ').title()} Differential GP Scores\",\n",
    "                    swap_axes=True,\n",
    "                    return_fig=True)\n",
    "# Save and display plot\n",
    "fig.savefig(f\"{figure_folder_path}/{analysis_label}_differential_gp_scores.png\")\n",
    "plt.show()\n",
    "\n",
    "# Inspect top up- and downregulated gene programs\n",
    "display(gp_summary_df[gp_summary_df[\"gp_name\"].isin(top_unique_gps)])\n",
    "\n",
    "top_cats = top_up_gp_df[\"category\"].append(top_down_gp_df[\"category\"]).to_list()\n",
    "top_gps = top_up_gp_df[\"gene_program\"].append(top_down_gp_df[\"gene_program\"]).to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8443574-f122-4f8b-aa69-ba34e7bdefc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cat_gp_score_gene_expr_summary(analysis_label=analysis_label,\n",
    "                                   model=model,\n",
    "                                   cats=top_cats,\n",
    "                                   gps=top_gps,\n",
    "                                   cat_key=latent_cluster_key,\n",
    "                                   adata=None,\n",
    "                                   feature_spaces=[\"latent\"], # [\"physical_embryo1\", \"physical_embryo2\", \"physical_embryo3\"],\n",
    "                                   plot_types=[\"gene_categories\", \"individual_genes\"],\n",
    "                                   n_top_genes_per_gp=5,\n",
    "                                   save_figs=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d12cb259-2df8-4d9c-8ca5-3f7a8dee41ae",
   "metadata": {},
   "source": [
    "##### 2.4.3.2 Explore All GPs Across Query-enriched Latent Clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aff07840-f813-40ee-9947-e31f777f2ca6",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_label = \"query_enriched_latent_cluster\"\n",
    "\n",
    "# Compute gene program enrichments and retrieve top up- and downregulated gene programs\n",
    "top_unique_gps = model.compute_differential_gp_scores(cat_key=latent_cluster_key,\n",
    "                                                      adata=model.adata,\n",
    "                                                      selected_gps=None,\n",
    "                                                      selected_cats=query_enriched_latent_clusters,\n",
    "                                                      gp_scores_weight_normalization=False,\n",
    "                                                      comparison_cats=\"rest\",\n",
    "                                                      n_sample=10000,\n",
    "                                                      key_added=latent_cluster_differential_gp_scores_key,\n",
    "                                                      n_top_up_gps_retrieved=n_top_up_gps,\n",
    "                                                      n_top_down_gps_retrieved=n_top_down_gps,\n",
    "                                                      seed=random_seed)\n",
    "\n",
    "# Display top upregulated gene programs\n",
    "top_up_gp_df = model.adata.uns[latent_cluster_differential_gp_scores_key][:n_top_up_gps]\n",
    "display(top_up_gp_df)\n",
    "\n",
    "# Display top downregulated gene programs\n",
    "top_down_gp_df = model.adata.uns[latent_cluster_differential_gp_scores_key][-n_top_down_gps:][::-1]\n",
    "display(top_down_gp_df)\n",
    "\n",
    "fig = sc.pl.dotplot(model.adata,\n",
    "                    top_unique_gps,\n",
    "                    groupby=cell_type_key,\n",
    "                    dendrogram=True, \n",
    "                    title=f\"{analysis_label.replace('_', ' ').title()} Differential GP Scores\",\n",
    "                    swap_axes=True,\n",
    "                    return_fig=True)\n",
    "# Save and display plot\n",
    "fig.savefig(f\"{figure_folder_path}/{analysis_label}_differential_gp_scores.png\")\n",
    "plt.show()\n",
    "\n",
    "# Inspect top up- and downregulated gene programs\n",
    "display(gp_summary_df[gp_summary_df[\"gp_name\"].isin(top_unique_gps)])\n",
    "\n",
    "top_cats = top_up_gp_df[\"category\"].append(top_down_gp_df[\"category\"]).to_list()\n",
    "top_gps = top_up_gp_df[\"gene_program\"].append(top_down_gp_df[\"gene_program\"]).to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef3d867b-85dd-49ee-b993-609c9b15415c",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cat_gp_score_gene_expr_summary(analysis_label=analysis_label,\n",
    "                                   model=model,\n",
    "                                   cats=top_cats,\n",
    "                                   gps=top_gps,\n",
    "                                   cat_key=latent_cluster_key,\n",
    "                                   adata=None,\n",
    "                                   feature_spaces=[\"latent\"], # [\"physical_embryo1\", \"physical_embryo2\", \"physical_embryo3\"],\n",
    "                                   plot_types=[\"gene_categories\", \"individual_genes\"],\n",
    "                                   n_top_genes_per_gp=5,\n",
    "                                   save_figs=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43d59624-580e-4813-bf53-15009c29ea0b",
   "metadata": {},
   "source": [
    "##### 2.4.3.3 Explore Specific GPs Across All Latent Clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d43e3a1-2d4c-47a3-880b-5ecc21167a7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_label = \"latent_cluster_bmp2_gp\"\n",
    "\n",
    "# Compute gene program enrichments and retrieve top up- and downregulated gene programs\n",
    "top_unique_gps = model.compute_differential_gp_scores(cat_key=latent_cluster_key,\n",
    "                                                      adata=model.adata,\n",
    "                                                      selected_gps=[\"BMP2_ligand_targetgenes_GP\"],\n",
    "                                                      selected_cats=None,\n",
    "                                                      gp_scores_weight_normalization=False,\n",
    "                                                      comparison_cats=\"rest\",\n",
    "                                                      n_sample=10000,\n",
    "                                                      key_added=latent_cluster_differential_gp_scores_key,\n",
    "                                                      n_top_up_gps_retrieved=n_top_up_gps,\n",
    "                                                      n_top_down_gps_retrieved=n_top_down_gps,\n",
    "                                                      seed=random_seed)\n",
    "\n",
    "# Display top upregulated gene programs\n",
    "top_up_gp_df = model.adata.uns[latent_cluster_differential_gp_scores_key][:n_top_up_gps]\n",
    "display(top_up_gp_df)\n",
    "\n",
    "# Display top downregulated gene programs\n",
    "top_down_gp_df = model.adata.uns[latent_cluster_differential_gp_scores_key][-n_top_down_gps:][::-1]\n",
    "display(top_down_gp_df)\n",
    "\n",
    "fig = sc.pl.dotplot(model.adata,\n",
    "                    top_unique_gps,\n",
    "                    groupby=cell_type_key,\n",
    "                    dendrogram=True, \n",
    "                    title=f\"{analysis_label.replace('_', ' ').title()} Differential GP Scores\",\n",
    "                    swap_axes=True,\n",
    "                    return_fig=True)\n",
    "# Save and display plot\n",
    "fig.savefig(f\"{figure_folder_path}/{analysis_label}_differential_gp_scores.png\")\n",
    "plt.show()\n",
    "\n",
    "# Inspect top up- and downregulated gene programs\n",
    "display(gp_summary_df[gp_summary_df[\"gp_name\"].isin(top_unique_gps)])\n",
    "\n",
    "top_cats = top_up_gp_df[\"category\"].append(top_down_gp_df[\"category\"]).to_list()\n",
    "top_gps = top_up_gp_df[\"gene_program\"].append(top_down_gp_df[\"gene_program\"]).to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e89740d-ca5a-4028-aa46-4f4a22c19c98",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cat_gp_score_gene_expr_summary(analysis_label=analysis_label,\n",
    "                                   model=model,\n",
    "                                   cats=top_cats,\n",
    "                                   gps=top_gps,\n",
    "                                   cat_key=latent_cluster_key,\n",
    "                                   adata=None,\n",
    "                                   feature_spaces=[\"latent\"], # [\"physical_embryo1\", \"physical_embryo2\", \"physical_embryo3\"],\n",
    "                                   plot_types=[\"gene_categories\", \"individual_genes\"],\n",
    "                                   n_top_genes_per_gp=5,\n",
    "                                   save_figs=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa759223-4a75-425c-88cf-ddf960bd33ef",
   "metadata": {},
   "source": [
    "## 3. Dataset-specific Analysis & Validation of Results from Lohoff et. al 2022"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3750853-200a-40e1-a9f4-9db2a4fc1b35",
   "metadata": {},
   "source": [
    "Cardiomyocytes are spatially and morphologically distinct from other cell types, endothelium is interspersed and spread across the entire embryo space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53cb4635-dba4-4338-bb4e-39bcacdaf3cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = [\"Cardiomyocytes\", \"Endothelium\"]\n",
    "\n",
    "# Create plot of cell type annotations in physical and latent space\n",
    "fig = plt.figure(figsize=(12, 14))\n",
    "title = fig.suptitle(t=\"Autotalker: Cardiomyocytes & Endothelium in Physical and Latent Space\",\n",
    "                     y=0.96,\n",
    "                     x=0.55,\n",
    "                     fontsize=20)\n",
    "spec = gridspec.GridSpec(ncols=1,\n",
    "                         nrows=2,\n",
    "                         width_ratios=[1],\n",
    "                         height_ratios=[1, 5])\n",
    "ax1 = fig.add_subplot(2, 3, 1)\n",
    "ax2 = fig.add_subplot(2, 3, 2)\n",
    "ax3 = fig.add_subplot(2, 3, 3)\n",
    "ax4 = fig.add_subplot(spec[1])\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo1\"],\n",
    "              color=[cell_type_key],\n",
    "              palette=cell_type_colors,\n",
    "              groups=cell_types,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 1\",\n",
    "              legend_loc=None,\n",
    "              ax=ax1,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "              color=[cell_type_key],\n",
    "              palette=cell_type_colors,\n",
    "              groups=cell_types,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 2\",\n",
    "              legend_loc=None,\n",
    "              ax=ax2,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo3\"],\n",
    "              color=[cell_type_key],\n",
    "              palette=cell_type_colors,\n",
    "              groups=cell_types,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 3\",\n",
    "              legend_loc=None,\n",
    "              ax=ax3,\n",
    "              show=False)\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[cell_type_key],\n",
    "           palette=cell_type_colors,\n",
    "           groups=cell_types,\n",
    "           size=1280000/len(model.adata),\n",
    "           title=\"Autotalker Latent Space\",\n",
    "           ax=ax4,\n",
    "           show=False)\n",
    "\n",
    "# Create and position shared legend\n",
    "handles, labels = ax4.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles, labels, bbox_to_anchor=(1.12, 0.655))\n",
    "ax4.get_legend().remove()\n",
    "\n",
    "# Adjust, save and display plot\n",
    "plt.subplots_adjust(wspace=0., hspace=0.85)\n",
    "fig.savefig(f\"{figure_folder_path}/cell_types_cardiomyocytes_endothelium_physical_latent_space.png\",\n",
    "            bbox_extra_artists=(lgd, title),\n",
    "            bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dbe8b0d-6056-4bea-9201-d80c7bb04a67",
   "metadata": {},
   "source": [
    "### 3.2 Mixed Mesenchymal Mesoderm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e73294bc-af70-434a-bfc5-27763476239c",
   "metadata": {},
   "source": [
    "\"mixed mesenchymal mesoderm, represent \n",
    "a cell state rather than a defined cell type. Mesenchyme represents  a  state  in  which  cells  express  markers  characteristic  of  migratory   cells loosely dispersed within an extracellular matrix56. This strong  overriding transcriptional signature of mesenchyme, irrespective of  location,  makes  it  challenging  to  distinguish  which  cell  types  this   mixed  mesenchymal  mesoderm  population  represents  using  clas- sical scRNA-seq data\"\n",
    "\n",
    "We can identify distinct subpopulations that are spatially defined."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c717a4ce-8a24-4166-8c51-c589e7854571",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type = \"mixed_mesenchymal_mesoderm\"\n",
    "cell_type_latent_resolution = 0.03 # 0.1\n",
    "cell_type_latent_cluster_key = f\"{cell_type}_latent_leiden_{cell_type_latent_resolution}\"\n",
    "cell_type_differential_gp_scores_key = f\"{cell_type}_differential_gp_scores\"\n",
    "cell_type_adata = model.adata[model.adata.obs[cell_type_key] == cell_type.replace(\"_\", \" \").capitalize()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc1cfa83-cc52-498a-addb-7e6310c34756",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute latent nearest neighbor graph for cell type only\n",
    "sc.pp.neighbors(cell_type_adata,\n",
    "                use_rep=latent_key,\n",
    "                key_added=f\"{cell_type}_latent_knng\")\n",
    "\n",
    "# Compute latent Leiden clustering for cell type\n",
    "sc.tl.leiden(adata=cell_type_adata,\n",
    "             resolution=cell_type_latent_resolution,\n",
    "             random_state=random_seed,\n",
    "             key_added=cell_type_latent_cluster_key,\n",
    "             adjacency=cell_type_adata.obsp[f\"{cell_type}_latent_knng_connectivities\"])\n",
    "\n",
    "# Use cell type latent space for UMAP generation\n",
    "sc.tl.umap(cell_type_adata,\n",
    "           neighbors_key=f\"{cell_type}_latent_knng\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a490772e-3227-4843-b119-e5dac33c503e",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.obs[cell_type_latent_cluster_key] = np.nan\n",
    "model.adata.obs.loc[model.adata.obs[cell_type_key] == cell_type.replace(\"_\", \" \").capitalize(),\n",
    "                    cell_type_latent_cluster_key] = cell_type_adata.obs[cell_type_latent_cluster_key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de762fad-aeb9-4de5-a2c7-b1e414b3b080",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create plot of cell type annotations in physical and latent space\n",
    "fig = plt.figure(figsize=(12, 14))\n",
    "title = fig.suptitle(t=f\"{cell_type.replace('_', ' ').title()} Latent Clusters in Latent and Physical Space\",\n",
    "                     y=0.96,\n",
    "                     x=0.55,\n",
    "                     fontsize=20)\n",
    "spec1 = gridspec.GridSpec(ncols=1,\n",
    "                          nrows=2,\n",
    "                          width_ratios=[1],\n",
    "                          height_ratios=[3, 2])\n",
    "spec2 = gridspec.GridSpec(ncols=3,\n",
    "                          nrows=2,\n",
    "                          width_ratios=[1, 1, 1],\n",
    "                          height_ratios=[3, 2])\n",
    "ax1 = fig.add_subplot(spec1[0])\n",
    "ax2 = fig.add_subplot(spec2[3])\n",
    "ax3 = fig.add_subplot(spec2[4])\n",
    "ax4 = fig.add_subplot(spec2[5])\n",
    "sc.pl.umap(adata=cell_type_adata,\n",
    "           color=[cell_type_latent_cluster_key],\n",
    "           palette=latent_cluster_colors,\n",
    "           size=1280000/len(model.adata),\n",
    "           title=f\"{cell_type.replace('_', ' ').title()} Latent Clusters in Latent Space\",\n",
    "           ax=ax1,\n",
    "           show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo1\"],\n",
    "              color=[cell_type_latent_cluster_key],\n",
    "              palette=latent_cluster_colors,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=f\"{cell_type.replace('_', ' ').title()} \\n Latent Clusters in \\n Physical Space Embryo 1\",\n",
    "              legend_loc=None,\n",
    "              ax=ax2,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "              color=[cell_type_latent_cluster_key],\n",
    "              palette=latent_cluster_colors,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=f\"{cell_type.replace('_', ' ').title()} \\n Latent Clusters in \\n Physical Space Embryo 2\",\n",
    "              legend_loc=None,\n",
    "              ax=ax3,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo3\"],\n",
    "              color=[cell_type_latent_cluster_key],\n",
    "              palette=latent_cluster_colors,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=f\"{cell_type.replace('_', ' ').title()} \\n Latent Clusters in \\n Physical Space Embryo 3\",\n",
    "              legend_loc=None,\n",
    "              ax=ax4,\n",
    "              show=False)\n",
    "\n",
    "# Create and position shared legend\n",
    "handles, labels = ax1.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles, labels, bbox_to_anchor=(1.05, 0.5))\n",
    "ax1.get_legend().remove()\n",
    "\n",
    "# Adjust, save and display plot\n",
    "plt.subplots_adjust(wspace=0.2, hspace=0.25)\n",
    "#fig.savefig(f\"{figure_folder_path}/{cell_type}_latent_clusters_physical_latent_space.png\",\n",
    "#            bbox_extra_artists=(lgd, title),\n",
    "#            bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e7c371a-e1a1-4358-9be4-93759a8a9d28",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.get_gp_summary()[model.get_gp_summary()[\"gp_name\"] == \"BMP2_ligand_targetgenes_GP\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc4e8409-0761-4552-8433-c66e33ea6975",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.obsm[\"autotalker_latent\"][:, 22]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60a5e0d1-af32-4d42-bbb9-06edd18ff365",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_adata.obsm[\"autotalker_latent\"][:, 22]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "416f1119-6fb1-4a2e-b54a-58731e347154",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_adata.obs.loc[cell_type_adata.obs[cell_type_latent_cluster_key] == \"1\", \"BMP2_ligand_targetgenes_GP\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "412b4e15-847c-4f85-9e05-877a834feeaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_adata.obs.loc[cell_type_adata.obs[cell_type_latent_cluster_key].isin([\"0\", \"2\"]), \"BMP2_ligand_targetgenes_GP\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18889228-619a-48d2-9019-01dd8d1e8cee",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_adata.X[cell_type_adata.obs[cell_type_latent_cluster_key] == \"1\", cell_type_adata.var_names == \"Myl9\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20fc0871-2b03-4a13-9c9a-4420b1d23916",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_adata.X[cell_type_adata.obs[cell_type_latent_cluster_key].isin([\"0\", \"2\"]), cell_type_adata.var_names == \"Myl9\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c8c3aed-23b8-4a39-a2fb-db4e0287ec25",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_adata.X[cell_type_adata.obs[cell_type_latent_cluster_key] == \"1\", cell_type_adata.var_names == \"Cd69\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b2e646b-d41d-4026-8b78-117de215861d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_adata.X[cell_type_adata.obs[cell_type_latent_cluster_key].isin([\"0\", \"2\"]), cell_type_adata.var_names == \"Cd69\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee63815c-34ea-43a1-83ef-c17b4aab3b32",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_adata.obs.loc[cell_type_adata.obs[cell_type_latent_cluster_key] == \"1\", \"BMP2_ligand_targetgenes_GP\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d30d87e5-4b01-421f-bf06-bed37608096b",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_label = \"mixed_mesenchymal_mesoderm_latent_cluster\"\n",
    "selected_cats = [\"1\"]\n",
    "comparison_cats = [\"0\", \"2\"]\n",
    "\n",
    "# Compute gene program enrichments and retrieve top up- and downregulated gene programs\n",
    "top_unique_gps = model.compute_differential_gp_scores(cat_key=cell_type_latent_cluster_key,\n",
    "                                                      adata=model.adata,\n",
    "                                                      selected_gps=None,\n",
    "                                                      selected_cats=selected_cats,\n",
    "                                                      gp_scores_weight_normalization=False,\n",
    "                                                      comparison_cats=comparison_cats,\n",
    "                                                      n_sample=10000,\n",
    "                                                      key_added=cell_type_differential_gp_scores_key,\n",
    "                                                      n_top_up_gps_retrieved=n_top_up_gps,\n",
    "                                                      n_top_down_gps_retrieved=n_top_down_gps,\n",
    "                                                      seed=random_seed)\n",
    "\n",
    "# Display top upregulated gene programs\n",
    "top_up_gp_df = model.adata.uns[cell_type_differential_gp_scores_key][:n_top_up_gps]\n",
    "display(top_up_gp_df)\n",
    "\n",
    "# Display top downregulated gene programs\n",
    "top_down_gp_df = model.adata.uns[cell_type_differential_gp_scores_key][-n_top_down_gps:][::-1]\n",
    "display(top_down_gp_df)\n",
    "\n",
    "fig = sc.pl.dotplot(model.adata,\n",
    "                    top_unique_gps,\n",
    "                    groupby=cell_type_latent_cluster_key,\n",
    "                    dendrogram=True, \n",
    "                    title=f\"{analysis_label.replace('_', ' ').title()} Differential GP Scores\",\n",
    "                    swap_axes=True,\n",
    "                    return_fig=True)\n",
    "# Save and display plot\n",
    "fig.savefig(f\"{figure_folder_path}/{analysis_label}_differential_gp_scores.png\")\n",
    "plt.show()\n",
    "\n",
    "# Inspect top up- and downregulated gene programs\n",
    "display(gp_summary_df[gp_summary_df[\"gp_name\"].isin(top_unique_gps)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9eac92bc-0469-4a9a-b8a1-b2805e94a3ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_cats = top_up_gp_df[\"category\"].append(top_down_gp_df[\"category\"]).to_list()\n",
    "top_gps = top_up_gp_df[\"gene_program\"].append(top_down_gp_df[\"gene_program\"]).to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53004ff8-5f20-4234-bf35-2656d703dc0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_adata.uns[cell_type_differential_gp_scores_key] = model.adata.uns[cell_type_differential_gp_scores_key]\n",
    "\n",
    "for gp in top_unique_gps:\n",
    "    cell_type_adata.obs[gp] = model.adata.obs.loc[cell_type_adata.obs.index, gp]\n",
    "\n",
    "top_cats = top_up_gp_df[\"category\"].append(top_down_gp_df[\"category\"]).to_list()\n",
    "top_gps = top_up_gp_df[\"gene_program\"].append(top_down_gp_df[\"gene_program\"]).to_list()\n",
    "\n",
    "get_cat_gp_score_gene_expr_summary(analysis_label=\"mixed_mesenchymal_mesoderm_latent_cluster\",\n",
    "                                   model=model,\n",
    "                                   cats=top_cats,\n",
    "                                   gps=top_gps,\n",
    "                                   cat_key=cell_type_latent_cluster_key,\n",
    "                                   adata=cell_type_adata,\n",
    "                                   feature_spaces=[\"latent\", \"physical_embryo1\", \"physical_embryo2\", \"physical_embryo3\"],\n",
    "                                   plot_types=[\"gene_categories\", \"individual_genes\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "917290bf-adb1-4489-8be2-0833903eb0e6",
   "metadata": {},
   "source": [
    "BMP signaling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40e85b96-b6cf-4c74-a168-f3de3963f793",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.compute_gp_gene_importances(selected_gp=\"CSH1_ligand_targetgenes_GP\")[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9be4e4f0-1bcc-484e-8386-25b06d28c388",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b1fdd5b-2737-4f46-b18a-33ed17b70a22",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "cell_type_latent_cluster_key = cell_type + \"_latent_cluster\"\n",
    "\n",
    "model.adata.obs[cell_type_latent_cluster_key] = None\n",
    "model.adata.obs.loc[model.adata.obs[cell_type_key] == cell_type.replace(\"_\", \" \").capitalize(),\n",
    "                    cell_type_latent_cluster_key] = model.adata.obs[latent_cluster_key]\n",
    "\n",
    "# Create plot of cell type annotations in physical and latent space\n",
    "fig = plt.figure(figsize=(12, 21))\n",
    "title = fig.suptitle(t=f\"{cell_type.replace('_', ' ').title()} Latent Clusters in Physical and Latent Space\",\n",
    "                     y=0.96,\n",
    "                     x=0.55,\n",
    "                     fontsize=20)\n",
    "spec1 = gridspec.GridSpec(ncols=1,\n",
    "                          nrows=3,\n",
    "                          width_ratios=[1],\n",
    "                          height_ratios=[3, 3, 2])\n",
    "spec2 = gridspec.GridSpec(ncols=3,\n",
    "                          nrows=3,\n",
    "                          width_ratios=[1, 1, 1],\n",
    "                          height_ratios=[3, 3, 2])\n",
    "ax1 = fig.add_subplot(spec1[0])\n",
    "ax2 = fig.add_subplot(spec1[1])\n",
    "ax3 = fig.add_subplot(spec2[6])\n",
    "ax4 = fig.add_subplot(spec2[7])\n",
    "ax5 = fig.add_subplot(spec2[8])\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[cell_type_key],\n",
    "           palette=cell_type_colors,\n",
    "           groups=cell_types,\n",
    "           size=1280000/len(model.adata),\n",
    "           title=f\"{cell_type.replace('_', ' ').title()} in Latent Space\",\n",
    "           ax=ax1,\n",
    "           show=False)\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[cell_type_latent_cluster_key],\n",
    "           palette=latent_cluster_colors,\n",
    "           size=1280000/len(model.adata),\n",
    "           title=f\"{cell_type.replace('_', ' ').title()} Latent Clusters in Latent Space\",\n",
    "           ax=ax2,\n",
    "           show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo1\"],\n",
    "              color=[cell_type_latent_cluster_key],\n",
    "              palette=latent_cluster_colors,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=f\"{cell_type.replace('_', ' ').title()} \\n Latent Clusters in \\n Physical Space Embryo 1\",\n",
    "              legend_loc=None,\n",
    "              ax=ax3,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "              color=[cell_type_latent_cluster_key],\n",
    "              palette=latent_cluster_colors,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=f\"{cell_type.replace('_', ' ').title()} \\n Latent Clusters in \\n Physical Space Embryo 2\",\n",
    "              legend_loc=None,\n",
    "              ax=ax4,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo3\"],\n",
    "              color=[\"mixed_mesenchymal_mesoderm_latent_cluster\"],\n",
    "              palette=latent_cluster_colors,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=f\"{cell_type.replace('_', ' ').title()} \\n Latent Clusters in \\n Physical Space Embryo 3\",\n",
    "              legend_loc=None,\n",
    "              ax=ax5,\n",
    "              show=False)\n",
    "\n",
    "# Create and position shared legend\n",
    "handles, labels = ax2.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles, labels, bbox_to_anchor=(1.05, 0.5))\n",
    "ax2.get_legend().remove()\n",
    "\n",
    "# Adjust, save and display plot\n",
    "plt.subplots_adjust(wspace=0.2, hspace=0.25)\n",
    "fig.savefig(f\"{figure_folder_path}/{cell_type}_latent_clusters_physical_latent_space.png\",\n",
    "            bbox_extra_artists=(lgd, title),\n",
    "            bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e9a93ad-7216-41de-be26-1c8f0d097c1c",
   "metadata": {},
   "source": [
    "Dorsal-ventral separation of esophageal and tracheal progenitor populations in the gut tube (not visible in scVI latent space)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac0b1690-edf0-4263-a4a0-053499ac24b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_latent_cluster = \"gut_tube_latent_cluster\"\n",
    "cell_type = \"Gut tube\"\n",
    "\n",
    "model.adata.obs[cell_type_latent_cluster] = None\n",
    "model.adata.obs.loc[model.adata.obs[cell_type_key] == cell_type,\n",
    "                    cell_type_latent_cluster] = model.adata.obs[latent_cluster_key]\n",
    "\n",
    "# Create plot of cell type annotations in physical and latent space\n",
    "fig = plt.figure(figsize=(12, 21))\n",
    "title = fig.suptitle(t=\"Autotalker: Gut Tube Latent Clusters in Physical and Latent Space\",\n",
    "                     y=0.96,\n",
    "                     x=0.55,\n",
    "                     fontsize=20)\n",
    "spec1 = gridspec.GridSpec(ncols=1,\n",
    "                          nrows=3,\n",
    "                          width_ratios=[1],\n",
    "                          height_ratios=[3, 3, 2])\n",
    "spec2 = gridspec.GridSpec(ncols=3,\n",
    "                          nrows=3,\n",
    "                          width_ratios=[1, 1, 1],\n",
    "                          height_ratios=[3, 3, 2])\n",
    "ax1 = fig.add_subplot(spec1[0])\n",
    "ax2 = fig.add_subplot(spec1[1])\n",
    "ax3 = fig.add_subplot(spec2[6])\n",
    "ax4 = fig.add_subplot(spec2[7])\n",
    "ax5 = fig.add_subplot(spec2[8])\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[cell_type_key],\n",
    "           palette=cell_type_colors,\n",
    "           groups=[cell_type],\n",
    "           size=1280000/len(model.adata),\n",
    "           title=\"Gut Tube in Latent Space\",\n",
    "           ax=ax1,\n",
    "           show=False)\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[cell_type_latent_cluster],\n",
    "           palette=latent_cluster_colors,\n",
    "           size=1280000/len(model.adata),\n",
    "           title=f\"{cell_type.title()} Latent Clusters in Latent Space\",\n",
    "           ax=ax2,\n",
    "           show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo1\"],\n",
    "              color=[cell_type_latent_cluster],\n",
    "              palette=latent_cluster_colors,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=f\"{cell_type.title()} \\n Latent Clusters in \\n Physical Space Embryo 1\",\n",
    "              legend_loc=None,\n",
    "              ax=ax3,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "              color=[cell_type_latent_cluster],\n",
    "              palette=latent_cluster_colors,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=f\"{cell_type.title()} \\n Latent Clusters in \\n Physical Space Embryo 2\",\n",
    "              legend_loc=None,\n",
    "              ax=ax4,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo3\"],\n",
    "              color=[cell_type_latent_cluster],\n",
    "              palette=latent_cluster_colors,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=f\"{cell_type.title()} \\n Latent Clusters in \\n Physical Space Embryo 3\",\n",
    "              legend_loc=None,\n",
    "              ax=ax5,\n",
    "              show=False)\n",
    "\n",
    "# Create and position shared legend\n",
    "handles, labels = ax2.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles, labels, bbox_to_anchor=(1.05, 0.5))\n",
    "ax2.get_legend().remove()\n",
    "\n",
    "# Adjust, save and display plot\n",
    "plt.subplots_adjust(wspace=0., hspace=0.85)\n",
    "fig.savefig(f\"{figure_folder_path}/gut_tube_latent_cluster_physical_latent_space.png\",\n",
    "            bbox_extra_artists=(lgd, title),\n",
    "            bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36314dbb-7a56-4e6c-948f-cee2247ecb9e",
   "metadata": {},
   "source": [
    "Dorsal-ventral and rostral-caudal spatially resolved patterns of the midbrain and hindbrain region."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8d42b1a-8284-425e-a3ee-ce2ba4d09671",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = [\"Forebrain/Midbrain/Hindbrain\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d14256cb-6506-4f26-b6d3-1d0d6e95c8e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create plot of cell type annotations in physical and latent space\n",
    "fig = plt.figure(figsize=(12, 14))\n",
    "title = fig.suptitle(t=\"Autotalker: Cell Types in Physical and Latent Space\",\n",
    "                     y=0.96,\n",
    "                     x=0.55,\n",
    "                     fontsize=20)\n",
    "spec = gridspec.GridSpec(ncols=1,\n",
    "                         nrows=2,\n",
    "                         width_ratios=[1],\n",
    "                         height_ratios=[1, 5])\n",
    "ax1 = fig.add_subplot(2, 3, 1)\n",
    "ax2 = fig.add_subplot(2, 3, 2)\n",
    "ax3 = fig.add_subplot(2, 3, 3)\n",
    "ax4 = fig.add_subplot(spec[1])\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo1\"],\n",
    "              color=[cell_type_key],\n",
    "              palette=\"Accent\",\n",
    "              groups=cell_types,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 1\",\n",
    "              legend_loc=None,\n",
    "              ax=ax1,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "              color=[cell_type_key],\n",
    "              palette=\"Accent\",\n",
    "              groups=cell_types,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 2\",\n",
    "              legend_loc=None,\n",
    "              ax=ax2,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo3\"],\n",
    "              color=[cell_type_key],\n",
    "              palette=\"Accent\",\n",
    "              groups=cell_types,\n",
    "              size=160000/len(model.adata),\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 3\",\n",
    "              legend_loc=None,\n",
    "              ax=ax3,\n",
    "              show=False)\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[cell_type_key],\n",
    "           palette=\"Accent\",\n",
    "           groups=cell_types,\n",
    "           size=1280000/len(model.adata),\n",
    "           title=\"Autotalker Latent Space\",\n",
    "           ax=ax4,\n",
    "           show=False)\n",
    "\n",
    "# Create and position shared legend\n",
    "handles, labels = ax4.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles, labels, bbox_to_anchor=(1.12, 0.6475))\n",
    "ax4.get_legend().remove()\n",
    "\n",
    "# Adjust, save and display plot\n",
    "plt.subplots_adjust(wspace=0., hspace=0.85)\n",
    "#fig.savefig(f\"{figure_folder_path}/cell_types_cardiomyocytes_endothelium_physical_latent_space.png\",\n",
    "#            bbox_extra_artists=(lgd, title),\n",
    "#            bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5db61ea-f79b-4536-9e04-933a645bbdc0",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 4. Learn De-Novo CCI GPs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb184039-274f-4017-804d-1700f963c4fb",
   "metadata": {},
   "source": [
    "### 4.1 Initialize, Train & Save Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83c4f04b-2f11-464c-bbf9-43a29e72d62b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load trained model\n",
    "if load_timestamp is not None:\n",
    "    model_artifacts_load_folder_path = f\"../artifacts/{dataset}/batch_integration/{load_timestamp}\"\n",
    "else:\n",
    "    model_artifacts_load_folder_path = model_artifacts_folder_path\n",
    "\n",
    "model = Autotalker.load(dir_path=f\"{model_artifacts_load_folder_path}/reference\",\n",
    "                        adata=None,\n",
    "                        adata_file_name=f\"{dataset}.h5ad\",\n",
    "                        n_addon_gps=10,\n",
    "                        gp_names_key=gp_names_key,\n",
    "                        genes_idx_key=genes_idx_key,\n",
    "                        unfreeze_all_weights=False,\n",
    "                        unfreeze_addon_gp_weights=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "538ac804-966c-46f3-8234-2211229c793a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for param_name, param in model.model.named_parameters():\n",
    "    print(param_name, param.requires_grad)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "033963da-40f8-4340-bcc0-753cb36b8fc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.train(n_epochs=3,\n",
    "            n_epochs_all_gps=3,\n",
    "            lr=lr,\n",
    "            lambda_edge_recon=1.,\n",
    "            lambda_gene_expr_recon=0.001,\n",
    "            lambda_l1_addon=0.003,\n",
    "            verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32baddaa-18d7-46be-9f1b-f438a9e7e21c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save trained model\n",
    "model.save(dir_path=model_artifacts_folder_path + \"/reference_query_denovo_gps\",\n",
    "           overwrite=True,\n",
    "           save_adata=True,\n",
    "           adata_file_name=f\"{dataset}.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d74a6739-b6b9-4a7c-8f08-eddfa411839e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load trained model\n",
    "model = Autotalker.load(dir_path=model_artifacts_folder_path + \"/reference_query_denovo_gps\",\n",
    "                        adata=None,\n",
    "                        adata_file_name=f\"{dataset}.h5ad\",\n",
    "                        gp_names_key=gp_names_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68ffdce1-d7d1-42a4-9478-a3e42172b086",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check number of active gene programs\n",
    "active_gps = model.get_active_gps()\n",
    "print(f\"Number of total gene programs: {len(model.adata.uns[gp_names_key])}.\")\n",
    "print(f\"Number of active gene programs: {len(active_gps)}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62c7ca8d-c9cf-4143-996c-2e70dfecc1f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(active_gps[-10:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0747a494-8b48-4db0-93cc-3668a1c990e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf7fb4c1-8c3a-463c-bb3a-7bfa7f79a5ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "addon_gp_summary_df = model.get_gp_summary()\n",
    "addon_gp_summary_df[(addon_gp_summary_df[\"gp_name\"].str.contains(\"addon\")) & \n",
    "                    (addon_gp_summary_df[\"gp_active\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf310c32-f5a9-4601-9cac-aac46b7efd7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get active addon gps\n",
    "active_addon_gp_names = [active_gp for active_gp in active_gps if \"addon\" in active_gp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40f492d1-c6aa-447f-b56f-fcc3086f2c48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute gene program enrichments and retrieve top up- and downregulated gene programs\n",
    "# for the query-enriched latent clusters\n",
    "top_unique_gps = model.compute_differential_gp_scores(cat_key=cell_type_key,\n",
    "                                                      adata=model.adata,\n",
    "                                                      selected_gps=active_addon_gp_names,\n",
    "                                                      selected_cats=None,\n",
    "                                                      gp_scores_weight_normalization=False,\n",
    "                                                      comparison_cats=\"rest\",\n",
    "                                                      n_sample=1000,\n",
    "                                                      key_added=denovo_cell_type_differential_gp_scores_key,\n",
    "                                                      n_top_up_gps_retrieved=n_top_up_gps,\n",
    "                                                      n_top_down_gps_retrieved=n_top_down_gps,\n",
    "                                                      seed=random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff13f27d-3b62-4fb4-bcd9-2345c18110db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display top upregulated gene programs\n",
    "top_up_gp_df = model.adata.uns[denovo_cell_type_differential_gp_scores_key][:n_top_up_gps]\n",
    "display(top_up_gp_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9de58d5c-72d7-4206-b05c-fc9ac81aa163",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display top downregulated gene programs\n",
    "top_down_gp_df = model.adata.uns[denovo_cell_type_differential_gp_scores_key][-n_top_down_gps:][::-1]\n",
    "display(top_down_gp_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cf7138f-f753-4754-ae25-dc23f0ce049f",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = sc.pl.dotplot(model.adata,\n",
    "                    top_unique_gps,\n",
    "                    groupby=cell_type_key,\n",
    "                    dendrogram=True, \n",
    "                    title=\"Differential GP Scores Add-on GPs\",\n",
    "                    swap_axes=True,\n",
    "                    return_fig=True)\n",
    "\n",
    "# Save and display plot\n",
    "fig.savefig(f\"{figure_folder_path}/differential_gp_scores_addon_gps.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffa3538c-3e50-4234-abca-5f9f9bed4902",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.compute_gp_gene_importances(selected_gp=active_addon_gp_names[0])[:50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38b383a1-4240-4a96-976f-68e92fede8bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.compute_gp_gene_importances(selected_gp=active_addon_gp_names[1])[:50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "340c4906-df0d-4887-b37d-bba735a1c0fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f08b0ef-1b61-49f0-960d-94a7931bc5b0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ac6acba-1b2f-403a-b0bc-bfb84ee60d21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create subplot of cell-type annotations in physical and latent space\n",
    "fig = plt.figure(figsize=(12, 20))\n",
    "title = fig.suptitle(t=\"Autotalker: Query-enriched Cell Types in Physical and Latent Space\",\n",
    "                     y=0.96,\n",
    "                     x=0.55,\n",
    "                     fontsize=20)\n",
    "spec = gridspec.GridSpec(ncols=1,\n",
    "                         nrows=3,\n",
    "                         width_ratios=[1],\n",
    "                         height_ratios=[1, 1, 0.7])\n",
    "spec2 = gridspec.GridSpec(ncols=3,\n",
    "                          nrows=3,\n",
    "                          width_ratios=[1, 1, 1],\n",
    "                          height_ratios=[1, 1, 0.7])\n",
    "ax1 = fig.add_subplot(spec[0])\n",
    "ax2 = fig.add_subplot(spec[1])\n",
    "ax3 = fig.add_subplot(spec2[6])\n",
    "ax4 = fig.add_subplot(spec2[7])\n",
    "ax5 = fig.add_subplot(spec2[8])\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[\"mapping_entity\"],\n",
    "           palette=\"Set1\",\n",
    "           size=240000/len(model.adata),\n",
    "           title=\"Autotalker Latent Space\",\n",
    "           ax=ax1,\n",
    "           show=False)\n",
    "sc.pl.umap(adata=model.adata,\n",
    "           color=[query_enriched_cell_type_key],\n",
    "           palette=\"coolwarm\",\n",
    "           size=240000/len(model.adata),\n",
    "           title=\"Autotalker Latent Space\",\n",
    "           ax=ax2,\n",
    "           show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo1\"],\n",
    "              color=[query_enriched_cell_type_key],\n",
    "              palette=\"coolwarm\",\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 1\",\n",
    "              legend_loc=None,\n",
    "              ax=ax3,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "              color=[query_enriched_cell_type_key],\n",
    "              palette=\"coolwarm\",\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 2\",\n",
    "              legend_loc=None,\n",
    "              ax=ax4,\n",
    "              show=False)\n",
    "sc.pl.spatial(adata=model.adata[model.adata.obs[\"sample\"] == \"embryo3\"],\n",
    "              color=[query_enriched_cell_type_key],\n",
    "              palette=\"coolwarm\",\n",
    "              spot_size=0.03,\n",
    "              title=\"Physical Space Embryo 3\",\n",
    "              legend_loc=None,\n",
    "              ax=ax5,\n",
    "              show=False)\n",
    "\n",
    "# Create and position shared legend\n",
    "handles, labels = ax2.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles, labels, bbox_to_anchor=(1.25, 0.465))\n",
    "ax2.get_legend().remove()\n",
    "\n",
    "# Adjust, save and display plot\n",
    "plt.subplots_adjust(wspace=0., hspace=0.2)\n",
    "fig.savefig(f\"{figure_folder_path}/query_enriched_cell_types_physical_latent_space.png\",\n",
    "            bbox_extra_artists=(lgd, title),\n",
    "            bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1829568-8bff-41fa-9d9b-cf1d4f2a0603",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
