{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bdc8086b-2b80-43b4-b88a-aa0c87c820a6",
   "metadata": {},
   "source": [
    "# Slurm Job Submission"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be88d0ca-4c09-44e3-83fc-fb6bcf4563fd",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>).\n",
    "- **Affiliation:** Helmholtz Munich, Institute of Computational Biology (ICB), Talavera-LÃ³pez Lab\n",
    "- **Date of Creation:** 20.03.2023\n",
    "- **Date of Last Modification:** 27.06.2023"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aeeeb1fe-eaa1-4cdd-aa21-ed78e0b7be81",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e58b0a3-c3fe-43d5-a420-14111aa545f6",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "48d297b0-e0dc-42b9-a349-e6ac2edba0e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aab82f4c-edd8-47dd-be56-559cef739439",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0491ee0e-3587-4243-856f-e7f71e6dced6",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda_env_name = \"nichecompass\"\n",
    "reproducibility_conda_env_name = \"nichecompass-reproducibility\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "004ab52a-9cc3-4e40-b0e6-f9fed7f4a549",
   "metadata": {},
   "source": [
    "### 1.3 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6b2d6cf0-b360-428f-9548-c7942378de68",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_python_script(\n",
    "        job_name_prefix,\n",
    "        job_id,\n",
    "        job_folder_path,\n",
    "        conda_env_name,\n",
    "        script_folder_path,\n",
    "        script_name,\n",
    "        script_args,\n",
    "        t=\"48:00:00\",\n",
    "        p=\"gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"gpu\",\n",
    "        nice=10000):\n",
    "    job_name = f\"{job_name_prefix}_{job_id}\"\n",
    "    # Account for fact that submit node has different home path than compute node\n",
    "    job_file_path = f\"{job_folder_path.replace('/aih', '')}/job_{job_name}.cmd\"\n",
    "    out_file_path = f\"{job_folder_path}/logs/out_{job_name}.txt\"\n",
    "    err_file_path = f\"{job_folder_path}/logs/err_{job_name}.txt\"\n",
    "    \n",
    "    os.makedirs(job_folder_path + \"/logs\", exist_ok=True)\n",
    "    \n",
    "    with open(job_file_path, \"w\") as handle:\n",
    "        handle.writelines(\"#!/bin/bash\\n\")\n",
    "        handle.writelines(f\"#SBATCH -J {job_name}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -o {out_file_path}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -e {err_file_path}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -t {t}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -p {p}\\n\")\n",
    "        handle.writelines(\"#SBATCH -c 6\\n\")\n",
    "        handle.writelines(f\"#SBATCH --gres={gres}\\n\")\n",
    "        handle.writelines(f\"#SBATCH --qos={qos}\\n\")       \n",
    "        handle.writelines(\"#SBATCH --mem=64GB\\n\")\n",
    "        handle.writelines(f\"#SBATCH --nice={nice}\\n\")\n",
    "        handle.writelines(\"source $HOME/.bashrc\\n\")\n",
    "        handle.writelines(f\"conda activate {conda_env_name}\\n\")\n",
    "        handle.writelines(\"cd /\\n\")\n",
    "        handle.writelines(f\"cd {script_folder_path}\\n\")\n",
    "        handle.writelines(f\"python ../{script_name}\")\n",
    "        handle.writelines(f\"{script_args}\")\n",
    "        handle.writelines(\"\\n\")\n",
    "        \n",
    "    os.system(f\"sbatch {job_file_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4624e92-33eb-457c-8767-9006e44d850e",
   "metadata": {},
   "source": [
    "## 2. NicheCompass Ablation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc99f810-89a4-4070-ac09-d3f6da04c95d",
   "metadata": {},
   "source": [
    "### 2.1 Loss Weights & Gene Program Mask"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b4617a3-879b-4591-95d9-76576cc0031f",
   "metadata": {},
   "source": [
    "Ablating:\n",
    "- Loss Weights: Gene Expression Reconstruction, Edge Reconstruction, KL Divergence\n",
    "- Gene Program Mask: FC GP Mask, NicheCompass Default GP Mask\n",
    "- Spatial Neighborhood Graph: Number of Neighbors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e95996-f6e0-459c-a400-5a2ae3193f47",
   "metadata": {},
   "source": [
    "#### 2.1.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84cfe23e-ec00-4c8d-af02-2aa210997565",
   "metadata": {},
   "source": [
    "##### 2.1.1.1 Fully Connected GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0621ca71-70a3-423e-9f79-91e706f5d4e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc62e692-6c03-42e6-9fd8-c6bdde35d409",
   "metadata": {},
   "source": [
    "##### 2.1.1.2 NicheCompass GP Mask (Keep All Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48a49459-8035-4d3e-82bb-9e594f12533b",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5bafadc-9eec-4815-860f-2230c4fabc1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1000\n",
    "for lambda_edge_recon in [50000]:\n",
    "    for lambda_gene_expr_recon in [300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b06564d-b53b-430e-91c5-9da00624290e",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 2000\n",
    "for lambda_edge_recon in [500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c46cace5-f7e8-4c76-bff1-3619ba15242c",
   "metadata": {},
   "source": [
    "##### 2.1.1.3 NicheCompass GP Mask (Weakly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a3fbb52-b4a6-4f1b-8361-fa068d7aa16a",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea24b9d8-45a1-4909-80c3-a2d96da1a9c1",
   "metadata": {},
   "source": [
    "##### 2.1.1.4 NicheCompass GP Mask (Strongly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4013b3f-cc14-41e6-bf75-fd3809199921",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f3fbdd9-c303-4781-91e6-9485dabf8e33",
   "metadata": {},
   "source": [
    "#### 2.1.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75e888c2-82b8-4d9f-b5e7-ac0d371bb0da",
   "metadata": {},
   "source": [
    "##### 2.1.2.1 Fully Connected GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "333a8e02-74cc-474e-9baa-49b02d348617",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f29be6dc-ee82-43bf-9a07-c8b4cacab09b",
   "metadata": {},
   "source": [
    "##### 2.1.2.2 NicheCompass GP Mask (Keep All Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "790911e2-ce3f-45cc-8e3d-3789cf1661e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7f18840-64b2-44dc-9cb8-e851ca3ddb27",
   "metadata": {},
   "source": [
    "##### 2.1.2.3 NicheCompass GP Mask (Weakly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50f5c146-9003-42a3-a434-b4bb9a34fe97",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c071b2e5-085a-444e-92f6-aebcd130767c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12416085\n",
      "Submitted batch job 12416086\n",
      "Submitted batch job 12416087\n",
      "Submitted batch job 12416088\n",
      "Submitted batch job 12416089\n",
      "Submitted batch job 12416090\n",
      "Submitted batch job 12416091\n",
      "Submitted batch job 12416092\n",
      "Submitted batch job 12416093\n",
      "Submitted batch job 12416094\n",
      "Submitted batch job 12416095\n",
      "Submitted batch job 12416096\n"
     ]
    }
   ],
   "source": [
    "job_id = 1000\n",
    "for lambda_edge_recon in [50000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0ba1c795-b445-4281-a0ad-09fe1b77826e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12416097\n"
     ]
    }
   ],
   "source": [
    "job_id = 2000\n",
    "for lambda_edge_recon in [0]:\n",
    "    for lambda_gene_expr_recon in [0]:\n",
    "        for n_neighbors in [16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0744941f-31c5-40fe-9835-31f36cff1e70",
   "metadata": {},
   "source": [
    "##### 2.1.2.4 NicheCompass GP Mask (Strongly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "962c9e96-ffed-4572-9b95-3d254ff285fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cb42f90-41a3-43e6-8747-18d91b527a92",
   "metadata": {},
   "source": [
    "#### 2.1.3 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e989cecb-0197-490e-8d11-5348bbdeffbc",
   "metadata": {},
   "source": [
    "##### 2.1.3.1 Fully Connected GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3a9d1e2b-a939-4b6f-bdee-87f767743618",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12363204\n",
      "Submitted batch job 12363205\n",
      "Submitted batch job 12363206\n",
      "Submitted batch job 12363207\n",
      "Submitted batch job 12363208\n",
      "Submitted batch job 12363209\n",
      "Submitted batch job 12363210\n",
      "Submitted batch job 12363211\n",
      "Submitted batch job 12363212\n",
      "Submitted batch job 12363213\n",
      "Submitted batch job 12363214\n",
      "Submitted batch job 12363215\n",
      "Submitted batch job 12363216\n",
      "Submitted batch job 12363217\n",
      "Submitted batch job 12363218\n",
      "Submitted batch job 12363219\n",
      "Submitted batch job 12363220\n",
      "Submitted batch job 12363221\n",
      "Submitted batch job 12363222\n",
      "Submitted batch job 12363223\n",
      "Submitted batch job 12363224\n",
      "Submitted batch job 12363225\n",
      "Submitted batch job 12363226\n",
      "Submitted batch job 12363227\n",
      "Submitted batch job 12363228\n",
      "Submitted batch job 12363229\n",
      "Submitted batch job 12363230\n",
      "Submitted batch job 12363231\n",
      "Submitted batch job 12363232\n",
      "Submitted batch job 12363233\n",
      "Submitted batch job 12363234\n",
      "Submitted batch job 12363235\n",
      "Submitted batch job 12363236\n",
      "Submitted batch job 12363237\n",
      "Submitted batch job 12363238\n",
      "Submitted batch job 12363239\n"
     ]
    }
   ],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ff6fd10-e19a-4d55-ab17-ae11a82665a3",
   "metadata": {},
   "source": [
    "##### 2.1.3.2 NicheCompass GP Mask (Keep All Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4125af6c-9aed-40a2-a5ad-ac00ccae7e72",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12420157\n",
      "Submitted batch job 12420158\n",
      "Submitted batch job 12420159\n",
      "Submitted batch job 12420160\n",
      "Submitted batch job 12420161\n",
      "Submitted batch job 12420162\n",
      "Submitted batch job 12420163\n",
      "Submitted batch job 12420164\n",
      "Submitted batch job 12420165\n",
      "Submitted batch job 12420166\n",
      "Submitted batch job 12420167\n",
      "Submitted batch job 12420168\n",
      "Submitted batch job 12420169\n",
      "Submitted batch job 12420170\n",
      "Submitted batch job 12420171\n",
      "Submitted batch job 12420172\n",
      "Submitted batch job 12420173\n",
      "Submitted batch job 12420174\n",
      "Submitted batch job 12420175\n",
      "Submitted batch job 12420176\n",
      "Submitted batch job 12420177\n",
      "Submitted batch job 12420178\n",
      "Submitted batch job 12420179\n",
      "Submitted batch job 12420180\n",
      "Submitted batch job 12420181\n",
      "Submitted batch job 12420182\n",
      "Submitted batch job 12420183\n",
      "Submitted batch job 12420184\n",
      "Submitted batch job 12420185\n",
      "Submitted batch job 12420186\n",
      "Submitted batch job 12420187\n",
      "Submitted batch job 12420188\n",
      "Submitted batch job 12420189\n",
      "Submitted batch job 12420190\n",
      "Submitted batch job 12420191\n",
      "Submitted batch job 12420192\n"
     ]
    }
   ],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "911dcd1e-0617-4116-aa06-0b9c596c292b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12442278\n"
     ]
    }
   ],
   "source": [
    "job_id = 5000\n",
    "for lambda_edge_recon in [500000]:\n",
    "    for lambda_gene_expr_recon in [0]:\n",
    "        for n_neighbors in [16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "375a4972-0b37-451e-bfbc-b152bcf45412",
   "metadata": {},
   "source": [
    "##### 2.1.3.3 NicheCompass GP Mask (Weakly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d5710e07-be54-4c0d-a76f-55ee625501c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12374961\n",
      "Submitted batch job 12374962\n",
      "Submitted batch job 12374963\n",
      "Submitted batch job 12374964\n",
      "Submitted batch job 12374965\n",
      "Submitted batch job 12374966\n",
      "Submitted batch job 12374967\n",
      "Submitted batch job 12374968\n",
      "Submitted batch job 12374969\n",
      "Submitted batch job 12374970\n",
      "Submitted batch job 12374971\n",
      "Submitted batch job 12374972\n",
      "Submitted batch job 12374973\n",
      "Submitted batch job 12374974\n",
      "Submitted batch job 12374975\n",
      "Submitted batch job 12374976\n",
      "Submitted batch job 12374977\n",
      "Submitted batch job 12374978\n",
      "Submitted batch job 12374979\n",
      "Submitted batch job 12374980\n",
      "Submitted batch job 12374981\n",
      "Submitted batch job 12374982\n",
      "Submitted batch job 12374983\n",
      "Submitted batch job 12374985\n",
      "Submitted batch job 12374986\n",
      "Submitted batch job 12374987\n",
      "Submitted batch job 12374988\n",
      "Submitted batch job 12374989\n",
      "Submitted batch job 12374990\n",
      "Submitted batch job 12374991\n",
      "Submitted batch job 12374992\n",
      "Submitted batch job 12374993\n",
      "Submitted batch job 12374994\n",
      "Submitted batch job 12374995\n",
      "Submitted batch job 12374996\n",
      "Submitted batch job 12374997\n"
     ]
    }
   ],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e47e0e96-832b-4d9f-a1ee-56761950904e",
   "metadata": {},
   "source": [
    "##### 2.1.3.4 NicheCompass GP Mask (Strongly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8ec32310-cb4f-40df-8036-b081d85874a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12411010\n",
      "Submitted batch job 12411011\n",
      "Submitted batch job 12411012\n",
      "Submitted batch job 12411013\n",
      "Submitted batch job 12411014\n",
      "Submitted batch job 12411015\n",
      "Submitted batch job 12411016\n",
      "Submitted batch job 12411017\n",
      "Submitted batch job 12411018\n",
      "Submitted batch job 12411019\n",
      "Submitted batch job 12411020\n",
      "Submitted batch job 12411021\n",
      "Submitted batch job 12411022\n",
      "Submitted batch job 12411023\n",
      "Submitted batch job 12411024\n",
      "Submitted batch job 12411025\n",
      "Submitted batch job 12411026\n",
      "Submitted batch job 12411027\n",
      "Submitted batch job 12411028\n",
      "Submitted batch job 12411029\n",
      "Submitted batch job 12411030\n",
      "Submitted batch job 12411031\n",
      "Submitted batch job 12411032\n",
      "Submitted batch job 12411033\n",
      "Submitted batch job 12411034\n",
      "Submitted batch job 12411035\n",
      "Submitted batch job 12411036\n",
      "Submitted batch job 12411037\n",
      "Submitted batch job 12411038\n",
      "Submitted batch job 12411039\n",
      "Submitted batch job 12411040\n",
      "Submitted batch job 12411041\n",
      "Submitted batch job 12411042\n",
      "Submitted batch job 12411043\n",
      "Submitted batch job 12411044\n",
      "Submitted batch job 12411045\n"
     ]
    }
   ],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ab4f058-e6e7-45d5-bce1-9432c1279590",
   "metadata": {},
   "source": [
    "#### 2.1.4 Metrics Computation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bc07a0d-3d70-40d3-8c90-d3ffd7c3f0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"ablation\"\n",
    "datasets = \"xenium_human_breast_cancer starmap_plus_mouse_cns\"\n",
    "ablation_task = \"loss_weights\"\n",
    "cell_type_keys = \"cell_states Main_molecular_cell_type\"\n",
    "condition_keys = \"None None\"\n",
    "experiment_ids = \"3 4\"\n",
    "\n",
    "job_name_prefix = f\"nichecompass_{ablation_task}_{task}_metrics_computation\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"compute_metrics.py\"\n",
    "script_args = f\" --ablation_task {ablation_task}\" \\\n",
    "              f\" --datasets {datasets}\" \\\n",
    "              f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "              f\" --condition_keys {condition_keys}\" \\\n",
    "              f\" --experiment_ids {experiment_ids}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=reproducibility_conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8734b56d-9239-4960-9689-f80ff56bf927",
   "metadata": {},
   "source": [
    "### 2.2 Gene Program & Gene Selection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7757c17-695f-4a02-9f39-645513a96ec4",
   "metadata": {},
   "source": [
    "Ablating:\n",
    "- Gene Program Selection: Active GP Threshold Ratio, Group Lasso Regularization\n",
    "- Gene Selection: L1 Regularization\n",
    "- Spatial Neighborhood Graph: Number of Neighbors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "985a6ac0-5138-497a-817d-4aca2cf73df9",
   "metadata": {},
   "source": [
    "#### 2.2.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0184c63e-5e14-4663-8970-751c1918d2e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for active_gp_thresh_ratio in [0., 0.05, 0.1]:\n",
    "    for lambda_group_lasso in [0, 5, 10]:\n",
    "        for lambda_l1_masked in [0., 5., 10., 50., 100.]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"gene_program_gene_selection\"\n",
    "                dataset = \"xenium_human_breast_cancer\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 512\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              \" --n_layers_encoder 1\" \\\n",
    "                              \" --n_hidden_encoder None\" \\\n",
    "                              \" --conv_layer_encoder gcnconv\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba1b231d-c2ba-4628-975a-9c8793568012",
   "metadata": {},
   "source": [
    "#### 2.2.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c74da10f-1b0e-482f-9028-f1172312a902",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for active_gp_thresh_ratio in [0., 0.05, 0.1]:\n",
    "    for lambda_group_lasso in [0, 5, 10]:\n",
    "        for lambda_l1_masked in [0., 5., 10., 50., 100.]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"gene_program_gene_selection\"\n",
    "                dataset = \"starmap_plus_mouse_cns\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"mouse\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 1024\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              \" --n_layers_encoder 1\" \\\n",
    "                              \" --n_hidden_encoder None\" \\\n",
    "                              \" --conv_layer_encoder gcnconv\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked 0.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21d51e95-d594-4fa6-bdcb-c44001904a73",
   "metadata": {},
   "source": [
    "#### 2.2.3 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f96de33-56ea-48db-bd1f-6437da817d90",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for active_gp_thresh_ratio in [0., 0.05, 0.1]:\n",
    "    for lambda_group_lasso in [0, 5, 10]:\n",
    "        for lambda_l1_masked in [0., 5., 10., 50., 100.]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"gene_program_gene_selection\"\n",
    "                dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "                reference_batches = \"batch2\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 1024\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              \" --n_layers_encoder 1\" \\\n",
    "                              \" --n_hidden_encoder None\" \\\n",
    "                              \" --conv_layer_encoder gcnconv\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked 0.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea685a7c-9ea8-428c-af96-c5bddfa48924",
   "metadata": {},
   "source": [
    "### 2.3 Encoder Architecture"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4085464-0b92-41f2-bc85-16bcbb8db164",
   "metadata": {},
   "source": [
    "#### 2.3.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e36b5743-f2a0-4b98-b68d-ab1525b96c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"encoder_architecture\"\n",
    "                dataset = \"xenium_human_breast_cancer\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 512\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "                active_gp_thresh_ratio = 0.05\n",
    "                lambda_group_lasso = 0.   \n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                              f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20369658-0353-4112-b2b1-c2f7b9e60565",
   "metadata": {},
   "source": [
    "#### 2.3.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7331e624-4c83-41a3-9a43-3dbeea67bb23",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"encoder_architecture\"\n",
    "                dataset = \"starmap_plus_mouse_cns\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"mouse\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 1024\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "                active_gp_thresh_ratio = 0.05\n",
    "                lambda_group_lasso = 0.\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                              f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked 0.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c11b0e1-14be-4abf-8381-c099bce0e3eb",
   "metadata": {},
   "source": [
    "#### 2.3.3 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af6c24a4-cecc-4ba0-81e2-8dcf7da09995",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"encoder_architecture\"\n",
    "                dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "                reference_batches = \"batch2\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 1024\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                              f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked 0.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29ad5d87-e074-423f-b9c5-43a0405ed227",
   "metadata": {},
   "source": [
    "### 2.5 Categorical Covariate Embedding & Contrastive Integration"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1e8207b-2b89-4fb1-89b0-c32cc207a8f3",
   "metadata": {},
   "source": [
    "#### 2.5.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d9ebf56-a5a1-4b64-b341-1108d269bbd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for cat_covariates_embeds_injection in [\"encoder\", \"gene_expr_decoder\"]:\n",
    "for lambda_cat_covariates_contrastive in [0., 1, 2]:\n",
    "    for contrastive_logits_pos_ratio in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"encoder_architecture\"\n",
    "                dataset = \"xenium_human_breast_cancer\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 512\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "                active_gp_thresh_ratio = 0.05\n",
    "                lambda_group_lasso = 0.\n",
    "                n_layers_encoder = 1\n",
    "                n_hidden_encoder = None\n",
    "                conv_layer_encoder = \"gcnconv\"\n",
    "                \n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                              f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              f\" --n_epochs_no_cat_covariates_contrastive {n_epochs_no_cat_covariates_contrastive}\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cc7c954-2df9-4224-9e57-7e6be8072b30",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c39d3f6-fe11-4056-bc48-9ef421bc9684",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a80b724d-8200-44b4-8dba-02269e4358e0",
   "metadata": {},
   "source": [
    "## 3. NicheCompass Single Sample Method Benchmarking Models Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eb67595-8ede-4823-a1fe-a313dee704f5",
   "metadata": {},
   "source": [
    "### 3.1 seqFISH Mouse Organogenesis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b52c28e8-6fc8-492f-9c31-d3ab9b815980",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.1.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe70711-9d03-47d5-892a-5cdf77714389",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_embryo2\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24ecebdb-2b38-4021-8116-b21061bef3b1",
   "metadata": {},
   "source": [
    "#### 3.1.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c3995e5-3c35-427e-838c-4831370d44c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct_embryo2\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19574d56-9c79-4874-9b4d-ee39248f0028",
   "metadata": {},
   "source": [
    "### 3.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe560799-9d98-4740-af7a-b972c7596d18",
   "metadata": {},
   "source": [
    "#### 3.2.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85bd3ffa-40d7-4414-b9fc-0a9e577f6a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"starmap_plus_mouse_cns_batch1\"\n",
    "cell_type_key = \"Main_molecular_cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b1548b5-4dca-4762-b2b1-3b70211ed57f",
   "metadata": {},
   "source": [
    "#### 3.2.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5242ada-198f-46a6-b545-048ab451c42e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"starmap_plus_mouse_cns_subsample_{subsample_pct}pct_batch1\"\n",
    "    cell_type_key = \"Main_molecular_cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4111eee6-88e7-4180-bc5b-a05cdde41061",
   "metadata": {},
   "source": [
    "### 3.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcce8ca8-124d-4814-a224-900bd546165e",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.3.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7842f6f0-664b-4405-ad30-add0af93ec7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59d54d8a-bea3-40a9-acfa-4e78d2cffffe",
   "metadata": {},
   "source": [
    "#### 3.3.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e7b792d-e60d-4a14-a971-7aca558dca0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c88c8e4-3f52-4ac4-8e9b-8ed601511180",
   "metadata": {},
   "source": [
    "### 3.4 Vizgen MERFISH Mouse Liver"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91da577-1daa-484b-809f-4c7c080aa129",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.4.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d4f57ce-ed16-4994-a6dd-6dafe87a719f",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d826697c-54bd-44ad-a33a-2fe56af946db",
   "metadata": {},
   "source": [
    "#### 3.4.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faa871f9-7357-4d38-a931-09a67728e0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c39bdcc-6c2f-45ae-bc75-305db782e1e2",
   "metadata": {},
   "source": [
    "### 3.5 Slide-seqV2 Mouse Hippocampus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "278ca85b-f667-460d-b54f-98101eb221d4",
   "metadata": {},
   "source": [
    "#### 3.5.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1f0bccd-24db-4252-933a-fffb5e17c618",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"slideseqv2_mouse_hippocampus\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ed715aa-9f15-445b-a37a-4ea469ca91b2",
   "metadata": {},
   "source": [
    "#### 3.5.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c94ac68-2133-41d2-913f-94438f7e26f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"slideseqv2_mouse_hippocampus_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "622f6c1d-65a1-4318-b03c-c12444fd0f4c",
   "metadata": {},
   "source": [
    "## 4. NicheCompass Sample Integration Method Benchmarking Models Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd147edf-4ccd-402d-aca9-ce19cbd1b34f",
   "metadata": {},
   "source": [
    "### 4.1 seqFISH Mouse Organogenesis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b865230b-ab67-4474-a29c-dcb1d2294340",
   "metadata": {},
   "source": [
    "#### 4.1.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f764998-8726-419f-bafe-d627f75a1257",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_embeds_nums = \"6\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dac84fe1-0698-4c04-868f-2dd6e7d73356",
   "metadata": {},
   "source": [
    "#### 4.1.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63f15cc6-4e6d-4473-a126-dccb02b889a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-norm\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_embeds_nums = \"6\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b334564-7198-4344-8e8b-b20897765f3c",
   "metadata": {},
   "source": [
    "### 4.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dfecebf-0b36-47f9-a3a4-b59938733399",
   "metadata": {},
   "source": [
    "#### 4.2.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "679dc44d-14de-493b-8407-6bcf0debf032",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"Main_molecular_cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61a05506-ee44-4d31-a43a-63a82c40ad1f",
   "metadata": {},
   "source": [
    "#### 4.2.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92069eca-64a0-4ab1-be3d-a5b0059145be",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"starmap_plus_mouse_cns_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"Main_molecular_cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-norm\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ab268b-d86a-4843-9623-6f687c4f0a11",
   "metadata": {},
   "source": [
    "### 4.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfda56b3-cb23-4b88-95f4-b3d8e8651ec2",
   "metadata": {},
   "source": [
    "#### 4.3.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79414d8b-71c9-41f8-9cbe-921a94589c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_embeds_nums = \"3 10\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c112831f-87a4-41f8-9478-02f71d678a3f",
   "metadata": {},
   "source": [
    "#### 4.3.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7c7fe02-1889-4fe4-99db-d7602bf87a3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    node_label_method = \"one-hop-norm\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_embeds_nums = \"3 10\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "828f673a-9f55-4b43-9ff2-ca286d8ed45a",
   "metadata": {},
   "source": [
    "## 5. NicheCompass Reference Model Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56180669-13e7-4873-8887-5804360c2b75",
   "metadata": {},
   "source": [
    "### 5.1 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e6bed3a-6613-4101-81e5-9c19bb217b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "n_neighbors = 12\n",
    "n_hvg = 4000\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"4096\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6d78ebff-8b17-47c7-b126-8f7c350cc350",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12420065\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 4\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "n_neighbors = 12\n",
    "n_hvg = 4000\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"4096\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 30.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aee0b50-4cfa-49f3-84b4-134a62b061b4",
   "metadata": {},
   "source": [
    "### 5.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11bdc0b5-5a82-4786-b53f-2e13cbe4f42c",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\" \\\n",
    "                    \" batch9 batch10 batch11 batch12 batch13 batch14 batch15 batch16 batch17 batch18\" \\\n",
    "                    \" batch19 batch20\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 8, 12\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 2048 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df29aa7c-fc1a-4de9-88a9-893dc54fb097",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 3\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\" \\\n",
    "                    \" batch9 batch10 batch11 batch12 batch13 batch14 batch15 batch16 batch17 batch18\" \\\n",
    "                    \" batch19 batch20\"\n",
    "n_neighbors = 12 # 8, 12\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "n_cond_embed = 20 # 3, 20, None\n",
    "lambda_cat_covariates_contrastive = 250000. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0.03125 # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --n_cond_embed {n_cond_embed}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fde3a4b-fd6f-43ba-9f37-436885f328df",
   "metadata": {},
   "source": [
    "### 5.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "463fcdec-a4d3-4d97-8f6c-6d4a9251158b",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "216ebd95-3479-4f63-bbdd-cb192d06a347",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"fov\"\n",
    "cat_covariates_no_edges = \"False\"\n",
    "cat_covariates_embeds_nums = \"10\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cc78e13-e77e-4234-acab-22ab7ec5bdfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 10\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 10\"\n",
    "lambda_cat_covariates_contrastive = 100000.\n",
    "contrastive_logits_pos_ratio = 0.0625\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a420a844-aa9e-4f9b-8f17-9c46a5fd921c",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 11\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = None\n",
    "cat_covariates_no_edges = None\n",
    "cat_covariates_embeds_nums = None\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6612db72-e027-4302-8518-7ff94a2493f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 15\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 512 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"patient batch fov\"\n",
    "cat_covariates_embeds_nums = \"5 8 20\"\n",
    "lambda_cat_covariates_contrastive = 250000.\n",
    "contrastive_logits_pos_ratio = 0.0625\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7209ec-5334-4239-a142-d44ae5971bd9",
   "metadata": {},
   "source": [
    "### 5.4 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "671d5295-2e42-4365-811d-dcbaab2b0b20",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_neighbors = 4 # 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 4096 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"patient batch\"\n",
    "cat_covariates_no_edges = \"True True\"\n",
    "cat_covariates_embeds_nums = \"2 4\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c96dfb0-6b04-4bed-ae41-5af7591090a8",
   "metadata": {},
   "source": [
    "### 5.4 Vizgen MERFISH Human Lung Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16844a7c-93da-4f56-bb1f-1cd17701d9b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"vizgen_merfish_human_lung_cancer\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 4 # 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 4096 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"2\"\n",
    "lambda_cat_covariates_contrastive = 250000. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0.125 # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7380bdd-79ba-48a7-aeaa-ba20b8978edd",
   "metadata": {},
   "source": [
    "### 5.5 Spatial ATAC-RNA-Seq Mouse Embryo & Brain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cde2a39-e1d1-49c2-a429-6e11eb8f84ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"spatial_atac_rna_seq_mouse_brain\"\n",
    "reference_batches = \"None\"\n",
    "n_neighbors = 12\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"4096\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              \" --n_hvg 3000\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --mebocost_species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              \" --include_atac_modality\" \\\n",
    "              \" --filter_peaks\" \\\n",
    "              \" --min_cell_peak_thresh_ratio 0.0005\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              \" --lambda_chrom_access_recon 100.\" \\\n",
    "              \" --lambda_cat_covariates_contrastive 0.\" \\\n",
    "              \" --contrastive_logits_ratio 0.\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              \" --lambda_l1_masked 5.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01341dcb-64ad-45bd-8ef6-939e77d5b0f8",
   "metadata": {},
   "source": [
    "### 5.6 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68c5b999-11e0-42e5-b886-8fb3a47d82f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"xenium_human_breast_cancer\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 4 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-attention\"\n",
    "edge_batch_size = 4096 # 4096 (4, 8), 1024 (12)\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"2\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0. # 0., 5., 10.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "001945c5-6d59-4bdb-9efe-d8bbb33edc33",
   "metadata": {},
   "source": [
    "## 6. NicheCompass Reference Query Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14bf6b4d-34b5-4324-8deb-095bdfa7d0e2",
   "metadata": {},
   "source": [
    "### 6.1 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a6a826d-f6ad-4ee5-85b2-66a3670a574e",
   "metadata": {},
   "source": [
    "#### 6.1.1 Reference Model Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cdc61c79-6652-49fb-8bef-71ba0dda3b81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12320752\n"
     ]
    }
   ],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_neighbors = 12\n",
    "n_hvg = 3000\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 4096\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_reference_only\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7db14469-4ee2-4be2-b00d-535bc0439620",
   "metadata": {},
   "source": [
    "#### 6.1.2 Query Mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "17761bc0-3178-42dd-9b0e-3935be8d2a5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12323212\n"
     ]
    }
   ],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 1\n",
    "query_batches = \"batch5 batch6\"\n",
    "n_neighbors = 12\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"01072023_165203_1\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "edge_batch_size = 4096\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label {node_label_method}_{task}_query_only\" \\\n",
    "              f\" --reference_query_model_label {node_label_method}_{task}_query_mapping\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5391f85b-21e3-4723-bcef-6f7b0d948b4c",
   "metadata": {},
   "source": [
    "### 6.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "210672eb-c6ef-4cf5-9953-c2e4fcdec335",
   "metadata": {},
   "source": [
    "#### 6.2.1 Reference Model Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "90d450f9-673b-42ba-b11d-03fcd295693d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12428141\n"
     ]
    }
   ],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 4\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 8 # 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 250000. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0.0625 # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "149598af-5167-4466-b1e2-89de2f7375e2",
   "metadata": {},
   "source": [
    "#### 6.2.2 Query Mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a73f7984-551c-49f4-90f6-47eecf5092d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12422983\n"
     ]
    }
   ],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 1\n",
    "query_batches = \"batch3\"\n",
    "n_neighbors = 12\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"06072023_114143_1\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "edge_batch_size = 1024\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label {node_label_method}_{task}_query_only\" \\\n",
    "              f\" --reference_query_model_label {node_label_method}_{task}_query_mapping\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da211ed3-d9d9-4e55-a008-a33c0ca4f34a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
