{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bdc8086b-2b80-43b4-b88a-aa0c87c820a6",
   "metadata": {},
   "source": [
    "# Slurm Job Submission"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be88d0ca-4c09-44e3-83fc-fb6bcf4563fd",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>).\n",
    "- **Affiliation:** Helmholtz Munich, Institute of Computational Biology (ICB), Talavera-LÃ³pez Lab\n",
    "- **Date of Creation:** 20.03.2023\n",
    "- **Date of Last Modification:** 28.07.2023"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aeeeb1fe-eaa1-4cdd-aa21-ed78e0b7be81",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e58b0a3-c3fe-43d5-a420-14111aa545f6",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "48d297b0-e0dc-42b9-a349-e6ac2edba0e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aab82f4c-edd8-47dd-be56-559cef739439",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0491ee0e-3587-4243-856f-e7f71e6dced6",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda_env_name = \"nichecompass\"\n",
    "reproducibility_conda_env_name = \"nichecompass-reproducibility\"\n",
    "reproducibility_conda_env_name = \"nichecompass-test\"\n",
    "conda_env_name = \"nichecompass-test\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "004ab52a-9cc3-4e40-b0e6-f9fed7f4a549",
   "metadata": {},
   "source": [
    "### 1.3 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6b2d6cf0-b360-428f-9548-c7942378de68",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_python_script(\n",
    "        job_name_prefix,\n",
    "        job_id,\n",
    "        job_folder_path,\n",
    "        conda_env_name,\n",
    "        script_folder_path,\n",
    "        script_name,\n",
    "        script_args,\n",
    "        t=\"48:00:00\",\n",
    "        p=\"gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"gpu\",\n",
    "        nice=10000):\n",
    "    job_name = f\"{job_name_prefix}_{job_id}\"\n",
    "    # Account for fact that submit node has different home path than compute node\n",
    "    job_file_path = f\"{job_folder_path.replace('/aih', '')}/job_{job_name}.cmd\"\n",
    "    out_file_path = f\"{job_folder_path}/logs/out_{job_name}.txt\"\n",
    "    err_file_path = f\"{job_folder_path}/logs/err_{job_name}.txt\"\n",
    "    \n",
    "    os.makedirs(job_folder_path + \"/logs\", exist_ok=True)\n",
    "    \n",
    "    with open(job_file_path, \"w\") as handle:\n",
    "        handle.writelines(\"#!/bin/bash\\n\")\n",
    "        handle.writelines(f\"#SBATCH -J {job_name}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -o {out_file_path}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -e {err_file_path}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -t {t}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -p {p}\\n\")\n",
    "        handle.writelines(\"#SBATCH -c 6\\n\")\n",
    "        handle.writelines(f\"#SBATCH --gres={gres}\\n\")\n",
    "        handle.writelines(f\"#SBATCH --qos={qos}\\n\")       \n",
    "        handle.writelines(\"#SBATCH --mem=128GB\\n\")\n",
    "        handle.writelines(f\"#SBATCH --nice={nice}\\n\")\n",
    "        handle.writelines(\"source $HOME/.bashrc\\n\")\n",
    "        handle.writelines(f\"conda activate {conda_env_name}\\n\")\n",
    "        handle.writelines(\"cd /\\n\")\n",
    "        handle.writelines(f\"cd {script_folder_path}\\n\")\n",
    "        handle.writelines(f\"python ../{script_name}\")\n",
    "        handle.writelines(f\"{script_args}\")\n",
    "        handle.writelines(\"\\n\")\n",
    "        \n",
    "    os.system(f\"sbatch {job_file_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4624e92-33eb-457c-8767-9006e44d850e",
   "metadata": {},
   "source": [
    "## 2. NicheCompass Ablation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc99f810-89a4-4070-ac09-d3f6da04c95d",
   "metadata": {},
   "source": [
    "### 2.1 Loss Weights, Gene Program Mask"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b4617a3-879b-4591-95d9-76576cc0031f",
   "metadata": {},
   "source": [
    "Ablating:\n",
    "- Loss weights: Gene Expression Reconstruction, Edge Reconstruction, KL Divergence\n",
    "- Gene program Mask: FC GP Mask, NicheCompass Default GP Mask (Not Filtered, Weakly Filtered, Strongly Filtered)\n",
    "\n",
    "Additional covariate:\n",
    "- Spatial Neighborhood Graph: Number of Neighbors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e95996-f6e0-459c-a400-5a2ae3193f47",
   "metadata": {},
   "source": [
    "#### 2.1.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84cfe23e-ec00-4c8d-af02-2aa210997565",
   "metadata": {},
   "source": [
    "##### 2.1.1.1 Fully Connected GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0621ca71-70a3-423e-9f79-91e706f5d4e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22f136c4-a8a6-4907-962d-3bc8a2282b23",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 37\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc62e692-6c03-42e6-9fd8-c6bdde35d409",
   "metadata": {},
   "source": [
    "##### 2.1.1.2 NicheCompass GP Mask (Keep All Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "48a49459-8035-4d3e-82bb-9e594f12533b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12662836\n",
      "Submitted batch job 12662837\n",
      "Submitted batch job 12662838\n",
      "Submitted batch job 12662839\n",
      "Submitted batch job 12662840\n",
      "Submitted batch job 12662841\n",
      "Submitted batch job 12662842\n",
      "Submitted batch job 12662843\n",
      "Submitted batch job 12662844\n",
      "Submitted batch job 12662845\n",
      "Submitted batch job 12662846\n",
      "Submitted batch job 12662847\n",
      "Submitted batch job 12662848\n",
      "Submitted batch job 12662849\n",
      "Submitted batch job 12662850\n",
      "Submitted batch job 12662851\n",
      "Submitted batch job 12662852\n",
      "Submitted batch job 12662853\n",
      "Submitted batch job 12662854\n",
      "Submitted batch job 12662855\n",
      "Submitted batch job 12662856\n",
      "Submitted batch job 12662857\n",
      "Submitted batch job 12662858\n",
      "Submitted batch job 12662859\n",
      "Submitted batch job 12662860\n",
      "Submitted batch job 12662861\n",
      "Submitted batch job 12662862\n",
      "Submitted batch job 12662863\n",
      "Submitted batch job 12662864\n",
      "Submitted batch job 12662865\n",
      "Submitted batch job 12662866\n",
      "Submitted batch job 12662867\n",
      "Submitted batch job 12662868\n",
      "Submitted batch job 12662869\n",
      "Submitted batch job 12662870\n",
      "Submitted batch job 12662871\n"
     ]
    }
   ],
   "source": [
    "job_id = 53\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a7c6489-5007-46b9-9d8f-650e3125e292",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 89\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c46cace5-f7e8-4c76-bff1-3619ba15242c",
   "metadata": {},
   "source": [
    "##### 2.1.1.3 NicheCompass GP Mask (Weakly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a3fbb52-b4a6-4f1b-8361-fa068d7aa16a",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 105\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03f0601a-62c5-485d-868c-738c0b6178fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 141\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea24b9d8-45a1-4909-80c3-a2d96da1a9c1",
   "metadata": {},
   "source": [
    "##### 2.1.1.4 NicheCompass GP Mask (Strongly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4013b3f-cc14-41e6-bf75-fd3809199921",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 157\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "965dbda4-065d-4ba4-a940-5cbd61166c43",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 193\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f3fbdd9-c303-4781-91e6-9485dabf8e33",
   "metadata": {},
   "source": [
    "#### 2.1.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75e888c2-82b8-4d9f-b5e7-ac0d371bb0da",
   "metadata": {},
   "source": [
    "##### 2.1.2.1 Fully Connected GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "333a8e02-74cc-474e-9baa-49b02d348617",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "038d37a5-46a9-4c7f-9d19-fe33a0398c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 37\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f29be6dc-ee82-43bf-9a07-c8b4cacab09b",
   "metadata": {},
   "source": [
    "##### 2.1.2.2 NicheCompass GP Mask (Keep All Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a41582a1-885f-400e-90d0-70cc47823eaa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12666140\n"
     ]
    }
   ],
   "source": [
    "# tmp\n",
    "job_id = 74\n",
    "for lambda_edge_recon in [50000]:\n",
    "    for lambda_gene_expr_recon in [300]:\n",
    "        for n_neighbors in [8]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0602953c-56ca-4dd9-b57b-79aa3c5a97ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12666141\n"
     ]
    }
   ],
   "source": [
    "# tmp\n",
    "job_id = 83\n",
    "for lambda_edge_recon in [500000]:\n",
    "    for lambda_gene_expr_recon in [30]:\n",
    "        for n_neighbors in [12]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1871d10d-83fc-4b99-b143-966dfd90572a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12666456\n"
     ]
    }
   ],
   "source": [
    "# tmp\n",
    "job_id = 84\n",
    "for lambda_edge_recon in [500000]:\n",
    "    for lambda_gene_expr_recon in [30]:\n",
    "        for n_neighbors in [16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4d2d5bd1-db7f-4f0a-ae36-bc135dfae0eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12666548\n"
     ]
    }
   ],
   "source": [
    "# tmp\n",
    "job_id = 85\n",
    "for lambda_edge_recon in [500000]:\n",
    "    for lambda_gene_expr_recon in [300]:\n",
    "        for n_neighbors in [4]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "790911e2-ce3f-45cc-8e3d-3789cf1661e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12661274\n",
      "Submitted batch job 12661275\n",
      "Submitted batch job 12661276\n",
      "Submitted batch job 12661277\n",
      "Submitted batch job 12661278\n",
      "Submitted batch job 12661279\n",
      "Submitted batch job 12661280\n",
      "Submitted batch job 12661281\n",
      "Submitted batch job 12661282\n",
      "Submitted batch job 12661283\n",
      "Submitted batch job 12661284\n",
      "Submitted batch job 12661285\n",
      "Submitted batch job 12661286\n",
      "Submitted batch job 12661287\n",
      "Submitted batch job 12661288\n",
      "Submitted batch job 12661289\n",
      "Submitted batch job 12661290\n",
      "Submitted batch job 12661291\n",
      "Submitted batch job 12661292\n",
      "Submitted batch job 12661293\n",
      "Submitted batch job 12661294\n",
      "Submitted batch job 12661295\n",
      "Submitted batch job 12661296\n",
      "Submitted batch job 12661297\n",
      "Submitted batch job 12661298\n",
      "Submitted batch job 12661299\n",
      "Submitted batch job 12661300\n",
      "Submitted batch job 12661301\n",
      "Submitted batch job 12661302\n",
      "Submitted batch job 12661303\n",
      "Submitted batch job 12661304\n",
      "Submitted batch job 12661305\n",
      "Submitted batch job 12661306\n",
      "Submitted batch job 12661307\n",
      "Submitted batch job 12661308\n",
      "Submitted batch job 12661309\n"
     ]
    }
   ],
   "source": [
    "job_id = 53\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4e9286b-b22c-441f-b3a4-25b9050eae73",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 89\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7f18840-64b2-44dc-9cb8-e851ca3ddb27",
   "metadata": {},
   "source": [
    "##### 2.1.2.3 NicheCompass GP Mask (Weakly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50f5c146-9003-42a3-a434-b4bb9a34fe97",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 105\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63ccf6d1-2cf6-4b63-9f99-97dd28d7974f",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 141\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0744941f-31c5-40fe-9835-31f36cff1e70",
   "metadata": {},
   "source": [
    "##### 2.1.2.4 NicheCompass GP Mask (Strongly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "962c9e96-ffed-4572-9b95-3d254ff285fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 157\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcbb1a0b-ae95-4e28-adfa-8620e93aac72",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 193\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cb42f90-41a3-43e6-8747-18d91b527a92",
   "metadata": {},
   "source": [
    "#### 2.1.3 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5558252c-7c8a-4755-997e-2553e7af9d5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 100006\n",
    "for lambda_edge_recon in [50000000]:\n",
    "    for lambda_gene_expr_recon in [3000]:\n",
    "        for n_neighbors in [16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "            n_sampled_neighbors = -1\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 79\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e989cecb-0197-490e-8d11-5348bbdeffbc",
   "metadata": {},
   "source": [
    "##### 2.1.3.1 Fully Connected GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a9d1e2b-a939-4b6f-bdee-87f767743618",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ce2d525-2c68-4816-bbea-e6ca62309774",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 37\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ff6fd10-e19a-4d55-ab17-ae11a82665a3",
   "metadata": {},
   "source": [
    "##### 2.1.3.2 NicheCompass GP Mask (Keep All Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4125af6c-9aed-40a2-a5ad-ac00ccae7e72",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12663074\n",
      "Submitted batch job 12663075\n",
      "Submitted batch job 12663076\n",
      "Submitted batch job 12663077\n",
      "Submitted batch job 12663078\n",
      "Submitted batch job 12663079\n",
      "Submitted batch job 12663080\n",
      "Submitted batch job 12663081\n",
      "Submitted batch job 12663082\n",
      "Submitted batch job 12663083\n",
      "Submitted batch job 12663084\n",
      "Submitted batch job 12663085\n",
      "Submitted batch job 12663086\n",
      "Submitted batch job 12663087\n",
      "Submitted batch job 12663088\n",
      "Submitted batch job 12663089\n",
      "Submitted batch job 12663090\n",
      "Submitted batch job 12663091\n",
      "Submitted batch job 12663092\n",
      "Submitted batch job 12663093\n",
      "Submitted batch job 12663094\n",
      "Submitted batch job 12663095\n",
      "Submitted batch job 12663096\n",
      "Submitted batch job 12663097\n",
      "Submitted batch job 12663098\n",
      "Submitted batch job 12663099\n",
      "Submitted batch job 12663100\n",
      "Submitted batch job 12663101\n",
      "Submitted batch job 12663102\n",
      "Submitted batch job 12663103\n",
      "Submitted batch job 12663104\n",
      "Submitted batch job 12663105\n",
      "Submitted batch job 12663106\n",
      "Submitted batch job 12663107\n",
      "Submitted batch job 12663108\n",
      "Submitted batch job 12663109\n"
     ]
    }
   ],
   "source": [
    "job_id = 53\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9dc3f3b-b7c1-485c-8291-e0e2ad4c25ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# missing\n",
    "job_id = 89\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "375a4972-0b37-451e-bfbc-b152bcf45412",
   "metadata": {},
   "source": [
    "##### 2.1.3.3 NicheCompass GP Mask (Weakly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5710e07-be54-4c0d-a76f-55ee625501c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 105\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "603dd800-bf66-4810-8aed-aa74c1e8a8c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 141\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e47e0e96-832b-4d9f-a1ee-56761950904e",
   "metadata": {},
   "source": [
    "##### 2.1.3.4 NicheCompass GP Mask (Strongly Filter Target Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ec32310-cb4f-40df-8036-b081d85874a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 157\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7f8c671-093d-4ed1-85c8-4c93aef4f9f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 193\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ab4f058-e6e7-45d5-bce1-9432c1279590",
   "metadata": {},
   "source": [
    "#### 2.1.4 Metrics Computation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4c49be3-1031-427f-bb45-3245c60e98b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 1\n",
    "for i in np.arange(1, 13, 1):\n",
    "    task = \"one-hop-norm_reference\"\n",
    "    dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"celltype_mapped_refined\"\n",
    "    batch_keys = \"batch\"\n",
    "    file_name = f\"mlflow_summary_loss_weights_ablation_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw bgc bilisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_loss_weights_ablation_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/ablation/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/ablation\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33a9fd4a-d69f-400d-8b1c-8c43e53f256f",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(20, 240, 20):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"xenium_human_breast_cancer\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"cell_states\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e83cd203-633c-409d-9e66-ae6ef47625f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12667108\n",
      "Submitted batch job 12667109\n",
      "Submitted batch job 12667110\n",
      "Submitted batch job 12667111\n",
      "Submitted batch job 12667112\n",
      "Submitted batch job 12667113\n",
      "Submitted batch job 12667114\n",
      "Submitted batch job 12667115\n",
      "Submitted batch job 12667116\n"
     ]
    }
   ],
   "source": [
    "# tmp\n",
    "job_id = 1\n",
    "for i in np.arange(4, 40, 4):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"xenium_human_breast_cancer\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"cell_states\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5d4d246-f0c6-4186-ab41-599ce8a4bb57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 1\n",
    "for i in np.arange(20, 240, 20):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"xenium_human_breast_cancer\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"cell_states\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4104bb9b-bf59-4ec6-b1f3-d43aa0e578ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12667077\n",
      "Submitted batch job 12667078\n",
      "Submitted batch job 12667079\n",
      "Submitted batch job 12667080\n",
      "Submitted batch job 12667081\n",
      "Submitted batch job 12667082\n",
      "Submitted batch job 12667083\n",
      "Submitted batch job 12667084\n"
     ]
    }
   ],
   "source": [
    "# tmp\n",
    "job_id = 1\n",
    "for i in np.arange(4, 36, 4):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"starmap_plus_mouse_cns\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"Main_molecular_cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "198ebdc9-e759-438b-a624-51e4d49ac5aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(20, 240, 20):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"starmap_plus_mouse_cns\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"Main_molecular_cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b72aa3ff-636d-49a7-b406-47b4a3d14745",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12667085\n",
      "Submitted batch job 12667086\n",
      "Submitted batch job 12667087\n",
      "Submitted batch job 12667088\n",
      "Submitted batch job 12667089\n",
      "Submitted batch job 12667090\n",
      "Submitted batch job 12667091\n",
      "Submitted batch job 12667092\n",
      "Submitted batch job 12667093\n"
     ]
    }
   ],
   "source": [
    "# tmp\n",
    "job_id = 1\n",
    "for i in np.arange(4, 40, 4):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db88fc23-3075-491e-bf2c-5fbdc4482a56",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(20, 240, 20):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ecdcbd2-4616-4175-b09f-3d853d51984d",
   "metadata": {},
   "source": [
    "### 2.2 Neighbor Sampling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "016fccd5-e4f7-4262-ad93-e2e27e848248",
   "metadata": {},
   "source": [
    "Best parameters from previous ablations:\n",
    "- lambda_edge_recon = 50,000,000\n",
    "- lambda_gene_expr_recon = 30,000\n",
    "- nichenet_keep_target_genes_ratio = 1.0\n",
    "\n",
    "Ablating:\n",
    "- Neighbor sampling: no sampling, 2 neighbors sampling, 4 neighbors sampling, 8 neighbors sampling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbecce69-187e-4bdc-8b6f-2a2c348ca4f8",
   "metadata": {},
   "source": [
    "#### 2.2.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22828726-54f6-4d3d-9f91-3a34d05b1944",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_neighbors in [4, 8, 12, 16]:\n",
    "    for n_sampled_neighbors in [2, 4, 8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"xenium_human_breast_cancer\"\n",
    "        reference_batches = \"batch1\"\n",
    "        species = \"human\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 128\n",
    "        lambda_edge_recon = 5000000\n",
    "        lambda_gene_expr_recon = 3000\n",
    "        nichenet_keep_target_genes_ratio = 1.0\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      f\" --nichenet_keep_target_genes_ratio {nichenet_keep_target_genes_ratio}\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6347069e-ea8f-4165-bd3b-63edf25a35a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 13\n",
    "for n_neighbors in [20, 24, 28, 32]:\n",
    "    for n_sampled_neighbors in [2, 4, 8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"xenium_human_breast_cancer\"\n",
    "        reference_batches = \"batch1\"\n",
    "        species = \"human\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 256\n",
    "        lambda_edge_recon = 5000000\n",
    "        lambda_gene_expr_recon = 3000\n",
    "        nichenet_keep_target_genes_ratio = 1.0\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      f\" --nichenet_keep_target_genes_ratio {nichenet_keep_target_genes_ratio}\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8b28b8e-943e-4636-80f4-2300aa11e01a",
   "metadata": {},
   "source": [
    "#### 2.2.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12566fe1-76c8-4034-bde2-b5440c7407cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_neighbors in [4, 8, 12, 16]:\n",
    "    for n_sampled_neighbors in [2, 4, 8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"starmap_plus_mouse_cns\"\n",
    "        reference_batches = \"batch1\"\n",
    "        species = \"mouse\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 1024\n",
    "        lambda_edge_recon = 500000\n",
    "        lambda_gene_expr_recon = 300\n",
    "        nichenet_keep_target_genes_ratio = 0.1\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      f\" --nichenet_keep_target_genes_ratio {nichenet_keep_target_genes_ratio}\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b3617d6-b2ab-4fa9-9c07-eb76538a88b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 13\n",
    "for n_neighbors in [20, 24, 28, 32]:\n",
    "    for n_sampled_neighbors in [2, 4, 8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"starmap_plus_mouse_cns\"\n",
    "        reference_batches = \"batch1\"\n",
    "        species = \"mouse\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 512\n",
    "        lambda_edge_recon = 50000000\n",
    "        lambda_gene_expr_recon = 30000\n",
    "        nichenet_keep_target_genes_ratio = 0.1\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      f\" --nichenet_keep_target_genes_ratio {nichenet_keep_target_genes_ratio}\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bd17731-1ce8-45a4-b02d-a0a92c56b895",
   "metadata": {},
   "source": [
    "#### 2.2.3 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e4e4b9b-20ef-42b8-b7e4-2279b4de9a78",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_neighbors in [4, 8, 12, 16]:\n",
    "    for n_sampled_neighbors in [2, 4, 8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "        reference_batches = \"batch2\"\n",
    "        species = \"human\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 512\n",
    "        lambda_edge_recon = \n",
    "        lambda_gene_expr_recon =\n",
    "        nichenet_keep_target_genes_ratio = 1.0\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      f\" --nichenet_keep_target_genes_ratio {nichenet_keep_target_genes_ratio}\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd07fca5-4076-4f65-9d9c-79109b886bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_neighbors in [20, 24, 28, 32]:\n",
    "    for n_sampled_neighbors in [2, 4, 8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "        reference_batches = \"batch2\"\n",
    "        species = \"human\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 256\n",
    "        lambda_edge_recon = \n",
    "        lambda_gene_expr_recon =\n",
    "        nichenet_keep_target_genes_ratio = 1.0\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      f\" --nichenet_keep_target_genes_ratio {nichenet_keep_target_genes_ratio}\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23fd2150-6666-445f-bada-f63bdc2ab9e6",
   "metadata": {},
   "source": [
    "#### 2.2.4 Metrics Computation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd34c575-3f01-44a1-8226-a5a49e805ff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "for job_id in range(1, 4):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"starmap_plus_mouse_cns\"\n",
    "    ablation_task = \"neighbor_sampling\"\n",
    "    cell_type_keys = \"Main_molecular_cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{job_id}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c34a6a35-d41b-4fbf-a595-4156efd1da0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"ablation\"\n",
    "ablation_task = \"neighbor_sampling\"\n",
    "datasets = \"starmap_plus_mouse_cns\"\n",
    "cell_type_keys = \"Main_molecular_cell_type\"\n",
    "batch_keys = \"None None None\"\n",
    "file_name = f\"mlflow_summary_{ablation_task}_{task}_processed.csv\"\n",
    "\n",
    "job_name_prefix = f\"nichecompass_{ablation_task}_{task}_metrics_computation\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"compute_metrics.py\"\n",
    "script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "              f\" --file_name {file_name}\" \\\n",
    "              f\" --datasets {datasets}\" \\\n",
    "              f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "              f\" --batch_keys {batch_keys}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=reproducibility_conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8b9a684-f40e-4cb3-a54e-936e6a7ed3bf",
   "metadata": {},
   "source": [
    "### 2.3 Encoder Architecture"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5501bd0a-337b-439b-9592-1abd7dd08c6e",
   "metadata": {},
   "source": [
    "Ablating:\n",
    "- Encoder Number of Message Passing Layers: 1, 2\n",
    "- Encoder Type of Message Passing Layers: GCNConv vs GATv2Conv (4 attention heads)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "544da563-d23d-4cfc-b288-e7e13f394f25",
   "metadata": {},
   "source": [
    "#### 2.3.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e36b5743-f2a0-4b98-b68d-ab1525b96c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"encoder_architecture\"\n",
    "                dataset = \"xenium_human_breast_cancer\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 128\n",
    "                lambda_edge_recon = 50000000.\n",
    "                lambda_gene_expr_recon = 3000.\n",
    "                n_sampled_neighbors = 4\n",
    "                \n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                              f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso 0.\" \\\n",
    "                              f\" --lambda_l1_masked 0.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7804d861-fc83-4e82-890d-3e61c32b16c5",
   "metadata": {},
   "source": [
    "#### 2.2.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7331e624-4c83-41a3-9a43-3dbeea67bb23",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"encoder_architecture\"\n",
    "                dataset = \"starmap_plus_mouse_cns\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"mouse\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 256\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 30.\n",
    "                n_sampled_neighbors = 4\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                              f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso 0.\" \\\n",
    "                              f\" --lambda_l1_masked 0.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe6c40e-d47c-40f2-ab49-194d977e0cd7",
   "metadata": {},
   "source": [
    "#### 2.2.3 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af6c24a4-cecc-4ba0-81e2-8dcf7da09995",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_neighbors in [16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"encoder_architecture\"\n",
    "                dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "                reference_batches = \"batch2\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 128\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                              f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso 0.\" \\\n",
    "                              f\" --lambda_l1_masked 0.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0354afb9-189f-4d5e-b52f-a1060dffa38c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c42823d5-0fd0-49ce-af0c-8cc314a0f098",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a022515e-a5c6-4011-9cc7-0eb40d8c48ab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8734b56d-9239-4960-9689-f80ff56bf927",
   "metadata": {},
   "source": [
    "### 2.3 Gene Program & Gene Selection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7757c17-695f-4a02-9f39-645513a96ec4",
   "metadata": {},
   "source": [
    "Ablating:\n",
    "- Gene Program Selection: Active GP Threshold Ratio, Group Lasso Regularization\n",
    "- Gene Selection: L1 Regularization\n",
    "- Spatial Neighborhood Graph: Number of Neighbors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "985a6ac0-5138-497a-817d-4aca2cf73df9",
   "metadata": {},
   "source": [
    "#### 2.3.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0184c63e-5e14-4663-8970-751c1918d2e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for active_gp_thresh_ratio in [0., 0.05, 0.1]:\n",
    "    for lambda_group_lasso in [0., 5., 10.]:\n",
    "        for lambda_l1_masked in [0., 5., 10., 50.]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"gene_program_gene_selection\"\n",
    "                dataset = \"xenium_human_breast_cancer\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 512\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              \" --n_layers_encoder 1\" \\\n",
    "                              \" --n_hidden_encoder None\" \\\n",
    "                              \" --conv_layer_encoder gcnconv\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba1b231d-c2ba-4628-975a-9c8793568012",
   "metadata": {},
   "source": [
    "#### 2.3.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c74da10f-1b0e-482f-9028-f1172312a902",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for active_gp_thresh_ratio in [0., 0.05, 0.1]:\n",
    "    for lambda_group_lasso in [0, 5, 10]:\n",
    "        for lambda_l1_masked in [0., 5., 10., 50., 100.]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"gene_program_gene_selection\"\n",
    "                dataset = \"starmap_plus_mouse_cns\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"mouse\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 1024\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              \" --n_layers_encoder 1\" \\\n",
    "                              \" --n_hidden_encoder None\" \\\n",
    "                              \" --conv_layer_encoder gcnconv\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked 0.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21d51e95-d594-4fa6-bdcb-c44001904a73",
   "metadata": {},
   "source": [
    "#### 2.3.3 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f96de33-56ea-48db-bd1f-6437da817d90",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for active_gp_thresh_ratio in [0., 0.05, 0.1]:\n",
    "    for lambda_group_lasso in [0, 5, 10]:\n",
    "        for lambda_l1_masked in [0., 5., 10., 50., 100.]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"gene_program_gene_selection\"\n",
    "                dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "                reference_batches = \"batch2\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 1024\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              \" --n_layers_encoder 1\" \\\n",
    "                              \" --n_hidden_encoder None\" \\\n",
    "                              \" --conv_layer_encoder gcnconv\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked 0.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29ad5d87-e074-423f-b9c5-43a0405ed227",
   "metadata": {},
   "source": [
    "### 2.5 Categorical Covariate Embedding & Contrastive Integration"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1e8207b-2b89-4fb1-89b0-c32cc207a8f3",
   "metadata": {},
   "source": [
    "#### 2.5.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d9ebf56-a5a1-4b64-b341-1108d269bbd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for cat_covariates_embeds_injection in [\"encoder\", \"gene_expr_decoder\"]:\n",
    "for lambda_cat_covariates_contrastive in [0., 1, 2]:\n",
    "    for contrastive_logits_pos_ratio in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"encoder_architecture\"\n",
    "                dataset = \"xenium_human_breast_cancer\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 512\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "                active_gp_thresh_ratio = 0.05\n",
    "                lambda_group_lasso = 0.\n",
    "                n_layers_encoder = 1\n",
    "                n_hidden_encoder = None\n",
    "                conv_layer_encoder = \"gcnconv\"\n",
    "                \n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                              f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              f\" --n_epochs_no_cat_covariates_contrastive {n_epochs_no_cat_covariates_contrastive}\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cc7c954-2df9-4224-9e57-7e6be8072b30",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c39d3f6-fe11-4056-bc48-9ef421bc9684",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a80b724d-8200-44b4-8dba-02269e4358e0",
   "metadata": {},
   "source": [
    "## 3. NicheCompass Single Sample Method Benchmarking Models Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eb67595-8ede-4823-a1fe-a313dee704f5",
   "metadata": {},
   "source": [
    "### 3.1 seqFISH Mouse Organogenesis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b52c28e8-6fc8-492f-9c31-d3ab9b815980",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.1.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe70711-9d03-47d5-892a-5cdf77714389",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_embryo2\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24ecebdb-2b38-4021-8116-b21061bef3b1",
   "metadata": {},
   "source": [
    "#### 3.1.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c3995e5-3c35-427e-838c-4831370d44c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct_embryo2\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19574d56-9c79-4874-9b4d-ee39248f0028",
   "metadata": {},
   "source": [
    "### 3.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe560799-9d98-4740-af7a-b972c7596d18",
   "metadata": {},
   "source": [
    "#### 3.2.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85bd3ffa-40d7-4414-b9fc-0a9e577f6a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"starmap_plus_mouse_cns_batch1\"\n",
    "cell_type_key = \"Main_molecular_cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b1548b5-4dca-4762-b2b1-3b70211ed57f",
   "metadata": {},
   "source": [
    "#### 3.2.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5242ada-198f-46a6-b545-048ab451c42e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"starmap_plus_mouse_cns_subsample_{subsample_pct}pct_batch1\"\n",
    "    cell_type_key = \"Main_molecular_cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4111eee6-88e7-4180-bc5b-a05cdde41061",
   "metadata": {},
   "source": [
    "### 3.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcce8ca8-124d-4814-a224-900bd546165e",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.3.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7842f6f0-664b-4405-ad30-add0af93ec7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59d54d8a-bea3-40a9-acfa-4e78d2cffffe",
   "metadata": {},
   "source": [
    "#### 3.3.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e7b792d-e60d-4a14-a971-7aca558dca0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c88c8e4-3f52-4ac4-8e9b-8ed601511180",
   "metadata": {},
   "source": [
    "### 3.4 Vizgen MERFISH Mouse Liver"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91da577-1daa-484b-809f-4c7c080aa129",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.4.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d4f57ce-ed16-4994-a6dd-6dafe87a719f",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d826697c-54bd-44ad-a33a-2fe56af946db",
   "metadata": {},
   "source": [
    "#### 3.4.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faa871f9-7357-4d38-a931-09a67728e0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c39bdcc-6c2f-45ae-bc75-305db782e1e2",
   "metadata": {},
   "source": [
    "### 3.5 Slide-seqV2 Mouse Hippocampus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "278ca85b-f667-460d-b54f-98101eb221d4",
   "metadata": {},
   "source": [
    "#### 3.5.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1f0bccd-24db-4252-933a-fffb5e17c618",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"slideseqv2_mouse_hippocampus\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ed715aa-9f15-445b-a37a-4ea469ca91b2",
   "metadata": {},
   "source": [
    "#### 3.5.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c94ac68-2133-41d2-913f-94438f7e26f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"slideseqv2_mouse_hippocampus_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "622f6c1d-65a1-4318-b03c-c12444fd0f4c",
   "metadata": {},
   "source": [
    "## 4. NicheCompass Sample Integration Method Benchmarking Models Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd147edf-4ccd-402d-aca9-ce19cbd1b34f",
   "metadata": {},
   "source": [
    "### 4.1 seqFISH Mouse Organogenesis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b865230b-ab67-4474-a29c-dcb1d2294340",
   "metadata": {},
   "source": [
    "#### 4.1.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f764998-8726-419f-bafe-d627f75a1257",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_embeds_nums = \"6\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dac84fe1-0698-4c04-868f-2dd6e7d73356",
   "metadata": {},
   "source": [
    "#### 4.1.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63f15cc6-4e6d-4473-a126-dccb02b889a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-norm\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_embeds_nums = \"6\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b334564-7198-4344-8e8b-b20897765f3c",
   "metadata": {},
   "source": [
    "### 4.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dfecebf-0b36-47f9-a3a4-b59938733399",
   "metadata": {},
   "source": [
    "#### 4.2.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "679dc44d-14de-493b-8407-6bcf0debf032",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"Main_molecular_cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61a05506-ee44-4d31-a43a-63a82c40ad1f",
   "metadata": {},
   "source": [
    "#### 4.2.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92069eca-64a0-4ab1-be3d-a5b0059145be",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"starmap_plus_mouse_cns_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"Main_molecular_cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-norm\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ab268b-d86a-4843-9623-6f687c4f0a11",
   "metadata": {},
   "source": [
    "### 4.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfda56b3-cb23-4b88-95f4-b3d8e8651ec2",
   "metadata": {},
   "source": [
    "#### 4.3.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79414d8b-71c9-41f8-9cbe-921a94589c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_embeds_nums = \"3 10\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c112831f-87a4-41f8-9478-02f71d678a3f",
   "metadata": {},
   "source": [
    "#### 4.3.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7c7fe02-1889-4fe4-99db-d7602bf87a3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    node_label_method = \"one-hop-norm\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_embeds_nums = \"3 10\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "828f673a-9f55-4b43-9ff2-ca286d8ed45a",
   "metadata": {},
   "source": [
    "## 5. NicheCompass Reference Model Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56180669-13e7-4873-8887-5804360c2b75",
   "metadata": {},
   "source": [
    "### 5.1 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8e6bed3a-6613-4101-81e5-9c19bb217b00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12669884\n",
      "Submitted batch job 12669885\n",
      "Submitted batch job 12669886\n",
      "Submitted batch job 12669887\n",
      "Submitted batch job 12669888\n",
      "Submitted batch job 12669889\n",
      "Submitted batch job 12669890\n",
      "Submitted batch job 12669891\n",
      "Submitted batch job 12669892\n",
      "Submitted batch job 12669893\n",
      "Submitted batch job 12669894\n",
      "Submitted batch job 12669895\n",
      "Submitted batch job 12669896\n",
      "Submitted batch job 12669897\n",
      "Submitted batch job 12669898\n",
      "Submitted batch job 12669899\n",
      "Submitted batch job 12669900\n",
      "Submitted batch job 12669901\n",
      "Submitted batch job 12669902\n",
      "Submitted batch job 12669903\n",
      "Submitted batch job 12669904\n",
      "Submitted batch job 12669905\n",
      "Submitted batch job 12669906\n",
      "Submitted batch job 12669907\n",
      "Submitted batch job 12669908\n",
      "Submitted batch job 12669909\n",
      "Submitted batch job 12669910\n",
      "Submitted batch job 12669911\n",
      "Submitted batch job 12669912\n",
      "Submitted batch job 12669913\n",
      "Submitted batch job 12669914\n",
      "Submitted batch job 12669915\n",
      "Submitted batch job 12669916\n",
      "Submitted batch job 12669917\n",
      "Submitted batch job 12669918\n",
      "Submitted batch job 12669919\n",
      "Submitted batch job 12669920\n",
      "Submitted batch job 12669921\n",
      "Submitted batch job 12669922\n",
      "Submitted batch job 12669923\n",
      "Submitted batch job 12669924\n",
      "Submitted batch job 12669925\n",
      "Submitted batch job 12669926\n",
      "Submitted batch job 12669927\n",
      "Submitted batch job 12669928\n",
      "Submitted batch job 12669929\n",
      "Submitted batch job 12669930\n",
      "Submitted batch job 12669931\n",
      "Submitted batch job 12669932\n",
      "Submitted batch job 12669933\n",
      "Submitted batch job 12669934\n",
      "Submitted batch job 12669935\n",
      "Submitted batch job 12669936\n",
      "Submitted batch job 12669937\n",
      "Submitted batch job 12669938\n",
      "Submitted batch job 12669939\n",
      "Submitted batch job 12669940\n",
      "Submitted batch job 12669941\n",
      "Submitted batch job 12669942\n",
      "Submitted batch job 12669943\n",
      "Submitted batch job 12669944\n",
      "Submitted batch job 12669945\n",
      "Submitted batch job 12669946\n",
      "Submitted batch job 12669947\n",
      "Submitted batch job 12669948\n",
      "Submitted batch job 12669949\n",
      "Submitted batch job 12669950\n",
      "Submitted batch job 12669951\n",
      "Submitted batch job 12669952\n",
      "Submitted batch job 12669953\n",
      "Submitted batch job 12669954\n",
      "Submitted batch job 12669955\n",
      "Submitted batch job 12669956\n",
      "Submitted batch job 12669957\n",
      "Submitted batch job 12669958\n",
      "Submitted batch job 12669959\n",
      "Submitted batch job 12669960\n",
      "Submitted batch job 12669961\n",
      "Submitted batch job 12669962\n",
      "Submitted batch job 12669963\n",
      "Submitted batch job 12669964\n",
      "Submitted batch job 12669965\n",
      "Submitted batch job 12669966\n",
      "Submitted batch job 12669967\n",
      "Submitted batch job 12669968\n",
      "Submitted batch job 12669969\n",
      "Submitted batch job 12669970\n",
      "Submitted batch job 12669971\n",
      "Submitted batch job 12669972\n",
      "Submitted batch job 12669973\n",
      "Submitted batch job 12669974\n",
      "Submitted batch job 12669975\n",
      "Submitted batch job 12669976\n",
      "Submitted batch job 12669977\n",
      "Submitted batch job 12669978\n",
      "Submitted batch job 12669979\n",
      "Submitted batch job 12669980\n",
      "Submitted batch job 12669981\n",
      "Submitted batch job 12669982\n",
      "Submitted batch job 12669983\n",
      "Submitted batch job 12669984\n",
      "Submitted batch job 12669985\n",
      "Submitted batch job 12669986\n",
      "Submitted batch job 12669987\n",
      "Submitted batch job 12669988\n",
      "Submitted batch job 12669989\n",
      "Submitted batch job 12669990\n",
      "Submitted batch job 12669991\n"
     ]
    }
   ],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [50000, 500000, 5000000]:\n",
    "    for lambda_gene_expr_recon in [30, 300, 3000]:\n",
    "            for active_gp_thresh_ratio in [0., 0.05]:\n",
    "                for lambda_l1_masked in [0, 5, 10]:\n",
    "                    for lambda_l1_addon in [5, 10]:\n",
    "                        for nichenet_keep_target_genes_ratio in [1.0]:\n",
    "                            for conv_layer_encoder in [\"gatv2conv\"]: \n",
    "                                task = \"reference\"\n",
    "                                dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "                                reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "                                n_neighbors = 12\n",
    "                                n_sampled_neighbors = -1\n",
    "                                n_hvg = 4000\n",
    "                                species = \"mouse\"\n",
    "                                node_label_method = \"one-hop-norm\"\n",
    "                                edge_batch_size = \"512\" # \"512\"\n",
    "                                cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "                                cat_covariates_keys = \"batch\"\n",
    "                                cat_covariates_no_edges = \"True\"\n",
    "                                cat_covariates_embeds_nums = \"3\"\n",
    "                                lambda_cat_covariates_contrastive = 0.\n",
    "                                contrastive_logits_pos_ratio = 0.\n",
    "                                contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "                                job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                                script_args = f\" --dataset {dataset}\" \\\n",
    "                                              f\" --reference_batches {reference_batches}\" \\\n",
    "                                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                              \" --filter_genes\" \\\n",
    "                                              f\" --n_hvg {n_hvg}\" \\\n",
    "                                              f\" --nichenet_keep_target_genes_ratio {nichenet_keep_target_genes_ratio}\" \\\n",
    "                                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                              \" --include_mebocost_gps\" \\\n",
    "                                              f\" --species {species}\" \\\n",
    "                                              \" --gp_filter_mode subset\" \\\n",
    "                                              \" --combine_overlap_gps\" \\\n",
    "                                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                                              \" --counts_key counts\" \\\n",
    "                                              \" --condition_key batch\" \\\n",
    "                                              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                              \" --spatial_key spatial\" \\\n",
    "                                              \" --adj_key spatial_connectivities\" \\\n",
    "                                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                              f\" --model_label {node_label_method}_{task}\" \\\n",
    "                                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                              \" --latent_key nichecompass_latent\" \\\n",
    "                                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                              \" --gene_expr_recon_dist nb\" \\\n",
    "                                              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                              \" --log_variational\" \\\n",
    "                                              f\" --node_label_method {node_label_method}\" \\\n",
    "                                              \" --n_layers_encoder 1\" \\\n",
    "                                              \" --n_hidden_encoder None\" \\\n",
    "                                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                              \" --n_epochs 100\" \\\n",
    "                                              \" --n_epochs_all_gps 25\" \\\n",
    "                                              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                              \" --lr 0.001\" \\\n",
    "                                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                              \" --lambda_group_lasso 0.\" \\\n",
    "                                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                              \" --node_batch_size None\" \\\n",
    "                                              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                                submit_python_script(\n",
    "                                        job_name_prefix=job_name_prefix,\n",
    "                                        job_id=job_id,\n",
    "                                        job_folder_path=job_folder_path,\n",
    "                                        conda_env_name=conda_env_name,\n",
    "                                        script_folder_path=script_folder_path,\n",
    "                                        script_name=script_name,\n",
    "                                        script_args=script_args,\n",
    "                                        nice=10000)\n",
    "\n",
    "                                job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cf736d52-ed9f-450b-b71c-0fe0809ab4af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12668637\n"
     ]
    }
   ],
   "source": [
    "# tmp\n",
    "job_id = 2\n",
    "for lambda_edge_recon in [500000]:\n",
    "    for lambda_gene_expr_recon in [300]:\n",
    "            for active_gp_thresh_ratio in [0.05]:\n",
    "                for lambda_l1_masked in [5.]:\n",
    "                    for lambda_l1_addon in [5.]:\n",
    "                        for nichenet_keep_target_genes_ratio in [1.0]:\n",
    "                            for conv_layer_encoder in [\"gcnconv\"]: \n",
    "                                task = \"reference\"\n",
    "                                dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "                                reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "                                n_neighbors = 4\n",
    "                                n_sampled_neighbors = -1\n",
    "                                n_hvg = 4000\n",
    "                                species = \"mouse\"\n",
    "                                node_label_method = \"one-hop-norm\"\n",
    "                                edge_batch_size = \"4096\"\n",
    "                                cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "                                cat_covariates_keys = \"batch\"\n",
    "                                cat_covariates_no_edges = \"True\"\n",
    "                                cat_covariates_embeds_nums = \"3\"\n",
    "                                lambda_cat_covariates_contrastive = 0.\n",
    "                                contrastive_logits_pos_ratio = 0.\n",
    "                                contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "                                job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                                script_args = f\" --dataset {dataset}\" \\\n",
    "                                              f\" --reference_batches {reference_batches}\" \\\n",
    "                                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                              \" --filter_genes\" \\\n",
    "                                              f\" --n_hvg {n_hvg}\" \\\n",
    "                                              f\" --nichenet_keep_target_genes_ratio {nichenet_keep_target_genes_ratio}\" \\\n",
    "                                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                              \" --include_mebocost_gps\" \\\n",
    "                                              f\" --species {species}\" \\\n",
    "                                              \" --gp_filter_mode subset\" \\\n",
    "                                              \" --combine_overlap_gps\" \\\n",
    "                                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                                              \" --counts_key counts\" \\\n",
    "                                              \" --condition_key batch\" \\\n",
    "                                              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                              \" --spatial_key spatial\" \\\n",
    "                                              \" --adj_key spatial_connectivities\" \\\n",
    "                                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                              f\" --model_label {node_label_method}_{task}\" \\\n",
    "                                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                              \" --latent_key nichecompass_latent\" \\\n",
    "                                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                              \" --gene_expr_recon_dist nb\" \\\n",
    "                                              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                              \" --log_variational\" \\\n",
    "                                              f\" --node_label_method {node_label_method}\" \\\n",
    "                                              \" --n_layers_encoder 1\" \\\n",
    "                                              \" --n_hidden_encoder None\" \\\n",
    "                                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                              \" --n_epochs 100\" \\\n",
    "                                              \" --n_epochs_all_gps 25\" \\\n",
    "                                              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                              \" --lr 0.001\" \\\n",
    "                                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                              \" --lambda_group_lasso 0.\" \\\n",
    "                                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                              \" --node_batch_size None\" \\\n",
    "                                              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                                submit_python_script(\n",
    "                                        job_name_prefix=job_name_prefix,\n",
    "                                        job_id=job_id,\n",
    "                                        job_folder_path=job_folder_path,\n",
    "                                        conda_env_name=conda_env_name,\n",
    "                                        script_folder_path=script_folder_path,\n",
    "                                        script_name=script_name,\n",
    "                                        script_args=script_args,\n",
    "                                        nice=10000)\n",
    "\n",
    "                                job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d78ebff-8b17-47c7-b126-8f7c350cc350",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 4\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "n_neighbors = 12\n",
    "n_hvg = 4000\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"4096\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 30.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 50000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 30000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aee0b50-4cfa-49f3-84b4-134a62b061b4",
   "metadata": {},
   "source": [
    "### 5.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11bdc0b5-5a82-4786-b53f-2e13cbe4f42c",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\" \\\n",
    "                    \" batch9 batch10 batch11 batch12 batch13 batch14 batch15 batch16 batch17 batch18\" \\\n",
    "                    \" batch19 batch20\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 8, 12\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 2048 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df29aa7c-fc1a-4de9-88a9-893dc54fb097",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 3\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\" \\\n",
    "                    \" batch9 batch10 batch11 batch12 batch13 batch14 batch15 batch16 batch17 batch18\" \\\n",
    "                    \" batch19 batch20\"\n",
    "n_neighbors = 12 # 8, 12\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "n_cond_embed = 20 # 3, 20, None\n",
    "lambda_cat_covariates_contrastive = 250000. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0.03125 # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --n_cond_embed {n_cond_embed}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fde3a4b-fd6f-43ba-9f37-436885f328df",
   "metadata": {},
   "source": [
    "### 5.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "463fcdec-a4d3-4d97-8f6c-6d4a9251158b",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc_filtered\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 20 # 4, 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 2048 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 0.01\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "216ebd95-3479-4f63-bbdd-cb192d06a347",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"fov\"\n",
    "cat_covariates_no_edges = \"False\"\n",
    "cat_covariates_embeds_nums = \"10\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cc78e13-e77e-4234-acab-22ab7ec5bdfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 10\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 10\"\n",
    "lambda_cat_covariates_contrastive = 100000.\n",
    "contrastive_logits_pos_ratio = 0.0625\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a420a844-aa9e-4f9b-8f17-9c46a5fd921c",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 11\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = None\n",
    "cat_covariates_no_edges = None\n",
    "cat_covariates_embeds_nums = None\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6612db72-e027-4302-8518-7ff94a2493f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 15\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 512 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"patient batch fov\"\n",
    "cat_covariates_embeds_nums = \"5 8 20\"\n",
    "lambda_cat_covariates_contrastive = 250000.\n",
    "contrastive_logits_pos_ratio = 0.0625\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7209ec-5334-4239-a142-d44ae5971bd9",
   "metadata": {},
   "source": [
    "### 5.4 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "671d5295-2e42-4365-811d-dcbaab2b0b20",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_neighbors = 4 # 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 4096 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"patient batch\"\n",
    "cat_covariates_no_edges = \"True True\"\n",
    "cat_covariates_embeds_nums = \"2 4\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c96dfb0-6b04-4bed-ae41-5af7591090a8",
   "metadata": {},
   "source": [
    "### 5.4 Vizgen MERFISH Human Lung Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16844a7c-93da-4f56-bb1f-1cd17701d9b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"vizgen_merfish_human_lung_cancer\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 4 # 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 4096 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"2\"\n",
    "lambda_cat_covariates_contrastive = 250000. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0.125 # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7380bdd-79ba-48a7-aeaa-ba20b8978edd",
   "metadata": {},
   "source": [
    "### 5.5 Spatial ATAC-RNA-Seq Mouse Embryo & Brain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5a246f1e-70e6-4760-84e6-373902ec86eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12640711\n"
     ]
    }
   ],
   "source": [
    "job_id = 100001\n",
    "for lambda_edge_recon in [500000]:\n",
    "    for lambda_gene_expr_recon in [300]:\n",
    "        for lambda_chrom_access_recon in [300]: # [100, 300, 1000, 3000, 10000, 30000]\n",
    "            for active_gp_thresh_ratio in [0.]: # [0., 0.03, 0.05, 0.1]\n",
    "                for lambda_l1_masked in [5.]: # [0, 5, 10, 50]\n",
    "                    for lambda_l1_addon in [5.]: # [0, 5, 10, 50]\n",
    "                        for nichenet_keep_target_genes_ratio in [1.0]: # [1.0, 0.1]\n",
    "                            task = \"reference\"\n",
    "                            dataset = \"spatial_atac_rna_seq_mouse_brain_batch2\"\n",
    "                            reference_batches = \"None\"\n",
    "                            n_neighbors = 8\n",
    "                            species = \"mouse\"\n",
    "                            node_label_method = \"one-hop-norm\"\n",
    "                            edge_batch_size = \"4096\"\n",
    "                            n_sampled_neighbors = 4\n",
    "                            lambda_cat_covariates_contrastive = 0.\n",
    "                            contrastive_logits_pos_ratio = 0.\n",
    "                            contrastive_logits_neg_ratio = 0.\n",
    "                            cat_covariates_embeds_injection = \"None\"\n",
    "                            cat_covariates_keys = \"batch\"\n",
    "                            cat_covariates_no_edges = \"True\"\n",
    "                            cat_covariates_embeds_nums = 0\n",
    "                            n_hvg = 3000\n",
    "\n",
    "                            job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                            script_name = \"train_nichecompass_reference_model.py\"\n",
    "                            script_args = f\" --dataset {dataset}\" \\\n",
    "                                          f\" --reference_batches {reference_batches}\" \\\n",
    "                                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                          \" --filter_genes\" \\\n",
    "                                          f\" --n_hvg {n_hvg}\" \\\n",
    "                                          f\" --nichenet_keep_target_genes_ratio {nichenet_keep_target_genes_ratio}\" \\\n",
    "                                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                          \" --include_mebocost_gps\" \\\n",
    "                                          \" --include_collectri_gps\" \\\n",
    "                                          f\" --species {species}\" \\\n",
    "                                          \" --gp_filter_mode subset\" \\\n",
    "                                          \" --combine_overlap_gps\" \\\n",
    "                                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                                          \" --counts_key counts\" \\\n",
    "                                          \" --condition_key batch\" \\\n",
    "                                          f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                          f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                          \" --spatial_key spatial\" \\\n",
    "                                          \" --adj_key spatial_connectivities\" \\\n",
    "                                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                          \" --include_atac_modality\" \\\n",
    "                                          \" --filter_peaks\" \\\n",
    "                                          \" --min_cell_peak_thresh_ratio 0.0005\" \\\n",
    "                                          f\" --model_label {node_label_method}_{task}\" \\\n",
    "                                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                          \" --latent_key nichecompass_latent\" \\\n",
    "                                          f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                          \" --gene_expr_recon_dist nb\" \\\n",
    "                                          f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                          f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                          \" --log_variational\" \\\n",
    "                                          f\" --node_label_method {node_label_method}\" \\\n",
    "                                          \" --n_layers_encoder 1\" \\\n",
    "                                          \" --n_hidden_encoder None\" \\\n",
    "                                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                                          \" --n_epochs 100\" \\\n",
    "                                          \" --n_epochs_all_gps 25\" \\\n",
    "                                          \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                          \" --lr 0.001\" \\\n",
    "                                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                          f\" --lambda_chrom_access_recon {lambda_chrom_access_recon}\" \\\n",
    "                                          f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                          f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                          f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                          \" --lambda_group_lasso 0.\" \\\n",
    "                                          f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                          \" --node_batch_size None\" \\\n",
    "                                          f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                            submit_python_script(\n",
    "                                    job_name_prefix=job_name_prefix,\n",
    "                                    job_id=job_id,\n",
    "                                    job_folder_path=job_folder_path,\n",
    "                                    conda_env_name=conda_env_name,\n",
    "                                    script_folder_path=script_folder_path,\n",
    "                                    script_name=script_name,\n",
    "                                    script_args=script_args,\n",
    "                                    nice=10000)\n",
    "\n",
    "                            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6cde2a39-e1d1-49c2-a429-6e11eb8f84ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12663365\n",
      "Submitted batch job 12663366\n",
      "Submitted batch job 12663367\n",
      "Submitted batch job 12663368\n",
      "Submitted batch job 12663369\n",
      "Submitted batch job 12663370\n",
      "Submitted batch job 12663371\n",
      "Submitted batch job 12663372\n",
      "Submitted batch job 12663373\n",
      "Submitted batch job 12663374\n",
      "Submitted batch job 12663375\n",
      "Submitted batch job 12663376\n",
      "Submitted batch job 12663377\n",
      "Submitted batch job 12663378\n",
      "Submitted batch job 12663379\n",
      "Submitted batch job 12663380\n",
      "Submitted batch job 12663381\n",
      "Submitted batch job 12663382\n",
      "Submitted batch job 12663383\n",
      "Submitted batch job 12663384\n",
      "Submitted batch job 12663385\n",
      "Submitted batch job 12663386\n",
      "Submitted batch job 12663387\n",
      "Submitted batch job 12663388\n",
      "Submitted batch job 12663389\n",
      "Submitted batch job 12663390\n",
      "Submitted batch job 12663391\n",
      "Submitted batch job 12663392\n",
      "Submitted batch job 12663393\n",
      "Submitted batch job 12663394\n",
      "Submitted batch job 12663395\n",
      "Submitted batch job 12663396\n",
      "Submitted batch job 12663397\n",
      "Submitted batch job 12663398\n",
      "Submitted batch job 12663399\n",
      "Submitted batch job 12663400\n",
      "Submitted batch job 12663401\n",
      "Submitted batch job 12663402\n",
      "Submitted batch job 12663403\n",
      "Submitted batch job 12663404\n",
      "Submitted batch job 12663405\n",
      "Submitted batch job 12663406\n",
      "Submitted batch job 12663407\n",
      "Submitted batch job 12663408\n",
      "Submitted batch job 12663409\n",
      "Submitted batch job 12663410\n",
      "Submitted batch job 12663411\n",
      "Submitted batch job 12663412\n",
      "Submitted batch job 12663413\n",
      "Submitted batch job 12663414\n",
      "Submitted batch job 12663415\n",
      "Submitted batch job 12663416\n",
      "Submitted batch job 12663417\n",
      "Submitted batch job 12663418\n",
      "Submitted batch job 12663419\n",
      "Submitted batch job 12663420\n",
      "Submitted batch job 12663421\n",
      "Submitted batch job 12663422\n",
      "Submitted batch job 12663423\n",
      "Submitted batch job 12663424\n",
      "Submitted batch job 12663425\n",
      "Submitted batch job 12663426\n",
      "Submitted batch job 12663427\n",
      "Submitted batch job 12663428\n",
      "Submitted batch job 12663429\n",
      "Submitted batch job 12663430\n",
      "Submitted batch job 12663431\n",
      "Submitted batch job 12663432\n",
      "Submitted batch job 12663433\n",
      "Submitted batch job 12663434\n",
      "Submitted batch job 12663435\n",
      "Submitted batch job 12663436\n",
      "Submitted batch job 12663437\n",
      "Submitted batch job 12663438\n",
      "Submitted batch job 12663439\n",
      "Submitted batch job 12663440\n",
      "Submitted batch job 12663441\n",
      "Submitted batch job 12663442\n",
      "Submitted batch job 12663443\n",
      "Submitted batch job 12663444\n",
      "Submitted batch job 12663445\n",
      "Submitted batch job 12663446\n",
      "Submitted batch job 12663447\n",
      "Submitted batch job 12663448\n",
      "Submitted batch job 12663449\n",
      "Submitted batch job 12663450\n",
      "Submitted batch job 12663451\n",
      "Submitted batch job 12663452\n",
      "Submitted batch job 12663453\n",
      "Submitted batch job 12663454\n",
      "Submitted batch job 12663455\n",
      "Submitted batch job 12663456\n",
      "Submitted batch job 12663457\n",
      "Submitted batch job 12663458\n",
      "Submitted batch job 12663459\n",
      "Submitted batch job 12663460\n",
      "Submitted batch job 12663461\n",
      "Submitted batch job 12663462\n",
      "Submitted batch job 12663463\n",
      "Submitted batch job 12663464\n",
      "Submitted batch job 12663465\n",
      "Submitted batch job 12663466\n",
      "Submitted batch job 12663467\n",
      "Submitted batch job 12663468\n",
      "Submitted batch job 12663469\n",
      "Submitted batch job 12663470\n",
      "Submitted batch job 12663471\n",
      "Submitted batch job 12663472\n",
      "Submitted batch job 12663473\n",
      "Submitted batch job 12663474\n",
      "Submitted batch job 12663475\n",
      "Submitted batch job 12663476\n",
      "Submitted batch job 12663477\n",
      "Submitted batch job 12663478\n",
      "Submitted batch job 12663479\n",
      "Submitted batch job 12663480\n",
      "Submitted batch job 12663481\n",
      "Submitted batch job 12663482\n",
      "Submitted batch job 12663483\n",
      "Submitted batch job 12663484\n",
      "Submitted batch job 12663485\n",
      "Submitted batch job 12663486\n",
      "Submitted batch job 12663487\n",
      "Submitted batch job 12663488\n",
      "Submitted batch job 12663489\n",
      "Submitted batch job 12663490\n",
      "Submitted batch job 12663491\n",
      "Submitted batch job 12663492\n",
      "Submitted batch job 12663493\n",
      "Submitted batch job 12663494\n",
      "Submitted batch job 12663495\n",
      "Submitted batch job 12663496\n",
      "Submitted batch job 12663497\n",
      "Submitted batch job 12663498\n",
      "Submitted batch job 12663499\n",
      "Submitted batch job 12663500\n",
      "Submitted batch job 12663501\n",
      "Submitted batch job 12663502\n",
      "Submitted batch job 12663503\n",
      "Submitted batch job 12663504\n",
      "Submitted batch job 12663505\n",
      "Submitted batch job 12663506\n",
      "Submitted batch job 12663507\n",
      "Submitted batch job 12663508\n",
      "Submitted batch job 12663509\n",
      "Submitted batch job 12663510\n",
      "Submitted batch job 12663511\n",
      "Submitted batch job 12663512\n",
      "Submitted batch job 12663513\n",
      "Submitted batch job 12663514\n",
      "Submitted batch job 12663515\n",
      "Submitted batch job 12663516\n",
      "Submitted batch job 12663517\n",
      "Submitted batch job 12663518\n",
      "Submitted batch job 12663519\n",
      "Submitted batch job 12663520\n",
      "Submitted batch job 12663521\n",
      "Submitted batch job 12663522\n",
      "Submitted batch job 12663523\n",
      "Submitted batch job 12663524\n",
      "Submitted batch job 12663525\n",
      "Submitted batch job 12663526\n"
     ]
    }
   ],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [500000, 5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [300, 3000, 30000]:\n",
    "        for lambda_chrom_access_recon in [300, 3000, 30000]: # [100, 300, 1000, 3000, 10000, 30000]\n",
    "            for active_gp_thresh_ratio in [0., 0.05, 0.1]: # [0., 0.03, 0.05, 0.1]\n",
    "                for lambda_l1_masked in [0, 5]: # [0, 5, 10, 50]\n",
    "                    for lambda_l1_addon in [5.]: # [0, 5, 10, 50]\n",
    "                        for nichenet_keep_target_genes_ratio in [0.1]: # [1.0, 0.1]\n",
    "                            task = \"reference\"\n",
    "                            dataset = \"spatial_atac_rna_seq_mouse_brain_batch2\"\n",
    "                            reference_batches = \"None\"\n",
    "                            n_neighbors = 8\n",
    "                            species = \"mouse\"\n",
    "                            node_label_method = \"one-hop-norm\"\n",
    "                            edge_batch_size = \"4096\"\n",
    "                            n_sampled_neighbors = 4\n",
    "                            lambda_cat_covariates_contrastive = 0.\n",
    "                            contrastive_logits_pos_ratio = 0.\n",
    "                            contrastive_logits_neg_ratio = 0.\n",
    "                            cat_covariates_embeds_injection = \"None\"\n",
    "                            cat_covariates_keys = \"batch\"\n",
    "                            cat_covariates_no_edges = \"True\"\n",
    "                            cat_covariates_embeds_nums = 0\n",
    "                            n_hvg = 3000\n",
    "\n",
    "                            job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                            script_name = \"train_nichecompass_reference_model.py\"\n",
    "                            script_args = f\" --dataset {dataset}\" \\\n",
    "                                          f\" --reference_batches {reference_batches}\" \\\n",
    "                                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                          \" --filter_genes\" \\\n",
    "                                          f\" --n_hvg {n_hvg}\" \\\n",
    "                                          f\" --nichenet_keep_target_genes_ratio {nichenet_keep_target_genes_ratio}\" \\\n",
    "                                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                          \" --include_mebocost_gps\" \\\n",
    "                                          \" --include_collectri_gps\" \\\n",
    "                                          f\" --species {species}\" \\\n",
    "                                          \" --gp_filter_mode subset\" \\\n",
    "                                          \" --combine_overlap_gps\" \\\n",
    "                                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                                          \" --counts_key counts\" \\\n",
    "                                          \" --condition_key batch\" \\\n",
    "                                          f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                          f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                          \" --spatial_key spatial\" \\\n",
    "                                          \" --adj_key spatial_connectivities\" \\\n",
    "                                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                          \" --include_atac_modality\" \\\n",
    "                                          \" --filter_peaks\" \\\n",
    "                                          \" --min_cell_peak_thresh_ratio 0.0005\" \\\n",
    "                                          f\" --model_label {node_label_method}_{task}\" \\\n",
    "                                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                          \" --latent_key nichecompass_latent\" \\\n",
    "                                          f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                          \" --gene_expr_recon_dist nb\" \\\n",
    "                                          f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                          f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                          \" --log_variational\" \\\n",
    "                                          f\" --node_label_method {node_label_method}\" \\\n",
    "                                          \" --n_layers_encoder 1\" \\\n",
    "                                          \" --n_hidden_encoder None\" \\\n",
    "                                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                                          \" --n_epochs 100\" \\\n",
    "                                          \" --n_epochs_all_gps 25\" \\\n",
    "                                          \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                          \" --lr 0.001\" \\\n",
    "                                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                          f\" --lambda_chrom_access_recon {lambda_chrom_access_recon}\" \\\n",
    "                                          f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                          f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                          f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                          \" --lambda_group_lasso 0.\" \\\n",
    "                                          f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                          \" --node_batch_size None\" \\\n",
    "                                          f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                            submit_python_script(\n",
    "                                    job_name_prefix=job_name_prefix,\n",
    "                                    job_id=job_id,\n",
    "                                    job_folder_path=job_folder_path,\n",
    "                                    conda_env_name=conda_env_name,\n",
    "                                    script_folder_path=script_folder_path,\n",
    "                                    script_name=script_name,\n",
    "                                    script_args=script_args,\n",
    "                                    nice=10000)\n",
    "\n",
    "                            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01341dcb-64ad-45bd-8ef6-939e77d5b0f8",
   "metadata": {},
   "source": [
    "### 5.6 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68c5b999-11e0-42e5-b886-8fb3a47d82f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"xenium_human_breast_cancer\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 4 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-attention\"\n",
    "edge_batch_size = 4096 # 4096 (4, 8), 1024 (12)\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"2\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0. # 0., 5., 10.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "001945c5-6d59-4bdb-9efe-d8bbb33edc33",
   "metadata": {},
   "source": [
    "## 6. NicheCompass Reference Query Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14bf6b4d-34b5-4324-8deb-095bdfa7d0e2",
   "metadata": {},
   "source": [
    "### 6.1 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a6a826d-f6ad-4ee5-85b2-66a3670a574e",
   "metadata": {},
   "source": [
    "#### 6.1.1 Reference Model Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdc61c79-6652-49fb-8bef-71ba0dda3b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_neighbors = 12\n",
    "n_hvg = 3000\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 4096\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_reference_only\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7db14469-4ee2-4be2-b00d-535bc0439620",
   "metadata": {},
   "source": [
    "#### 6.1.2 Query Mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17761bc0-3178-42dd-9b0e-3935be8d2a5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 1\n",
    "query_batches = \"batch5 batch6\"\n",
    "n_neighbors = 12\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"01072023_165203_1\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "edge_batch_size = 4096\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label {node_label_method}_{task}_query_only\" \\\n",
    "              f\" --reference_query_model_label {node_label_method}_{task}_query_mapping\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5391f85b-21e3-4723-bcef-6f7b0d948b4c",
   "metadata": {},
   "source": [
    "### 6.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "210672eb-c6ef-4cf5-9953-c2e4fcdec335",
   "metadata": {},
   "source": [
    "#### 6.2.1 Reference Model Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90d450f9-673b-42ba-b11d-03fcd295693d",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 4\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 8 # 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 250000. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0.0625 # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "149598af-5167-4466-b1e2-89de2f7375e2",
   "metadata": {},
   "source": [
    "#### 6.2.2 Query Mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a73f7984-551c-49f4-90f6-47eecf5092d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 1\n",
    "query_batches = \"batch3\"\n",
    "n_neighbors = 12\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"06072023_114143_1\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "edge_batch_size = 1024\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label {node_label_method}_{task}_query_only\" \\\n",
    "              f\" --reference_query_model_label {node_label_method}_{task}_query_mapping\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da211ed3-d9d9-4e55-a008-a33c0ca4f34a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "toc-showcode": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
