{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bdc8086b-2b80-43b4-b88a-aa0c87c820a6",
   "metadata": {},
   "source": [
    "# Slurm Job Submission"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be88d0ca-4c09-44e3-83fc-fb6bcf4563fd",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>).\n",
    "- **Affiliation:** Helmholtz Munich, Institute of Computational Biology (ICB), Talavera-LÃ³pez Lab\n",
    "- **Date of Creation:** 20.03.2023\n",
    "- **Date of Last Modification:** 27.06.2023"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aeeeb1fe-eaa1-4cdd-aa21-ed78e0b7be81",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e58b0a3-c3fe-43d5-a420-14111aa545f6",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "48d297b0-e0dc-42b9-a349-e6ac2edba0e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aab82f4c-edd8-47dd-be56-559cef739439",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0491ee0e-3587-4243-856f-e7f71e6dced6",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda_env_name = \"nichecompass\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "004ab52a-9cc3-4e40-b0e6-f9fed7f4a549",
   "metadata": {},
   "source": [
    "### 1.3 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6b2d6cf0-b360-428f-9548-c7942378de68",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_python_script(\n",
    "        job_name_prefix,\n",
    "        job_id,\n",
    "        job_folder_path,\n",
    "        conda_env_name,\n",
    "        script_folder_path,\n",
    "        script_name,\n",
    "        script_args,\n",
    "        nice=10000):\n",
    "    job_name = f\"{job_name_prefix}_{job_id}\"\n",
    "    # Account for fact that submit node has different home path than compute node\n",
    "    job_file_path = f\"{job_folder_path.replace('/aih', '')}/job_{job_name}.cmd\"\n",
    "    out_file_path = f\"{job_folder_path}/logs/out_{job_name}.txt\"\n",
    "    err_file_path = f\"{job_folder_path}/logs/err_{job_name}.txt\"\n",
    "    \n",
    "    os.makedirs(job_folder_path + \"/logs\", exist_ok=True)\n",
    "    \n",
    "    with open(job_file_path, \"w\") as handle:\n",
    "        handle.writelines(\"#!/bin/bash\\n\")\n",
    "        handle.writelines(f\"#SBATCH -J {job_name}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -o {out_file_path}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -e {err_file_path}\\n\")\n",
    "        handle.writelines(\"#SBATCH -t 48:00:00\\n\")\n",
    "        handle.writelines(\"#SBATCH -p gpu_p\\n\")\n",
    "        handle.writelines(\"#SBATCH -c 6\\n\")\n",
    "        handle.writelines(\"#SBATCH --gres=gpu:1\\n\")\n",
    "        handle.writelines(\"#SBATCH --qos=gpu\\n\")       \n",
    "        handle.writelines(\"#SBATCH --mem=64GB\\n\")\n",
    "        handle.writelines(f\"#SBATCH --nice={nice}\\n\")\n",
    "        handle.writelines(\"source $HOME/.bashrc\\n\")\n",
    "        handle.writelines(f\"conda activate {conda_env_name}\\n\")\n",
    "        handle.writelines(\"cd /\\n\")\n",
    "        handle.writelines(f\"cd {script_folder_path}\\n\")\n",
    "        handle.writelines(f\"python ../{script_name}\")\n",
    "        handle.writelines(f\"{script_args}\")\n",
    "        handle.writelines(\"\\n\")\n",
    "        \n",
    "    os.system(f\"sbatch {job_file_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4624e92-33eb-457c-8767-9006e44d850e",
   "metadata": {},
   "source": [
    "## 2. NicheCompass Ablation Model Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc99f810-89a4-4070-ac09-d3f6da04c95d",
   "metadata": {},
   "source": [
    "### 2.1 Loss Weights"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c84d94e-4fa8-4edc-8e55-2ec3cf708127",
   "metadata": {},
   "source": [
    "#### 2.1.1 seqFISH Mouse Organogenesis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "604fcb62-7a98-4ba8-96dc-6daf55067e84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12273517\n",
      "Submitted batch job 12273518\n",
      "Submitted batch job 12273519\n",
      "Submitted batch job 12273520\n",
      "Submitted batch job 12273521\n",
      "Submitted batch job 12273522\n",
      "Submitted batch job 12273523\n",
      "Submitted batch job 12273524\n",
      "Submitted batch job 12273525\n",
      "Submitted batch job 12273526\n",
      "Submitted batch job 12273527\n",
      "Submitted batch job 12273528\n",
      "Submitted batch job 12273529\n",
      "Submitted batch job 12273530\n",
      "Submitted batch job 12273531\n",
      "Submitted batch job 12273532\n",
      "Submitted batch job 12273533\n",
      "Submitted batch job 12273534\n",
      "Submitted batch job 12273535\n",
      "Submitted batch job 12273536\n",
      "Submitted batch job 12273537\n",
      "Submitted batch job 12273538\n",
      "Submitted batch job 12273539\n",
      "Submitted batch job 12273540\n",
      "Submitted batch job 12273541\n",
      "Submitted batch job 12273542\n",
      "Submitted batch job 12273543\n",
      "Submitted batch job 12273544\n",
      "Submitted batch job 12273545\n",
      "Submitted batch job 12273546\n",
      "Submitted batch job 12273547\n",
      "Submitted batch job 12273548\n",
      "Submitted batch job 12273549\n",
      "Submitted batch job 12273550\n",
      "Submitted batch job 12273551\n",
      "Submitted batch job 12273552\n",
      "Submitted batch job 12273553\n",
      "Submitted batch job 12273554\n",
      "Submitted batch job 12273555\n",
      "Submitted batch job 12273556\n",
      "Submitted batch job 12273557\n",
      "Submitted batch job 12273558\n",
      "Submitted batch job 12273559\n",
      "Submitted batch job 12273560\n",
      "Submitted batch job 12273561\n"
     ]
    }
   ],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16, 20]:\n",
    "            task = \"ablation\"\n",
    "            dataset = \"seqfish_mouse_organogenesis\"\n",
    "            reference_batches = \"batch1 batch2\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 4096\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_loss_weights_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label loss_weights_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e95996-f6e0-459c-a400-5a2ae3193f47",
   "metadata": {},
   "source": [
    "#### 2.1.2 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0621ca71-70a3-423e-9f79-91e706f5d4e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12274606\n",
      "Submitted batch job 12274607\n",
      "Submitted batch job 12274608\n",
      "Submitted batch job 12274609\n",
      "Submitted batch job 12274610\n",
      "Submitted batch job 12274611\n",
      "Submitted batch job 12274612\n",
      "Submitted batch job 12274613\n",
      "Submitted batch job 12274614\n",
      "Submitted batch job 12274615\n",
      "Submitted batch job 12274616\n",
      "Submitted batch job 12274617\n",
      "Submitted batch job 12274618\n",
      "Submitted batch job 12274619\n",
      "Submitted batch job 12274620\n",
      "Submitted batch job 12274621\n",
      "Submitted batch job 12274622\n",
      "Submitted batch job 12274623\n",
      "Submitted batch job 12274624\n",
      "Submitted batch job 12274625\n",
      "Submitted batch job 12274626\n",
      "Submitted batch job 12274627\n",
      "Submitted batch job 12274628\n",
      "Submitted batch job 12274629\n",
      "Submitted batch job 12274630\n",
      "Submitted batch job 12274631\n",
      "Submitted batch job 12274632\n",
      "Submitted batch job 12274633\n",
      "Submitted batch job 12274634\n",
      "Submitted batch job 12274635\n",
      "Submitted batch job 12274636\n",
      "Submitted batch job 12274637\n",
      "Submitted batch job 12274638\n",
      "Submitted batch job 12274639\n",
      "Submitted batch job 12274640\n",
      "Submitted batch job 12274641\n",
      "Submitted batch job 12274642\n",
      "Submitted batch job 12274643\n",
      "Submitted batch job 12274644\n",
      "Submitted batch job 12274645\n",
      "Submitted batch job 12274646\n",
      "Submitted batch job 12274647\n",
      "Submitted batch job 12274648\n",
      "Submitted batch job 12274649\n",
      "Submitted batch job 12274650\n"
     ]
    }
   ],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16, 20]:\n",
    "            task = \"ablation\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_loss_weights_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label loss_weights_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea685a7c-9ea8-428c-af96-c5bddfa48924",
   "metadata": {},
   "source": [
    "### 2.1 Encoder Architecture"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "591b030c-6f6f-4cb5-b049-5087220f7856",
   "metadata": {},
   "source": [
    "### 2.2 Gene Program Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "79f5917b-eead-4b63-bef6-7aa6d8bedcb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12270217\n",
      "Submitted batch job 12270218\n",
      "Submitted batch job 12270219\n",
      "Submitted batch job 12270220\n"
     ]
    }
   ],
   "source": [
    "job_id = 1\n",
    "for cat_covariates_embeds_injection in [\"gene_expr_decoder\", \"encoder gene_expr_decoder\"]:\n",
    "    for cat_covariates_keys, cat_covariates_no_edges in zip(\n",
    "        [\"patient batch\", \"batch\"], [\"True True\", \"True\"]):\n",
    "        task = \"ablation\"\n",
    "        dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "        reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "        n_neighbors = 12\n",
    "        species = \"human\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 1024\n",
    "        cat_covariates_embeds_nums = None\n",
    "        lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "        contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "        contrastive_logits_neg_ratio = 0.\n",
    "        lambda_l1_masked = 0.\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                      f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {node_label_method}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                      f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      \" --lambda_edge_recon 500000.\" \\\n",
    "                      \" --lambda_gene_expr_recon 300.\" \\\n",
    "                      f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                      f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                      f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f34f4bc-50a7-42ff-938a-8890d3ffa6ff",
   "metadata": {},
   "source": [
    "### 2.3 Gene Program Selection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f553bea4-6178-4225-a07a-cc67c5c92d74",
   "metadata": {},
   "source": [
    "### 2.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77543dff-356e-45e4-b0a5-473f7dd5ea00",
   "metadata": {},
   "source": [
    "### 2.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a80b724d-8200-44b4-8dba-02269e4358e0",
   "metadata": {},
   "source": [
    "## 3. NicheCompass Single Sample Method Benchmarking Models Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eb67595-8ede-4823-a1fe-a313dee704f5",
   "metadata": {},
   "source": [
    "### 3.1 seqFISH Mouse Organogenesis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b52c28e8-6fc8-492f-9c31-d3ab9b815980",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.1.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6fe70711-9d03-47d5-892a-5cdf77714389",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12262744\n"
     ]
    }
   ],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_embryo2\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24ecebdb-2b38-4021-8116-b21061bef3b1",
   "metadata": {},
   "source": [
    "#### 3.1.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3c3995e5-3c35-427e-838c-4831370d44c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12228081\n",
      "Submitted batch job 12228082\n",
      "Submitted batch job 12228083\n",
      "Submitted batch job 12228084\n",
      "Submitted batch job 12228085\n"
     ]
    }
   ],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct_embryo2\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19574d56-9c79-4874-9b4d-ee39248f0028",
   "metadata": {},
   "source": [
    "### 3.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe560799-9d98-4740-af7a-b972c7596d18",
   "metadata": {},
   "source": [
    "#### 3.2.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "85bd3ffa-40d7-4414-b9fc-0a9e577f6a4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12084714\n"
     ]
    }
   ],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"starmap_plus_mouse_cns_batch1\"\n",
    "cell_type_key = \"Main_molecular_cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b1548b5-4dca-4762-b2b1-3b70211ed57f",
   "metadata": {},
   "source": [
    "#### 3.2.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5242ada-198f-46a6-b545-048ab451c42e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"starmap_plus_mouse_cns_subsample_{subsample_pct}pct_batch1\"\n",
    "    cell_type_key = \"Main_molecular_cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4111eee6-88e7-4180-bc5b-a05cdde41061",
   "metadata": {},
   "source": [
    "### 3.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcce8ca8-124d-4814-a224-900bd546165e",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.3.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "7842f6f0-664b-4405-ad30-add0af93ec7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12227832\n"
     ]
    }
   ],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59d54d8a-bea3-40a9-acfa-4e78d2cffffe",
   "metadata": {},
   "source": [
    "#### 3.3.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "1e7b792d-e60d-4a14-a971-7aca558dca0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12228081\n",
      "Submitted batch job 12228082\n",
      "Submitted batch job 12228083\n",
      "Submitted batch job 12228084\n",
      "Submitted batch job 12228085\n"
     ]
    }
   ],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c88c8e4-3f52-4ac4-8e9b-8ed601511180",
   "metadata": {},
   "source": [
    "### 3.4 Vizgen MERFISH Mouse Liver"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91da577-1daa-484b-809f-4c7c080aa129",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.4.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "9d4f57ce-ed16-4994-a6dd-6dafe87a719f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12227832\n"
     ]
    }
   ],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d826697c-54bd-44ad-a33a-2fe56af946db",
   "metadata": {},
   "source": [
    "#### 3.4.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "faa871f9-7357-4d38-a931-09a67728e0f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12228081\n",
      "Submitted batch job 12228082\n",
      "Submitted batch job 12228083\n",
      "Submitted batch job 12228084\n",
      "Submitted batch job 12228085\n"
     ]
    }
   ],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c39bdcc-6c2f-45ae-bc75-305db782e1e2",
   "metadata": {},
   "source": [
    "### 3.5 Slide-seqV2 Mouse Hippocampus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "278ca85b-f667-460d-b54f-98101eb221d4",
   "metadata": {},
   "source": [
    "#### 3.5.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1f0bccd-24db-4252-933a-fffb5e17c618",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"slideseqv2_mouse_hippocampus\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\" # one-hop-norm, one-hop-attention\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ed715aa-9f15-445b-a37a-4ea469ca91b2",
   "metadata": {},
   "source": [
    "#### 3.5.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c94ac68-2133-41d2-913f-94438f7e26f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"slideseqv2_mouse_hippocampus_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-attention\" # one-hop-norm, one-hop-attention\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\" # 16384 is full dataset\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "622f6c1d-65a1-4318-b03c-c12444fd0f4c",
   "metadata": {},
   "source": [
    "## 4. NicheCompass Sample Integration Method Benchmarking Models Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd147edf-4ccd-402d-aca9-ce19cbd1b34f",
   "metadata": {},
   "source": [
    "### 4.1 seqFISH Mouse Organogenesis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b865230b-ab67-4474-a29c-dcb1d2294340",
   "metadata": {},
   "source": [
    "#### 4.1.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "0f764998-8726-419f-bafe-d627f75a1257",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12263342\n"
     ]
    }
   ],
   "source": [
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_embeds_nums = \"6\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dac84fe1-0698-4c04-868f-2dd6e7d73356",
   "metadata": {},
   "source": [
    "#### 4.1.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "63f15cc6-4e6d-4473-a126-dccb02b889a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12263343\n",
      "Submitted batch job 12263344\n",
      "Submitted batch job 12263345\n",
      "Submitted batch job 12263346\n",
      "Submitted batch job 12263347\n"
     ]
    }
   ],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-norm\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_embeds_nums = \"6\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b334564-7198-4344-8e8b-b20897765f3c",
   "metadata": {},
   "source": [
    "### 4.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dfecebf-0b36-47f9-a3a4-b59938733399",
   "metadata": {},
   "source": [
    "#### 4.2.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "679dc44d-14de-493b-8407-6bcf0debf032",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"Main_molecular_cell_type\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61a05506-ee44-4d31-a43a-63a82c40ad1f",
   "metadata": {},
   "source": [
    "#### 4.2.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92069eca-64a0-4ab1-be3d-a5b0059145be",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"starmap_plus_mouse_cns_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"Main_molecular_cell_type\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-norm\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ab268b-d86a-4843-9623-6f687c4f0a11",
   "metadata": {},
   "source": [
    "### 4.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfda56b3-cb23-4b88-95f4-b3d8e8651ec2",
   "metadata": {},
   "source": [
    "#### 4.3.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79414d8b-71c9-41f8-9cbe-921a94589c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_embeds_nums = \"3 10\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "               \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "               \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {node_label_method}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.05\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c112831f-87a4-41f8-9478-02f71d678a3f",
   "metadata": {},
   "source": [
    "#### 4.3.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7c7fe02-1889-4fe4-99db-d7602bf87a3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    node_label_method = \"one-hop-norm\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_embeds_nums = \"3 10\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4 8 8 12 12 16 16 20 20\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None None None\" \\\n",
    "                   \" --seeds 0 1 2 3 4 5 6 7 8 9\" \\\n",
    "                   \" --run_index 1 2 3 4 5 6 7 8 9 10\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {node_label_method}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.05\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method {node_label_method}\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 500000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 300.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "828f673a-9f55-4b43-9ff2-ca286d8ed45a",
   "metadata": {},
   "source": [
    "## 5. NicheCompass Reference Model Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56180669-13e7-4873-8887-5804360c2b75",
   "metadata": {},
   "source": [
    "### 5.1 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8e6bed3a-6613-4101-81e5-9c19bb217b00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12272955\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "n_neighbors = 12\n",
    "n_hvg = 4000\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"4096\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aee0b50-4cfa-49f3-84b4-134a62b061b4",
   "metadata": {},
   "source": [
    "### 5.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "11bdc0b5-5a82-4786-b53f-2e13cbe4f42c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12208376\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\" \\\n",
    "                    \" batch9 batch10 batch11 batch12 batch13 batch14 batch15 batch16 batch17 batch18\" \\\n",
    "                    \" batch19 batch20\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 8, 12\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 2048 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "df29aa7c-fc1a-4de9-88a9-893dc54fb097",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12225823\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 3\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\" \\\n",
    "                    \" batch9 batch10 batch11 batch12 batch13 batch14 batch15 batch16 batch17 batch18\" \\\n",
    "                    \" batch19 batch20\"\n",
    "n_neighbors = 12 # 8, 12\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "n_cond_embed = 20 # 3, 20, None\n",
    "lambda_cat_covariates_contrastive = 250000. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0.03125 # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --n_cond_embed {n_cond_embed}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fde3a4b-fd6f-43ba-9f37-436885f328df",
   "metadata": {},
   "source": [
    "### 5.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "463fcdec-a4d3-4d97-8f6c-6d4a9251158b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12270955\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "216ebd95-3479-4f63-bbdd-cb192d06a347",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12275697\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 3\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"fov\"\n",
    "cat_covariates_no_edges = \"False\"\n",
    "cat_covariates_embeds_nums = \"10\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "8cc78e13-e77e-4234-acab-22ab7ec5bdfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12275698\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 4\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 10\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6612db72-e027-4302-8518-7ff94a2493f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12262680\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 15\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\"\n",
    "n_neighbors = 12 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 512 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"patient batch fov\"\n",
    "cat_covariates_embeds_nums = \"5 8 20\"\n",
    "lambda_cat_covariates_contrastive = 250000.\n",
    "contrastive_logits_pos_ratio = 0.0625\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7209ec-5334-4239-a142-d44ae5971bd9",
   "metadata": {},
   "source": [
    "### 5.4 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "671d5295-2e42-4365-811d-dcbaab2b0b20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12269781\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_neighbors = 4 # 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 4096 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"patient batch\"\n",
    "cat_covariates_no_edges = \"True True\"\n",
    "cat_covariates_embeds_nums = \"2 4\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c96dfb0-6b04-4bed-ae41-5af7591090a8",
   "metadata": {},
   "source": [
    "### 5.4 Vizgen MERFISH Human Lung Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "16844a7c-93da-4f56-bb1f-1cd17701d9b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12222718\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"vizgen_merfish_human_lung_cancer\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 4 # 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 4096 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"2\"\n",
    "lambda_cat_covariates_contrastive = 250000. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0.125 # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7380bdd-79ba-48a7-aeaa-ba20b8978edd",
   "metadata": {},
   "source": [
    "### 5.5 Spatial ATAC-RNA-Seq Mouse Embryo & Brain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cde2a39-e1d1-49c2-a429-6e11eb8f84ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"spatial_atac_rna_seq_mouse_brain\"\n",
    "reference_batches = \"None\"\n",
    "n_neighbors = 12\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"4096\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              \" --n_hvg 3000\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --mebocost_species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              \" --include_atac_modality\" \\\n",
    "              \" --filter_peaks\" \\\n",
    "              \" --min_cell_peak_thresh_ratio 0.0005\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              \" --lambda_chrom_access_recon 100.\" \\\n",
    "              \" --lambda_cat_covariates_contrastive 0.\" \\\n",
    "              \" --contrastive_logits_ratio 0.\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              \" --lambda_l1_masked 5.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01341dcb-64ad-45bd-8ef6-939e77d5b0f8",
   "metadata": {},
   "source": [
    "### 5.6 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "68c5b999-11e0-42e5-b886-8fb3a47d82f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12222799\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"xenium_human_breast_cancer\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 4 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-attention\"\n",
    "edge_batch_size = 4096 # 4096 (4, 8), 1024 (12)\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"2\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0. # 0., 5., 10.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "001945c5-6d59-4bdb-9efe-d8bbb33edc33",
   "metadata": {},
   "source": [
    "## 6. NicheCompass Query Mapping on Reference Model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14bf6b4d-34b5-4324-8deb-095bdfa7d0e2",
   "metadata": {},
   "source": [
    "### 6.1 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8f7f5f3-abe8-4518-8f7c-e4191ec3e6d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 12224333\n"
     ]
    }
   ],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_neighbors = 12\n",
    "n_hvg = 3000\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-attention\"\n",
    "edge_batch_size = 1024\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"4\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17761bc0-3178-42dd-9b0e-3935be8d2a5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_name_prefix = \"nichecompass_nanostring_cosmx_human_liver_query\"\n",
    "job_id = 1\n",
    "job_folder_path = \"/home/aih/sebastian.birk/workspace/projects/nichecompass-repro-new/slurm_jobs\"\n",
    "conda_env_name = \"nichecompass_hpc\"\n",
    "script_folder_path = \"/home/aih/sebastian.birk/workspace/projects/nichecompass-repro-new/scripts\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = \" --dataset nanostring_cosmx_human_liver\" \\\n",
    "              \" --query_batches sample2\" \\\n",
    "              \" --reference_batch sample1\" \\\n",
    "              \" --load_timestamp 10032023_145839\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp=20000\" \\\n",
    "              \" --n_epochs=40\" \\\n",
    "              \" --n_epochs_all_gps=0\" \\\n",
    "              \" --lambda_group_lasso=0.\" \\\n",
    "              \" --lambda_l1_masked=0.\" \\\n",
    "              \" --edge_batch_size=256\" \\\n",
    "              \" --node_batch_size=32\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb67b128-db16-4c2c-9e6f-b34fb390e5ea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
